// openclaw/dist/plugin-sdk/acp-runtime.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../utils-CS0Ikux6.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import "../message-channel-BliByQBl.js";
import "../store-_QDcIMYk.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import "../sessions-BwqyiEaq.js";
import "../paths-rhN9LKM_.js";
import "../session-write-lock-Db3SWNU3.js";
import { a as registerAcpRuntimeBackend, i as getAcpRuntimeBackend, n as readAcpSessionEntry, o as requireAcpRuntimeBackend, s as unregisterAcpRuntimeBackend, t as getAcpSessionManager, v as AcpRuntimeError, y as isAcpRuntimeError } from "../manager-DmufLZK2.js";
import "../acp-runtime-BS3aFfwU.js";
export { AcpRuntimeError, getAcpRuntimeBackend, getAcpSessionManager, isAcpRuntimeError, readAcpSessionEntry, registerAcpRuntimeBackend, requireAcpRuntimeBackend, unregisterAcpRuntimeBackend };
