// openclaw/dist/plugin-sdk/compat.js
import "../redact-BDinS1q9.js";
import "../errors-BxyFnvP3.js";
import "../unhandled-rejections-CDJ8dOVP.js";
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../warning-filter-CBhOcgHd.js";
import "../utils-CS0Ikux6.js";
import "../links-8xRhWBQL.js";
import "../paths-C4Oy7wjk.js";
import "../auth-profiles-CLtbIyEz.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import { o as ToolPolicySchema } from "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import { a as DmPolicySchema, c as GroupPolicySchema, m as MarkdownConfigSchema } from "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import "../file-lock-BJsnXQjL.js";
import "../audit-fs-DTQIjp2r.js";
import "../resolve-B7S4PlnO.js";
import "../profiles-B_nscSgH.js";
import "../tailscale-CxY-91GZ.js";
import "../tailnet-DHiKSG7V.js";
import "../net-JAg9zL0G.js";
import "../auth-Buq0Niri.js";
import "../credentials-D8pfRimu.js";
import "../message-channel-BliByQBl.js";
import "../store-_QDcIMYk.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import "../sessions-BwqyiEaq.js";
import "../paths-rhN9LKM_.js";
import "../session-write-lock-Db3SWNU3.js";
import "../method-scopes-B2vIWWxl.js";
import "../call-BhjE3oIo.js";
import "../prompt-style-HkzccPBT.js";
import "../ports-lsof-Dlzvm55V.js";
import "../restart-stale-pids-OP_Aybe0.js";
import "../ports-B3yeEL-n.js";
import "../logging-D4CVa7JL.js";
import "../commands-2T-1FAo7.js";
import "../issue-format-C8oDJazD.js";
import "../identity-D07jLUrU.js";
import "../heartbeat-B91tt3Lu.js";
import { $_ as createScopedAccountConfigAccessors, A_ as formatNormalizedAllowFromEntries, Av as collectOpenGroupPolicyRestrictSendersWarnings, Bv as createOpenGroupPolicyRestrictSendersWarningCollector, Dv as collectAllowlistProviderGroupPolicyWarnings, Ev as buildOpenGroupPolicyWarning, Fv as createAllowlistProviderGroupPolicyWarningCollector, Gv as projectConfigAccountIdWarningCollector, Hv as createOpenProviderGroupPolicyWarningCollector, Iv as createAllowlistProviderOpenWarningCollector, Kv as projectConfigWarningCollector, Lv as createAllowlistProviderRestrictSendersWarningCollector, Mv as collectOpenProviderGroupPolicyWarnings, N_ as mapAllowlistResolutionInputs, Nv as composeAccountWarningCollectors, Ov as collectAllowlistProviderRestrictSendersWarnings, Pv as composeWarningCollectors, Q_ as createHybridChannelConfigBase, Rv as createAllowlistProviderRouteAllowlistWarningCollector, Tv as buildOpenGroupPolicyRestrictSendersWarning, Uv as projectAccountConfigWarningCollector, Vv as createOpenProviderConfiguredRouteWarningCollector, Wv as projectAccountWarningCollector, Z_ as createHybridChannelConfigAdapter, ev as createScopedChannelConfigAdapter, iv as createTopLevelChannelConfigBase, jv as collectOpenGroupPolicyRouteAllowlistWarnings, k_ as formatAllowFromLowercase, nv as createScopedDmSecurityResolver, qv as projectWarningCollector, rv as createTopLevelChannelConfigAdapter, sb as delegateCompactionToRuntime, sv as mapAllowFromEntries, tv as createScopedChannelConfigBase, wv as buildOpenGroupPolicyConfigureRouteAllowlistWarning, yh as createRestrictSendersChannelSecurity, zv as createConditionalWarningCollector } from "../pi-embedded-CswW9luA.js";
import "../internal-hooks-ClBId-OS.js";
import "../multimodal-rLHaC7mC.js";
import "../memory-search-C_KRQGyn.js";
import "../provider-catalog-BGWDC_kb.js";
import "../secret-input-BOkqlW2S.js";
import "../bindings-DFhSj1AD.js";
import "../resolve-route-jw3TrvbG.js";
import "../routing-GIGwPHsU.js";
import "../identity-file-AtA39mSs.js";
import "../outbound-runtime-QVIwx-7l.js";
import "../provider-env-vars-BCe4Mf67.js";
import "../provider-auth-input-WYCWu6tD.js";
import "../provider-model-minimax-CYJqVplI.js";
import "../provider-models-ld3k2AX4.js";
import "../anthropic-vertex-provider-7stzcdx2.js";
import "../models-config.providers.discovery-DnQmXa4L.js";
import "../text-runtime-BXjeqyjp.js";
import "../tool-catalog-DP3BJZz3.js";
import "../docker-Cq_Zfc99.js";
import "../sandbox-Dt1ZOIQc.js";
import "../common-DArxwyX8.js";
import "../image-ops-339pOknK.js";
import "../thinking-rAawjv9l.js";
import "../path-alias-guards-DMcFUz5-.js";
import "../sandbox-paths-D-zPD6r4.js";
import "../channel-actions-DZIaeZzh.js";
import "../mime-C0asv0HQ.js";
import "../ssrf-BifXvxQR.js";
import "../fetch-guard-BTY_UfUY.js";
import "../provider-web-search-BHcSeZzc.js";
import "../manager-DmufLZK2.js";
import { t as KeyedAsyncQueue } from "../keyed-async-queue-s6sZRjxj.js";
import { t as emptyPluginConfigSchema } from "../config-schema-DOXyQvBE.js";
import { i as buildNestedDmConfigSchema, n as buildCatchallMultiAccountChannelSchema, r as buildChannelConfigSchema, t as AllowFromListSchema } from "../config-schema-CKb8N9O1.js";
import "../channel-plugin-common-BX9jRob2.js";
import { t as buildAccountScopedDmSecurityPolicy } from "../helpers-DvzONIyU.js";
import "../status-helpers-D9rzemcb.js";
import "../conversation-runtime-C56HP9yG.js";
import "../runtime-whatsapp-boundary-BMepWtnk.js";
import "../pairing-store-Bh2YkvRc.js";
import "../json-store-BIglSbRG.js";
import "../secret-file-CEppYIjz.js";
import "../channel-config-schema-CrubAncY.js";
import "../setup-binary-BLFhRd_P.js";
import "../archive-Df6iZdzY.js";
import "../fs-safe-B2bSW_OI.js";
import "../signal-cli-install-UDdZA4-C.js";
import "../setup-wizard-proxy--gFyw5Eh.js";
import "../setup-C0zuBL_6.js";
import { n as readStoreAllowFromForDmPolicy, o as resolveDmGroupAccessWithLists, s as resolveEffectiveAllowFromLists, t as DM_GROUP_ACCESS_REASON, u as resolveControlCommandGate } from "../dm-policy-shared-ChGQzSwJ.js";
import "../hook-runtime-sGC9reX7.js";
import { r as onDiagnosticEvent } from "../diagnostic-events-Cg5dNp82.js";
import "../diagnostic-BtBb10Ry.js";
import "../templating-Bxm_rtRG.js";
import "../channel-reply-pipeline-DGoktEKi.js";
import { a as buildPendingHistoryContextFromMap, c as evictOldHistoryKeys, i as buildHistoryContextFromMap, l as recordPendingHistoryEntry, n as buildHistoryContext, o as clearHistoryEntries, r as buildHistoryContextFromEntries, s as clearHistoryEntriesIfEnabled, t as DEFAULT_GROUP_HISTORY_LIMIT, u as recordPendingHistoryEntryIfEnabled } from "../reply-history-yYXk1ivL.js";
import "../commands-registry.data-_PzNk48W.js";
import "../commands-registry-ABOHrld2.js";
import "../frontmatter-Cd_6iJyw.js";
import "../env-overrides-B5iBORFt.js";
import "../skills-DAP6Yelo.js";
import "../skills-remote-BWQnXFJW.js";
import "../workspace-dirs-DhbPcAGw.js";
import "../pairing-token-CQLjZNmB.js";
import "../skill-commands-I_17sL0j.js";
import "../level-overrides-Ds8ppNbt.js";
import "../config-BnJ6TZJ3.js";
import "../routes-sPyjSxSY.js";
import "../ssh-tunnel-BSFzY8ID.js";
import "../server-middleware-Nz-5LEFG.js";
import "../logging-DFpZCDzJ.js";
import { c as resolveToolsBySender, o as resolveChannelGroupRequireMention, s as resolveChannelGroupToolsPolicy } from "../config-runtime-gAoemo4L.js";
import "../exec-approvals-CFIa5yg0.js";
import "../webhook-ingress-BBdXea0x.js";
import "../system-events-C2ngfmKm.js";
import { t as createRuntimeDirectoryLiveAdapter } from "../runtime-forwarders-B4Fz8OBx.js";
import "../ssrf-policy-BlgZt63j.js";
import "../provider-auth-ref-Dd8Ctvr9.js";
import "../provider-auth-helpers-CK3uH19P.js";
import "../provider-api-key-auth-CyO5NBeZ.js";
import "../runtime-env-CAlGDNNR.js";
import "../pairing-labels-BAkUxk5K.js";
import { a as listDirectoryGroupEntriesFromMapKeysAndAllowFrom, c as listInspectedDirectoryEntriesFromSources, d as listResolvedDirectoryUserEntriesFromAllowFrom, f as toDirectoryEntries, g as nullChannelDirectorySelf, h as emptyChannelDirectoryList, i as listDirectoryGroupEntriesFromMapKeys, l as listResolvedDirectoryEntriesFromSources, m as createEmptyChannelDirectoryAdapter, n as collectNormalizedDirectoryIds, o as listDirectoryUserEntriesFromAllowFrom, p as createChannelDirectoryAdapter, r as listDirectoryEntriesFromSources, s as listDirectoryUserEntriesFromAllowFromAndMapKeys, t as applyDirectoryQueryAndLimit, u as listResolvedDirectoryGroupEntriesFromMapKeys } from "../directory-runtime-BgPR38m_.js";
import { t as inspectReadOnlyChannelAccount } from "../read-only-account-inspect-B8C32NVU.js";
import "../src-Bc4Fm2Ge.js";
import "../web-media-Bo_Nxwjv.js";
import "../temp-path-Dgv6Kihk.js";
import "../media-understanding-Ch_w554d.js";
import "../web-media-BvqWsq_l.js";
import "../state-paths-C-WJBTWG.js";
import "../llm-task-Djjqflw_.js";
import "../pi-model-discovery-BNyYuKdC.js";
import "../exec-inline-eval-Di27nCtK.js";
import "../target-registry-CofRQwTE.js";
import "../external-content-Bq0zNXEh.js";
import "../brave-Hgb0hhUH.js";
import "../duckduckgo-B0xgAU7B.js";
import "../exa-Ce69v825.js";
import "../security-runtime-CoZ_sfgg.js";
import "../provider-usage-qQL8PJ0p.js";
import "../provider-onboard-CqAihpGi.js";
import "../perplexity-BNVyHCAP.js";
import "../stagger-DZXzssdr.js";
import "../command-secret-targets-wRPHRRib.js";
import "../delivery-queue-1A7pbB9e.js";
import "../channel-summary-DHY50TCU.js";
import "../follow-redirects-0oww1mdf.js";
import "../tool-policy-match-DnEs8-IP.js";
import "../runtime-CUG7pnyw.js";
import "../channel-status-C_NTcHvs.js";
import "../discord-core-DpMUVJg9.js";
import "../cli-runtime-CYhFvgcO.js";
import "../config-presence-Cqse_N2e.js";
import "../query-expansion-CZbIeYUY.js";
import "../search-manager-DsKEM_vh.js";
import "../acp-runtime-BS3aFfwU.js";
import "../telegram-core-BFUHv9d-.js";
import "../audit--q0z8yGd.js";
import "../gateway-runtime-Czf7LoCY.js";
import "../connection-auth-ZVYO9ARZ.js";
import { t as createAccountStatusSink } from "../channel-lifecycle-CGlugnkL.js";
import "../mcp-config-CpCZwCwq.js";
import { t as createPluginRuntimeStore } from "../runtime-store-AF98vbqz.js";
import { n as resolveBlueBubblesGroupRequireMention, r as resolveBlueBubblesGroupToolPolicy, t as collectBlueBubblesStatusIssues } from "../bluebubbles-BajuLDlM.js";
//#region src/plugin-sdk/compat.ts
if (process.env.VITEST !== "true" && process.env.OPENCLAW_SUPPRESS_PLUGIN_SDK_COMPAT_WARNING !== "1") process.emitWarning("openclaw/plugin-sdk/compat is deprecated for new plugins. Migrate to focused openclaw/plugin-sdk/<subpath> imports. See https://docs.openclaw.ai/plugins/sdk-migration", {
	code: "OPENCLAW_PLUGIN_SDK_COMPAT_DEPRECATED",
	detail: "Bundled plugins must use scoped plugin-sdk subpaths. External plugins may keep compat temporarily while migrating. Migration guide: https://docs.openclaw.ai/plugins/sdk-migration"
});
//#endregion
export { AllowFromListSchema, DEFAULT_GROUP_HISTORY_LIMIT, DM_GROUP_ACCESS_REASON, DmPolicySchema, GroupPolicySchema, KeyedAsyncQueue, MarkdownConfigSchema, ToolPolicySchema, applyDirectoryQueryAndLimit, buildAccountScopedDmSecurityPolicy, buildCatchallMultiAccountChannelSchema, buildChannelConfigSchema, buildHistoryContext, buildHistoryContextFromEntries, buildHistoryContextFromMap, buildNestedDmConfigSchema, buildOpenGroupPolicyConfigureRouteAllowlistWarning, buildOpenGroupPolicyRestrictSendersWarning, buildOpenGroupPolicyWarning, buildPendingHistoryContextFromMap, clearHistoryEntries, clearHistoryEntriesIfEnabled, collectAllowlistProviderGroupPolicyWarnings, collectAllowlistProviderRestrictSendersWarnings, collectBlueBubblesStatusIssues, collectNormalizedDirectoryIds, collectOpenGroupPolicyRestrictSendersWarnings, collectOpenGroupPolicyRouteAllowlistWarnings, collectOpenProviderGroupPolicyWarnings, composeAccountWarningCollectors, composeWarningCollectors, createAccountStatusSink, createAllowlistProviderGroupPolicyWarningCollector, createAllowlistProviderOpenWarningCollector, createAllowlistProviderRestrictSendersWarningCollector, createAllowlistProviderRouteAllowlistWarningCollector, createChannelDirectoryAdapter, createConditionalWarningCollector, createEmptyChannelDirectoryAdapter, createHybridChannelConfigAdapter, createHybridChannelConfigBase, createOpenGroupPolicyRestrictSendersWarningCollector, createOpenProviderConfiguredRouteWarningCollector, createOpenProviderGroupPolicyWarningCollector, createPluginRuntimeStore, createRestrictSendersChannelSecurity, createRuntimeDirectoryLiveAdapter, createScopedAccountConfigAccessors, createScopedChannelConfigAdapter, createScopedChannelConfigBase, createScopedDmSecurityResolver, createTopLevelChannelConfigAdapter, createTopLevelChannelConfigBase, delegateCompactionToRuntime, emptyChannelDirectoryList, emptyPluginConfigSchema, evictOldHistoryKeys, formatAllowFromLowercase, formatNormalizedAllowFromEntries, inspectReadOnlyChannelAccount, listDirectoryEntriesFromSources, listDirectoryGroupEntriesFromMapKeys, listDirectoryGroupEntriesFromMapKeysAndAllowFrom, listDirectoryUserEntriesFromAllowFrom, listDirectoryUserEntriesFromAllowFromAndMapKeys, listInspectedDirectoryEntriesFromSources, listResolvedDirectoryEntriesFromSources, listResolvedDirectoryGroupEntriesFromMapKeys, listResolvedDirectoryUserEntriesFromAllowFrom, mapAllowFromEntries, mapAllowlistResolutionInputs, nullChannelDirectorySelf, onDiagnosticEvent, projectAccountConfigWarningCollector, projectAccountWarningCollector, projectConfigAccountIdWarningCollector, projectConfigWarningCollector, projectWarningCollector, readStoreAllowFromForDmPolicy, recordPendingHistoryEntry, recordPendingHistoryEntryIfEnabled, resolveBlueBubblesGroupRequireMention, resolveBlueBubblesGroupToolPolicy, resolveChannelGroupRequireMention, resolveChannelGroupToolsPolicy, resolveControlCommandGate, resolveDmGroupAccessWithLists, resolveEffectiveAllowFromLists, resolveToolsBySender, toDirectoryEntries };
