// openclaw/dist/plugin-sdk/config-runtime.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../utils-CS0Ikux6.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import { Mt as resolveActiveTalkProviderConfig, Nt as resolveAgentMaxConcurrent, a as getRuntimeConfigSnapshot, f as readConfigFileSnapshotForWrite, g as writeConfigFile, s as loadConfig } from "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import { i as coerceSecretRef } from "../types.secrets-DgX397vI.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import { A as resolveSlackNativeStreaming, M as resolveTelegramPreviewStreamMode, O as mapStreamingModeToSlackLegacyDraftStreamMode, d as normalizeTelegramCommandName, f as resolveTelegramCustomCommands, j as resolveSlackStreamingMode, k as resolveDiscordPreviewStreamMode, u as TELEGRAM_COMMAND_NAME_PATTERN } from "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import "../message-channel-BliByQBl.js";
import { A as resolveGroupSessionKey, a as recordSessionMetaFromInbound, c as updateLastRoute, i as readSessionUpdatedAt, l as updateSessionStore, n as loadSessionStore, o as resolveSessionStoreEntry } from "../store-_QDcIMYk.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import { _ as resolveChannelResetConfig, b as resolveThreadFlag, g as evaluateSessionFreshness, m as resolveSessionKey, v as resolveSessionResetPolicy, y as resolveSessionResetType } from "../sessions-BwqyiEaq.js";
import { l as resolveStorePath } from "../paths-rhN9LKM_.js";
import "../session-write-lock-Db3SWNU3.js";
import { a as resolveNativeSkillsEnabled, i as resolveNativeCommandsEnabled, n as isNativeCommandsExplicitlyDisabled } from "../commands-2T-1FAo7.js";
import "../issue-format-C8oDJazD.js";
import { a as warnMissingProviderGroupPolicyFallbackOnce, i as resolveOpenProviderRuntimeGroupPolicy, n as resolveAllowlistProviderRuntimeGroupPolicy, r as resolveDefaultGroupPolicy, t as GROUP_POLICY_BLOCKED_LABEL } from "../runtime-group-policy-DqDSPQSR.js";
import { n as logConfigUpdated } from "../logging-DFpZCDzJ.js";
import { a as resolveChannelGroupPolicy, l as resolveMarkdownTableMode, n as resolveCronStorePath, o as resolveChannelGroupRequireMention, r as saveCronStore, t as loadCronStore, x as updateConfig } from "../config-runtime-gAoemo4L.js";
import { t as applyModelOverrideToSessionEntry } from "../model-overrides-BN8miQfK.js";
import { n as isDangerousNameMatchingEnabled, r as resolveDangerousNameMatchingEnabled } from "../dangerous-name-matching-TMkiuJ0K.js";
export { GROUP_POLICY_BLOCKED_LABEL, TELEGRAM_COMMAND_NAME_PATTERN, applyModelOverrideToSessionEntry, coerceSecretRef, evaluateSessionFreshness, getRuntimeConfigSnapshot, isDangerousNameMatchingEnabled, isNativeCommandsExplicitlyDisabled, loadConfig, loadCronStore, loadSessionStore, logConfigUpdated, mapStreamingModeToSlackLegacyDraftStreamMode, normalizeTelegramCommandName, readConfigFileSnapshotForWrite, readSessionUpdatedAt, recordSessionMetaFromInbound, resolveActiveTalkProviderConfig, resolveAgentMaxConcurrent, resolveAllowlistProviderRuntimeGroupPolicy, resolveChannelGroupPolicy, resolveChannelGroupRequireMention, resolveChannelResetConfig, resolveCronStorePath, resolveDangerousNameMatchingEnabled, resolveDefaultGroupPolicy, resolveDiscordPreviewStreamMode, resolveGroupSessionKey, resolveMarkdownTableMode, resolveNativeCommandsEnabled, resolveNativeSkillsEnabled, resolveOpenProviderRuntimeGroupPolicy, resolveSessionKey, resolveSessionResetPolicy, resolveSessionResetType, resolveSessionStoreEntry, resolveSlackNativeStreaming, resolveSlackStreamingMode, resolveStorePath, resolveTelegramCustomCommands, resolveTelegramPreviewStreamMode, resolveThreadFlag, saveCronStore, updateConfig, updateLastRoute, updateSessionStore, warnMissingProviderGroupPolicyFallbackOnce, writeConfigFile };
