// openclaw/dist/plugin-sdk/discord-core.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import { n as DiscordConfigSchema } from "../zod-schema.providers-core-COy5nBQ0.js";
import { r as getChatChannelMeta } from "../registry-C5UkPpaO.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import { a as withNormalizedTimestamp } from "../date-time-BwTq3C5B.js";
import { c as jsonResult, d as readNumberParam, f as readReactionParams, h as readStringParam, l as parseAvailableTags, p as readStringArrayParam } from "../common-DArxwyX8.js";
import "../image-ops-339pOknK.js";
import "../path-alias-guards-DMcFUz5-.js";
import { t as assertMediaNotDataUrl } from "../sandbox-paths-D-zPD6r4.js";
import "../mime-C0asv0HQ.js";
import { r as buildChannelConfigSchema } from "../config-schema-CKb8N9O1.js";
import "../channel-plugin-common-BX9jRob2.js";
import { r as resolvePollMaxSelections } from "../polls-GziVvR8X.js";
import "../discord-core-DpMUVJg9.js";
export { DiscordConfigSchema, assertMediaNotDataUrl, buildChannelConfigSchema, getChatChannelMeta, jsonResult, parseAvailableTags, readNumberParam, readReactionParams, readStringArrayParam, readStringParam, resolvePollMaxSelections, withNormalizedTimestamp };
