// openclaw/dist/plugin-sdk/infra-runtime.js
import "../redact-BDinS1q9.js";
import { a as hasErrnoCode, i as formatUncaughtError, n as extractErrorCode, o as isErrno, r as formatErrorMessage, s as readErrorName, t as collectErrorGraphCandidates } from "../errors-BxyFnvP3.js";
import "../unhandled-rejections-CDJ8dOVP.js";
import "../logger-kwZIqwuw.js";
import { C as resolveRequiredHomeDir, S as resolveOsHomeRelativePath, b as resolveHomeRelativePath, v as expandHomePrefix, w as resolveRequiredOsHomeDir, x as resolveOsHomeDir, y as resolveEffectiveHomeDir } from "../paths-ViKUYWUK.js";
import { n as resolvePreferredOpenClawTmpDir, t as POSIX_OPENCLAW_TMP_DIR } from "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import { i as normalizeZaiEnv, n as logAcceptedEnvOption, r as normalizeEnv, t as isTruthyEnvValue } from "../env-Dnra1IpT.js";
import "../warning-filter-CBhOcgHd.js";
import "../utils-CS0Ikux6.js";
import "../links-8xRhWBQL.js";
import "../paths-C4Oy7wjk.js";
import "../auth-profiles-CLtbIyEz.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import { $ as splitCommandChainWithOperators, G as analyzeArgvCommand, J as buildSafeBinsShellCommand, K as analyzeShellCommand, Q as splitCommandChain, X as isWindowsPlatform, Y as buildSafeShellCommand, Z as resolvePlannedSegmentArgv, et as matchAllowlist, it as resolveCommandResolutionFromArgv, nt as resolveAllowlistCandidatePath, q as buildEnforcedShellCommand, rt as resolveCommandResolution, tt as parseExecArgvToken, yt as DEFAULT_SAFE_BINS } from "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import { C as normalizeScpRemotePath, S as normalizeScpRemoteHost, b as isSafeScpRemoteHost, x as isSafeScpRemotePath } from "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import { n as resetFileLockStateForTest, r as withFileLock, t as acquireFileLock } from "../file-lock-BJsnXQjL.js";
import "../audit-fs-DTQIjp2r.js";
import "../resolve-B7S4PlnO.js";
import "../profiles-B_nscSgH.js";
import "../tailscale-CxY-91GZ.js";
import "../tailnet-DHiKSG7V.js";
import "../net-JAg9zL0G.js";
import "../auth-Buq0Niri.js";
import "../credentials-D8pfRimu.js";
import "../message-channel-BliByQBl.js";
import "../store-_QDcIMYk.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import "../sessions-BwqyiEaq.js";
import "../paths-rhN9LKM_.js";
import "../session-write-lock-Db3SWNU3.js";
import { i as writeTextAtomic, n as readJsonFile, r as writeJsonAtomic, t as createAsyncLock } from "../json-files-CW7rEs_Q.js";
import "../method-scopes-B2vIWWxl.js";
import "../call-BhjE3oIo.js";
import { a as resetWSLStateForTests, i as isWSLSync, n as isWSL2Sync, r as isWSLEnv, t as isWSL } from "../wsl-b7s-IOwT.js";
import "../prompt-style-HkzccPBT.js";
import "../ports-lsof-Dlzvm55V.js";
import "../restart-stale-pids-OP_Aybe0.js";
import "../ports-B3yeEL-n.js";
import "../logging-D4CVa7JL.js";
import "../commands-2T-1FAo7.js";
import "../issue-format-C8oDJazD.js";
import "../identity-D07jLUrU.js";
import "../heartbeat-B91tt3Lu.js";
import { $h as PROXY_FETCH_PROXY_URL, Cg as sanitizeExecApprovalDisplayText, Dg as getChannelActivity, Eg as resolveDiagnosticFlags, Gh as waitForTransportReady, Jh as createTelegramRetryRunner, Kh as TELEGRAM_RETRY_DEFAULTS, Og as recordChannelActivity, Qh as resetGlobalUndiciStreamTimeoutsForTests, Sg as resolveExecApprovalCommandDisplay, Tg as matchesDiagnosticFlag, Xh as ensureGlobalUndiciEnvProxyDispatcher, Yh as DEFAULT_UNDICI_STREAM_TIMEOUT_MS, Zh as ensureGlobalUndiciStreamTimeouts, ag as getLastHeartbeatEvent, bg as getExecApprovalApproverDmNoticeText, cg as resolveIndicatorType, dg as resolveExecApprovalSessionTarget, eg as getProxyUrlFromFetch, ig as emitHeartbeatEvent, kg as resetChannelActivityForTest, lg as resolveFetch, ng as resolveProxyFetchFromEnv, og as onHeartbeatEvent, qh as createRateLimitRetryRunner, rg as resolveHeartbeatVisibility, sg as resetHeartbeatEventsForTest, tg as makeProxyFetch, ug as wrapFetchWithAbortSignal, vg as buildExecApprovalPendingReplyPayload, wg as isDiagnosticFlagEnabled, xg as getExecApprovalReplyMetadata, yg as buildExecApprovalUnavailableReplyPayload } from "../pi-embedded-CswW9luA.js";
import "../internal-hooks-ClBId-OS.js";
import { n as generateSecureUuid, t as generateSecureToken } from "../secure-random-Dgumttcn.js";
import "../multimodal-rLHaC7mC.js";
import "../memory-search-C_KRQGyn.js";
import "../provider-catalog-BGWDC_kb.js";
import "../secret-input-BOkqlW2S.js";
import "../bindings-DFhSj1AD.js";
import "../resolve-route-jw3TrvbG.js";
import "../routing-GIGwPHsU.js";
import "../identity-file-AtA39mSs.js";
import { n as resolveAgentOutboundIdentity, r as resolveOutboundSendDep, t as normalizeOutboundIdentity } from "../outbound-runtime-QVIwx-7l.js";
import "../provider-env-vars-BCe4Mf67.js";
import "../provider-auth-input-WYCWu6tD.js";
import "../provider-model-minimax-CYJqVplI.js";
import { D as resolveRetryConfig, O as retryAsync } from "../provider-models-ld3k2AX4.js";
import "../anthropic-vertex-provider-7stzcdx2.js";
import "../models-config.providers.discovery-DnQmXa4L.js";
import "../text-runtime-BXjeqyjp.js";
import "../tool-catalog-DP3BJZz3.js";
import "../docker-Cq_Zfc99.js";
import "../sandbox-Dt1ZOIQc.js";
import "../common-DArxwyX8.js";
import "../image-ops-339pOknK.js";
import "../thinking-rAawjv9l.js";
import { a as trySafeFileURLToPath, i as safeFileURLToPath, n as basenameFromMediaSource, r as isWindowsNetworkPath, t as assertNoWindowsNetworkPath } from "../local-file-access-BdDE0bQ3.js";
import "../path-alias-guards-DMcFUz5-.js";
import "../sandbox-paths-D-zPD6r4.js";
import "../channel-actions-DZIaeZzh.js";
import "../mime-C0asv0HQ.js";
import { i as resolveEnvHttpProxyUrl, n as hasEnvHttpProxyConfigured, r as hasProxyEnvConfigured, t as PROXY_ENV_KEYS } from "../proxy-env-Bp6SXFEY.js";
import { a as createPinnedLookup, c as isPrivateIpAddress, d as resolvePinnedHostnameWithPolicy, i as createPinnedDispatcher, l as isPrivateNetworkAllowedByPolicy, n as assertPublicHostname, o as isBlockedHostname, p as normalizeHostname, r as closeDispatcher, s as isBlockedHostnameOrIp, t as SsrFBlockedError, u as resolvePinnedHostname } from "../ssrf-BifXvxQR.js";
import { i as withTrustedEnvProxyGuardedFetchMode, n as fetchWithSsrFGuard, r as withStrictGuardedFetchMode, t as GUARDED_FETCH_MODE } from "../fetch-guard-BTY_UfUY.js";
import "../provider-web-search-BHcSeZzc.js";
import "../manager-DmufLZK2.js";
import "../config-schema-CKb8N9O1.js";
import "../channel-plugin-common-BX9jRob2.js";
import "../status-helpers-D9rzemcb.js";
import "../conversation-runtime-C56HP9yG.js";
import "../runtime-whatsapp-boundary-BMepWtnk.js";
import "../pairing-store-Bh2YkvRc.js";
import "../json-store-BIglSbRG.js";
import { i as tryReadSecretFileSync, n as loadSecretFileSync, r as readSecretFileSync, t as DEFAULT_SECRET_FILE_MAX_BYTES } from "../secret-file-CEppYIjz.js";
import "../channel-config-schema-CrubAncY.js";
import "../setup-binary-BLFhRd_P.js";
import "../archive-Df6iZdzY.js";
import { a as openFileWithinRoot, c as readLocalFileSafely, d as writeFileFromPathWithinRoot, f as writeFileWithinRoot, i as createRootScopedReadFile, l as readPathWithinRoot, n as appendFileWithinRoot, o as openWritableFileWithinRoot, r as copyFileWithinRoot, s as readFileWithinRoot, t as SafeOpenError, u as resolveOpenedFileRealPathForHandle } from "../fs-safe-B2bSW_OI.js";
import "../signal-cli-install-UDdZA4-C.js";
import "../setup-wizard-proxy--gFyw5Eh.js";
import "../setup-C0zuBL_6.js";
import "../dm-policy-shared-ChGQzSwJ.js";
import "../hook-runtime-sGC9reX7.js";
import { i as resetDiagnosticEventsForTest, n as isDiagnosticsEnabled, r as onDiagnosticEvent, t as emitDiagnosticEvent } from "../diagnostic-events-Cg5dNp82.js";
import "../diagnostic-BtBb10Ry.js";
import { t as pruneMapToMaxSize } from "../map-size-8PXXQPcj.js";
import { n as resolveGlobalDedupeCache, t as createDedupeCache } from "../dedupe-CCJJrryE.js";
import "../templating-Bxm_rtRG.js";
import "../channel-reply-pipeline-DGoktEKi.js";
import "../reply-history-yYXk1ivL.js";
import "../commands-registry.data-_PzNk48W.js";
import "../commands-registry-ABOHrld2.js";
import "../frontmatter-Cd_6iJyw.js";
import "../env-overrides-B5iBORFt.js";
import "../skills-DAP6Yelo.js";
import "../skills-remote-BWQnXFJW.js";
import "../workspace-dirs-DhbPcAGw.js";
import "../pairing-token-CQLjZNmB.js";
import "../skill-commands-I_17sL0j.js";
import { f as computeBackoff, p as sleepWithAbort } from "../level-overrides-Ds8ppNbt.js";
import "../config-BnJ6TZJ3.js";
import "../routes-sPyjSxSY.js";
import "../ssh-tunnel-BSFzY8ID.js";
import "../server-middleware-Nz-5LEFG.js";
import "../logging-DFpZCDzJ.js";
import "../config-runtime-gAoemo4L.js";
import { C as isSafeBinUsage, E as resolveSafeBins, S as evaluateShellAllowlist, T as resolveAllowAlwaysPatterns, _ as resolveExecApprovalsFromFile, a as maxAsk, b as saveExecApprovals, c as normalizeExecApprovals, d as normalizeExecSecurity, f as readExecApprovalsSnapshot, g as resolveExecApprovals, h as requiresExecApproval, i as loadExecApprovals, l as normalizeExecAsk, m as requestExecApprovalViaSocket, n as addAllowlistEntry, o as mergeExecApprovalsSocketDefaults, p as recordAllowlistUse, r as ensureExecApprovals, s as minSecurity, t as DEFAULT_EXEC_APPROVAL_TIMEOUT_MS, u as normalizeExecHost, v as resolveExecApprovalsPath, w as normalizeSafeBins, x as evaluateExecAllowlist, y as resolveExecApprovalsSocketPath } from "../exec-approvals-CFIa5yg0.js";
import { i as formatDurationSeconds, n as formatDurationHuman, r as formatDurationPrecise, t as formatDurationCompact } from "../format-duration-gUgoY3JD.js";
import { A as DEFAULT_WEBHOOK_MAX_BODY_BYTES, F as readRequestBodyWithLimit, I as requestBodyErrorToText, M as installRequestBodyLimitGuard, N as isRequestBodyLimitError, P as readJsonBodyWithLimit, j as RequestBodyLimitError, k as DEFAULT_WEBHOOK_BODY_TIMEOUT_MS } from "../webhook-ingress-BBdXea0x.js";
import { a as isSystemEventContextChanged, c as resetSystemEventsForTest, i as hasSystemEvents, n as drainSystemEvents, o as peekSystemEventEntries, r as enqueueSystemEvent, s as peekSystemEvents, t as drainSystemEventEntries } from "../system-events-C2ngfmKm.js";
import { n as hasSystemMark, r as prefixSystemMessage, t as SYSTEM_MARK } from "../system-message-CbYLzyfL.js";
import { n as createRuntimeOutboundDelegates } from "../runtime-forwarders-B4Fz8OBx.js";
import { a as ssrfPolicyFromAllowPrivateNetwork, i as normalizeHostnameSuffixAllowlist, n as buildHostnameAllowlistPolicyFromSuffixAllowlist, r as isHttpsUrlAllowedByHostnameSuffixAllowlist, t as assertHttpUrlTargetsPrivateNetwork } from "../ssrf-policy-BlgZt63j.js";
import "../provider-auth-ref-Dd8Ctvr9.js";
import "../provider-auth-helpers-CK3uH19P.js";
import "../provider-api-key-auth-CyO5NBeZ.js";
import "../runtime-env-CAlGDNNR.js";
import "../pairing-labels-BAkUxk5K.js";
import "../directory-runtime-BgPR38m_.js";
import "../read-only-account-inspect-B8C32NVU.js";
import "../src-Bc4Fm2Ge.js";
import "../web-media-Bo_Nxwjv.js";
import "../temp-path-Dgv6Kihk.js";
import "../media-understanding-Ch_w554d.js";
import "../web-media-BvqWsq_l.js";
import "../state-paths-C-WJBTWG.js";
import "../llm-task-Djjqflw_.js";
import "../pi-model-discovery-BNyYuKdC.js";
import "../exec-inline-eval-Di27nCtK.js";
import "../target-registry-CofRQwTE.js";
import "../external-content-Bq0zNXEh.js";
import "../brave-Hgb0hhUH.js";
import "../duckduckgo-B0xgAU7B.js";
import "../exa-Ce69v825.js";
import "../security-runtime-CoZ_sfgg.js";
import "../provider-usage-qQL8PJ0p.js";
import "../provider-onboard-CqAihpGi.js";
import "../perplexity-BNVyHCAP.js";
import "../stagger-DZXzssdr.js";
import "../command-secret-targets-wRPHRRib.js";
import "../delivery-queue-1A7pbB9e.js";
import "../channel-summary-DHY50TCU.js";
import "../follow-redirects-0oww1mdf.js";
import "../tool-policy-match-DnEs8-IP.js";
import "../runtime-CUG7pnyw.js";
import "../channel-status-C_NTcHvs.js";
import "../discord-core-DpMUVJg9.js";
import "../cli-runtime-CYhFvgcO.js";
import "../config-presence-Cqse_N2e.js";
import "../query-expansion-CZbIeYUY.js";
import "../search-manager-DsKEM_vh.js";
import "../acp-runtime-BS3aFfwU.js";
import "../telegram-core-BFUHv9d-.js";
import "../audit--q0z8yGd.js";
import "../gateway-runtime-Czf7LoCY.js";
import "../connection-auth-ZVYO9ARZ.js";
import "../mcp-config-CpCZwCwq.js";
export { DEFAULT_EXEC_APPROVAL_TIMEOUT_MS, DEFAULT_SAFE_BINS, DEFAULT_SECRET_FILE_MAX_BYTES, DEFAULT_UNDICI_STREAM_TIMEOUT_MS, DEFAULT_WEBHOOK_BODY_TIMEOUT_MS, DEFAULT_WEBHOOK_MAX_BODY_BYTES, GUARDED_FETCH_MODE, POSIX_OPENCLAW_TMP_DIR, PROXY_ENV_KEYS, PROXY_FETCH_PROXY_URL, RequestBodyLimitError, SYSTEM_MARK, SafeOpenError, SsrFBlockedError, TELEGRAM_RETRY_DEFAULTS, acquireFileLock, addAllowlistEntry, analyzeArgvCommand, analyzeShellCommand, appendFileWithinRoot, assertHttpUrlTargetsPrivateNetwork, assertNoWindowsNetworkPath, assertPublicHostname, basenameFromMediaSource, buildEnforcedShellCommand, buildExecApprovalPendingReplyPayload, buildExecApprovalUnavailableReplyPayload, buildHostnameAllowlistPolicyFromSuffixAllowlist, buildSafeBinsShellCommand, buildSafeShellCommand, closeDispatcher, collectErrorGraphCandidates, computeBackoff, copyFileWithinRoot, createAsyncLock, createDedupeCache, createPinnedDispatcher, createPinnedLookup, createRateLimitRetryRunner, createRootScopedReadFile, createRuntimeOutboundDelegates, createTelegramRetryRunner, drainSystemEventEntries, drainSystemEvents, emitDiagnosticEvent, emitHeartbeatEvent, enqueueSystemEvent, ensureExecApprovals, ensureGlobalUndiciEnvProxyDispatcher, ensureGlobalUndiciStreamTimeouts, evaluateExecAllowlist, evaluateShellAllowlist, expandHomePrefix, extractErrorCode, fetchWithSsrFGuard, formatDurationCompact, formatDurationHuman, formatDurationPrecise, formatDurationSeconds, formatErrorMessage, formatUncaughtError, generateSecureToken, generateSecureUuid, getChannelActivity, getExecApprovalApproverDmNoticeText, getExecApprovalReplyMetadata, getLastHeartbeatEvent, getProxyUrlFromFetch, hasEnvHttpProxyConfigured, hasErrnoCode, hasProxyEnvConfigured, hasSystemEvents, hasSystemMark, installRequestBodyLimitGuard, isBlockedHostname, isBlockedHostnameOrIp, isDiagnosticFlagEnabled, isDiagnosticsEnabled, isErrno, isHttpsUrlAllowedByHostnameSuffixAllowlist, isPrivateIpAddress, isPrivateNetworkAllowedByPolicy, isRequestBodyLimitError, isSafeBinUsage, isSafeScpRemoteHost, isSafeScpRemotePath, isSystemEventContextChanged, isTruthyEnvValue, isWSL, isWSL2Sync, isWSLEnv, isWSLSync, isWindowsNetworkPath, isWindowsPlatform, loadExecApprovals, loadSecretFileSync, logAcceptedEnvOption, makeProxyFetch, matchAllowlist, matchesDiagnosticFlag, maxAsk, mergeExecApprovalsSocketDefaults, minSecurity, normalizeEnv, normalizeExecApprovals, normalizeExecAsk, normalizeExecHost, normalizeExecSecurity, normalizeHostname, normalizeHostnameSuffixAllowlist, normalizeOutboundIdentity, normalizeSafeBins, normalizeScpRemoteHost, normalizeScpRemotePath, normalizeZaiEnv, onDiagnosticEvent, onHeartbeatEvent, openFileWithinRoot, openWritableFileWithinRoot, parseExecArgvToken, peekSystemEventEntries, peekSystemEvents, prefixSystemMessage, pruneMapToMaxSize, readErrorName, readExecApprovalsSnapshot, readFileWithinRoot, readJsonBodyWithLimit, readJsonFile, readLocalFileSafely, readPathWithinRoot, readRequestBodyWithLimit, readSecretFileSync, recordAllowlistUse, recordChannelActivity, requestBodyErrorToText, requestExecApprovalViaSocket, requiresExecApproval, resetChannelActivityForTest, resetDiagnosticEventsForTest, resetFileLockStateForTest, resetGlobalUndiciStreamTimeoutsForTests, resetHeartbeatEventsForTest, resetSystemEventsForTest, resetWSLStateForTests, resolveAgentOutboundIdentity, resolveAllowAlwaysPatterns, resolveAllowlistCandidatePath, resolveCommandResolution, resolveCommandResolutionFromArgv, resolveDiagnosticFlags, resolveEffectiveHomeDir, resolveEnvHttpProxyUrl, resolveExecApprovalCommandDisplay, resolveExecApprovalSessionTarget, resolveExecApprovals, resolveExecApprovalsFromFile, resolveExecApprovalsPath, resolveExecApprovalsSocketPath, resolveFetch, resolveGlobalDedupeCache, resolveHeartbeatVisibility, resolveHomeRelativePath, resolveIndicatorType, resolveOpenedFileRealPathForHandle, resolveOsHomeDir, resolveOsHomeRelativePath, resolveOutboundSendDep, resolvePinnedHostname, resolvePinnedHostnameWithPolicy, resolvePlannedSegmentArgv, resolvePreferredOpenClawTmpDir, resolveProxyFetchFromEnv, resolveRequiredHomeDir, resolveRequiredOsHomeDir, resolveRetryConfig, resolveSafeBins, retryAsync, safeFileURLToPath, sanitizeExecApprovalDisplayText, saveExecApprovals, sleepWithAbort, splitCommandChain, splitCommandChainWithOperators, ssrfPolicyFromAllowPrivateNetwork, tryReadSecretFileSync, trySafeFileURLToPath, waitForTransportReady, withFileLock, withStrictGuardedFetchMode, withTrustedEnvProxyGuardedFetchMode, wrapFetchWithAbortSignal, writeFileFromPathWithinRoot, writeFileWithinRoot, writeJsonAtomic, writeTextAtomic };
