// openclaw/dist/plugin-sdk/matrix.js
import { n as redactSensitiveText } from "../redact-BDinS1q9.js";
import "../errors-BxyFnvP3.js";
import "../unhandled-rejections-CDJ8dOVP.js";
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../warning-filter-CBhOcgHd.js";
import "../utils-CS0Ikux6.js";
import { t as formatDocsLink } from "../links-8xRhWBQL.js";
import "../paths-C4Oy7wjk.js";
import "../auth-profiles-CLtbIyEz.js";
import { _ as normalizeAccountId, g as DEFAULT_ACCOUNT_ID, u as resolveAgentIdFromSessionKey, v as normalizeOptionalAccountId } from "../session-key-DAhnzjyr.js";
import { r as normalizeStringEntries } from "../string-normalization-CohoSMRS.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import { a as hasConfiguredSecretInput, c as normalizeResolvedSecretInputString, l as normalizeSecretInputString } from "../types.secrets-DgX397vI.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import { o as ToolPolicySchema } from "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import { m as MarkdownConfigSchema } from "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import "../file-lock-BJsnXQjL.js";
import "../audit-fs-DTQIjp2r.js";
import "../resolve-B7S4PlnO.js";
import "../profiles-B_nscSgH.js";
import "../tailscale-CxY-91GZ.js";
import "../tailnet-DHiKSG7V.js";
import { i as isPrivateOrLoopbackHost } from "../net-JAg9zL0G.js";
import "../auth-Buq0Niri.js";
import "../credentials-D8pfRimu.js";
import "../message-channel-BliByQBl.js";
import "../store-_QDcIMYk.js";
import "../runtime-Iz8uZ7EU.js";
import { i as resolveAllowlistMatchByCandidates, n as formatAllowlistMatchMeta, o as resolveCompiledAllowlistMatch, r as resolveAllowlistCandidates, t as compileAllowlist } from "../plugins-B09-vgme.js";
import { i as resolveChannelEntryMatch, n as buildChannelKeyCandidates } from "../channel-config-D4OY8L18.js";
import "../sessions-BwqyiEaq.js";
import "../paths-rhN9LKM_.js";
import "../session-write-lock-Db3SWNU3.js";
import "../method-scopes-B2vIWWxl.js";
import "../call-BhjE3oIo.js";
import "../prompt-style-HkzccPBT.js";
import "../ports-lsof-Dlzvm55V.js";
import "../restart-stale-pids-OP_Aybe0.js";
import "../ports-B3yeEL-n.js";
import "../logging-D4CVa7JL.js";
import "../commands-2T-1FAo7.js";
import "../issue-format-C8oDJazD.js";
import { t as resolveAckReaction } from "../identity-D07jLUrU.js";
import "../heartbeat-B91tt3Lu.js";
import { B_ as summarizeMapping, F_ as addAllowlistUserEntriesFromConfigEntry, I_ as buildAllowlistResolutionSummary, L_ as canonicalizeAllowlistWithResolvedIds, Lg as formatLocationText, R_ as mergeAllowlist, Rg as toLocationContext, z_ as patchAllowlistUsersInConfigEntries } from "../pi-embedded-CswW9luA.js";
import "../internal-hooks-ClBId-OS.js";
import "../multimodal-rLHaC7mC.js";
import "../memory-search-C_KRQGyn.js";
import "../provider-catalog-BGWDC_kb.js";
import { t as createAccountListHelpers } from "../account-helpers-BiC682rw.js";
import { r as buildSecretInputSchema } from "../secret-input-BOkqlW2S.js";
import "../bindings-DFhSj1AD.js";
import "../resolve-route-jw3TrvbG.js";
import "../routing-GIGwPHsU.js";
import "../identity-file-AtA39mSs.js";
import { r as resolveOutboundSendDep } from "../outbound-runtime-QVIwx-7l.js";
import "../provider-env-vars-BCe4Mf67.js";
import "../provider-auth-input-WYCWu6tD.js";
import "../provider-model-minimax-CYJqVplI.js";
import "../provider-models-ld3k2AX4.js";
import "../anthropic-vertex-provider-7stzcdx2.js";
import "../models-config.providers.discovery-DnQmXa4L.js";
import "../text-runtime-BXjeqyjp.js";
import "../tool-catalog-DP3BJZz3.js";
import "../docker-Cq_Zfc99.js";
import "../sandbox-Dt1ZOIQc.js";
import { c as jsonResult, d as readNumberParam, f as readReactionParams, h as readStringParam, i as createActionGate, p as readStringArrayParam } from "../common-DArxwyX8.js";
import "../image-ops-339pOknK.js";
import "../thinking-rAawjv9l.js";
import "../path-alias-guards-DMcFUz5-.js";
import "../sandbox-paths-D-zPD6r4.js";
import "../channel-actions-DZIaeZzh.js";
import "../mime-C0asv0HQ.js";
import "../ssrf-BifXvxQR.js";
import { n as fetchWithSsrFGuard } from "../fetch-guard-BTY_UfUY.js";
import "../provider-web-search-BHcSeZzc.js";
import "../manager-DmufLZK2.js";
import { t as emptyPluginConfigSchema } from "../config-schema-DOXyQvBE.js";
import { o as moveSingleAccountChannelSectionToDefaultAccount, t as applyAccountNameToChannelSection } from "../setup-helpers-BHFbY0Qo.js";
import { r as buildChannelConfigSchema } from "../config-schema-CKb8N9O1.js";
import { n as deleteAccountFromConfigSection, r as setAccountEnabledInConfigSection } from "../channel-plugin-common-BX9jRob2.js";
import { n as formatPairingApproveHint } from "../helpers-DvzONIyU.js";
import { t as PAIRING_APPROVED_MESSAGE } from "../pairing-message-BICCrA63.js";
import { a as warnMissingProviderGroupPolicyFallbackOnce, n as resolveAllowlistProviderRuntimeGroupPolicy, r as resolveDefaultGroupPolicy, t as GROUP_POLICY_BLOCKED_LABEL } from "../runtime-group-policy-DqDSPQSR.js";
import { i as buildProbeChannelStatusSummary, r as buildComputedAccountStatusSnapshot, s as collectStatusIssuesFromLastError } from "../status-helpers-D9rzemcb.js";
import { K as registerSessionBindingAdapter, W as getSessionBindingService, n as resolveThreadBindingFarewellText, q as unregisterSessionBindingAdapter } from "../conversation-runtime-C56HP9yG.js";
import "../runtime-whatsapp-boundary-BMepWtnk.js";
import "../pairing-store-Bh2YkvRc.js";
import { n as writeJsonFileAtomically, t as readJsonFileWithFallback } from "../json-store-BIglSbRG.js";
import { n as createChannelPairingController } from "../channel-pairing-CA9BsqQV.js";
import "../secret-file-CEppYIjz.js";
import "../channel-config-schema-CrubAncY.js";
import "../setup-binary-BLFhRd_P.js";
import "../archive-Df6iZdzY.js";
import "../fs-safe-B2bSW_OI.js";
import "../signal-cli-install-UDdZA4-C.js";
import { D as mergeAllowFromEntries, L as promptAccountId, U as promptSingleChannelSecretInput, d as buildSingleChannelSecretPromptState, ot as setTopLevelChannelGroupPolicy, u as addWildcardAllowFrom } from "../setup-wizard-proxy--gFyw5Eh.js";
import { t as formatResolvedUnresolvedNote } from "../setup-C0zuBL_6.js";
import { u as resolveControlCommandGate } from "../dm-policy-shared-ChGQzSwJ.js";
import { a as resolveSenderScopedGroupPolicy, t as evaluateGroupRouteAccessForPolicy } from "../group-access-DbAQwRuM.js";
import "../hook-runtime-sGC9reX7.js";
import "../diagnostic-BtBb10Ry.js";
import "../templating-Bxm_rtRG.js";
import { n as createTypingCallbacks, o as createReplyPrefixOptions, t as createChannelReplyPipeline } from "../channel-reply-pipeline-DGoktEKi.js";
import { n as formatZonedTimestamp } from "../format-datetime-BnXmKV6h.js";
import "../reply-history-yYXk1ivL.js";
import "../commands-registry.data-_PzNk48W.js";
import "../commands-registry-ABOHrld2.js";
import { d as logInboundDrop, f as logTypingFailure } from "../channel-feedback-mKZrhk-A.js";
import "../frontmatter-Cd_6iJyw.js";
import "../env-overrides-B5iBORFt.js";
import "../skills-DAP6Yelo.js";
import "../skills-remote-BWQnXFJW.js";
import "../workspace-dirs-DhbPcAGw.js";
import "../pairing-token-CQLjZNmB.js";
import "../skill-commands-I_17sL0j.js";
import "../level-overrides-Ds8ppNbt.js";
import "../config-BnJ6TZJ3.js";
import "../routes-sPyjSxSY.js";
import "../ssh-tunnel-BSFzY8ID.js";
import "../server-middleware-Nz-5LEFG.js";
import "../logging-DFpZCDzJ.js";
import "../config-runtime-gAoemo4L.js";
import "../exec-approvals-CFIa5yg0.js";
import "../webhook-ingress-BBdXea0x.js";
import "../system-events-C2ngfmKm.js";
import "../ssrf-policy-BlgZt63j.js";
import "../provider-auth-ref-Dd8Ctvr9.js";
import "../provider-auth-helpers-CK3uH19P.js";
import "../provider-api-key-auth-CyO5NBeZ.js";
import "../runtime-env-CAlGDNNR.js";
import { s as resolveThreadBindingIdleTimeoutMsForChannel, u as resolveThreadBindingMaxAgeMsForChannel } from "../thread-bindings-policy-DPRpdev-.js";
import "../pairing-labels-BAkUxk5K.js";
import "../directory-runtime-BgPR38m_.js";
import "../read-only-account-inspect-B8C32NVU.js";
import "../src-Bc4Fm2Ge.js";
import { o as getAgentScopedMediaLocalRoots } from "../web-media-Bo_Nxwjv.js";
import "../temp-path-Dgv6Kihk.js";
import "../media-understanding-Ch_w554d.js";
import { n as normalizePollInput } from "../polls-GziVvR8X.js";
import "../web-media-BvqWsq_l.js";
import "../state-paths-C-WJBTWG.js";
import "../llm-task-Djjqflw_.js";
import "../pi-model-discovery-BNyYuKdC.js";
import "../exec-inline-eval-Di27nCtK.js";
import "../target-registry-CofRQwTE.js";
import "../external-content-Bq0zNXEh.js";
import "../brave-Hgb0hhUH.js";
import "../duckduckgo-B0xgAU7B.js";
import "../exa-Ce69v825.js";
import "../security-runtime-CoZ_sfgg.js";
import "../provider-usage-qQL8PJ0p.js";
import "../provider-onboard-CqAihpGi.js";
import "../perplexity-BNVyHCAP.js";
import "../stagger-DZXzssdr.js";
import "../command-secret-targets-wRPHRRib.js";
import "../delivery-queue-1A7pbB9e.js";
import "../channel-summary-DHY50TCU.js";
import "../follow-redirects-0oww1mdf.js";
import "../tool-policy-match-DnEs8-IP.js";
import { n as resolveRuntimeEnv, t as createLoggerBackedRuntime } from "../runtime-CUG7pnyw.js";
import "../channel-status-C_NTcHvs.js";
import "../discord-core-DpMUVJg9.js";
import "../cli-runtime-CYhFvgcO.js";
import "../config-presence-Cqse_N2e.js";
import "../query-expansion-CZbIeYUY.js";
import "../search-manager-DsKEM_vh.js";
import "../acp-runtime-BS3aFfwU.js";
import "../telegram-core-BFUHv9d-.js";
import "../audit--q0z8yGd.js";
import "../gateway-runtime-Czf7LoCY.js";
import "../connection-auth-ZVYO9ARZ.js";
import "../mcp-config-CpCZwCwq.js";
import { t as runPluginCommandWithTimeout } from "../run-command-1zvv2MMi.js";
import "../channel-setup-B0zShv38.js";
import { C as resolveMatrixDefaultOrOnlyAccountId, S as resolveMatrixChannelConfig, T as resolveMatrixAccountStringValues, _ as resolveMatrixCredentialsPath, b as requiresExplicitMatrixDefaultAccount, f as setMatrixThreadBindingMaxAgeBySessionKey, g as resolveMatrixCredentialsDir, h as resolveMatrixAccountStorageRoot, n as matrixSetupWizard, t as matrixSetupAdapter, u as setMatrixThreadBindingIdleTimeoutBySessionKey, v as resolveMatrixLegacyFlatStoragePaths, w as getMatrixScopedEnvVarNames, x as resolveConfiguredMatrixAccountIds, y as findMatrixAccountEntry } from "../matrix-De2qOSaQ.js";
import { t as promptChannelAccessConfig } from "../setup-group-access-DplyRblj.js";
export { DEFAULT_ACCOUNT_ID, GROUP_POLICY_BLOCKED_LABEL, MarkdownConfigSchema, PAIRING_APPROVED_MESSAGE, ToolPolicySchema, addAllowlistUserEntriesFromConfigEntry, addWildcardAllowFrom, applyAccountNameToChannelSection, buildAllowlistResolutionSummary, buildChannelConfigSchema, buildChannelKeyCandidates, buildComputedAccountStatusSnapshot, buildProbeChannelStatusSummary, buildSecretInputSchema, buildSingleChannelSecretPromptState, canonicalizeAllowlistWithResolvedIds, collectStatusIssuesFromLastError, compileAllowlist, createAccountListHelpers, createActionGate, createChannelPairingController, createChannelReplyPipeline, createLoggerBackedRuntime, createReplyPrefixOptions, createTypingCallbacks, deleteAccountFromConfigSection, emptyPluginConfigSchema, evaluateGroupRouteAccessForPolicy, fetchWithSsrFGuard, findMatrixAccountEntry, formatAllowlistMatchMeta, formatDocsLink, formatLocationText, formatPairingApproveHint, formatResolvedUnresolvedNote, formatZonedTimestamp, getAgentScopedMediaLocalRoots, getMatrixScopedEnvVarNames, getSessionBindingService, hasConfiguredSecretInput, isPrivateOrLoopbackHost, jsonResult, logInboundDrop, logTypingFailure, matrixSetupAdapter, matrixSetupWizard, mergeAllowFromEntries, mergeAllowlist, moveSingleAccountChannelSectionToDefaultAccount, normalizeAccountId, normalizeOptionalAccountId, normalizePollInput, normalizeResolvedSecretInputString, normalizeSecretInputString, normalizeStringEntries, patchAllowlistUsersInConfigEntries, promptAccountId, promptChannelAccessConfig, promptSingleChannelSecretInput, readJsonFileWithFallback, readNumberParam, readReactionParams, readStringArrayParam, readStringParam, redactSensitiveText, registerSessionBindingAdapter, requiresExplicitMatrixDefaultAccount, resolveAckReaction, resolveAgentIdFromSessionKey, resolveAllowlistCandidates, resolveAllowlistMatchByCandidates, resolveAllowlistProviderRuntimeGroupPolicy, resolveChannelEntryMatch, resolveCompiledAllowlistMatch, resolveConfiguredMatrixAccountIds, resolveControlCommandGate, resolveDefaultGroupPolicy, resolveMatrixAccountStorageRoot, resolveMatrixAccountStringValues, resolveMatrixChannelConfig, resolveMatrixCredentialsDir, resolveMatrixCredentialsPath, resolveMatrixDefaultOrOnlyAccountId, resolveMatrixLegacyFlatStoragePaths, resolveOutboundSendDep, resolveRuntimeEnv, resolveSenderScopedGroupPolicy, resolveThreadBindingFarewellText, resolveThreadBindingIdleTimeoutMsForChannel, resolveThreadBindingMaxAgeMsForChannel, runPluginCommandWithTimeout, setAccountEnabledInConfigSection, setMatrixThreadBindingIdleTimeoutBySessionKey, setMatrixThreadBindingMaxAgeBySessionKey, setTopLevelChannelGroupPolicy, summarizeMapping, toLocationContext, unregisterSessionBindingAdapter, warnMissingProviderGroupPolicyFallbackOnce, writeJsonFileAtomically };
