// openclaw/dist/plugin-sdk/media-runtime.js
import "../redact-BDinS1q9.js";
import "../errors-BxyFnvP3.js";
import "../unhandled-rejections-CDJ8dOVP.js";
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../warning-filter-CBhOcgHd.js";
import "../utils-CS0Ikux6.js";
import "../links-8xRhWBQL.js";
import "../paths-C4Oy7wjk.js";
import "../auth-profiles-CLtbIyEz.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import { _ as normalizeInboundPathRoots, g as mergeInboundPathRoots, h as isValidInboundPathRootPattern, m as isInboundPathAllowed, p as DEFAULT_IMESSAGE_ATTACHMENT_ROOTS, v as resolveIMessageAttachmentRoots, y as resolveIMessageRemoteAttachmentRoots } from "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import "../file-lock-BJsnXQjL.js";
import "../audit-fs-DTQIjp2r.js";
import "../resolve-B7S4PlnO.js";
import "../profiles-B_nscSgH.js";
import "../tailscale-CxY-91GZ.js";
import "../tailnet-DHiKSG7V.js";
import "../net-JAg9zL0G.js";
import "../auth-Buq0Niri.js";
import "../credentials-D8pfRimu.js";
import "../message-channel-BliByQBl.js";
import "../store-_QDcIMYk.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import "../sessions-BwqyiEaq.js";
import "../paths-rhN9LKM_.js";
import "../session-write-lock-Db3SWNU3.js";
import "../method-scopes-B2vIWWxl.js";
import "../call-BhjE3oIo.js";
import "../prompt-style-HkzccPBT.js";
import "../ports-lsof-Dlzvm55V.js";
import "../restart-stale-pids-OP_Aybe0.js";
import "../ports-B3yeEL-n.js";
import "../logging-D4CVa7JL.js";
import "../commands-2T-1FAo7.js";
import "../issue-format-C8oDJazD.js";
import "../identity-D07jLUrU.js";
import "../heartbeat-B91tt3Lu.js";
import { $m as MEDIA_FFMPEG_MAX_AUDIO_DURATION_SECS, Am as normalizeMediaAttachments, Bm as resolveChannelMediaMaxBytes, Cm as createDirectTextMediaOutbound, Dm as buildProviderRegistry, Em as transcribeFirstAudio, Gm as pngChunk, Hm as crc32, Jm as resolveOutboundMediaLocalRoots, Km as resolveOutboundAttachmentFromUrl, Mm as resolveMediaAttachmentLocalRoots, Nm as runCapability, Om as clearMediaUnderstandingBinaryCacheForTests, Qm as runFfprobe, Tm as resolveScopedChannelMediaMaxBytes, Um as encodePngRgba, Vm as unlinkIfExists, Wm as fillPixel, Xm as parseFfprobeCsvFields, Ym as parseFfprobeCodecAndSampleRate, Zm as runFfmpeg, ah as isTelegramVoiceCompatibleAudio, eh as MEDIA_FFMPEG_MAX_BUFFER_BYTES, ih as TELEGRAM_VOICE_MIME_TYPES, jm as resolveAutoImageModel, km as createMediaAttachmentCache, nh as MEDIA_FFPROBE_TIMEOUT_MS, oh as isVoiceCompatibleAudio, qm as buildOutboundMediaLoadOptions, rh as TELEGRAM_VOICE_AUDIO_EXTENSIONS, th as MEDIA_FFMPEG_TIMEOUT_MS, wm as createScopedChannelMediaMaxBytesResolver, zm as buildAgentMediaPayload } from "../pi-embedded-CswW9luA.js";
import "../internal-hooks-ClBId-OS.js";
import "../multimodal-rLHaC7mC.js";
import "../memory-search-C_KRQGyn.js";
import "../provider-catalog-BGWDC_kb.js";
import "../secret-input-BOkqlW2S.js";
import "../bindings-DFhSj1AD.js";
import "../resolve-route-jw3TrvbG.js";
import "../routing-GIGwPHsU.js";
import "../identity-file-AtA39mSs.js";
import "../outbound-runtime-QVIwx-7l.js";
import "../provider-env-vars-BCe4Mf67.js";
import "../provider-auth-input-WYCWu6tD.js";
import "../provider-model-minimax-CYJqVplI.js";
import "../provider-models-ld3k2AX4.js";
import "../anthropic-vertex-provider-7stzcdx2.js";
import "../models-config.providers.discovery-DnQmXa4L.js";
import "../text-runtime-BXjeqyjp.js";
import "../tool-catalog-DP3BJZz3.js";
import "../docker-Cq_Zfc99.js";
import "../sandbox-Dt1ZOIQc.js";
import "../common-DArxwyX8.js";
import { a as hasAlphaChannel, c as resizeToJpeg, i as getImageMetadata, l as resizeToPng, n as buildImageResizeSideGrid, o as normalizeExifOrientation, r as convertHeicToJpeg, s as optimizeImageToPng, t as IMAGE_REDUCE_QUALITY_STEPS } from "../image-ops-339pOknK.js";
import "../thinking-rAawjv9l.js";
import "../path-alias-guards-DMcFUz5-.js";
import "../sandbox-paths-D-zPD6r4.js";
import "../channel-actions-DZIaeZzh.js";
import { a as isAudioFileName, c as normalizeMimeType, d as MAX_IMAGE_BYTES, f as MAX_VIDEO_BYTES, i as imageMimeFromFormat, l as MAX_AUDIO_BYTES, m as mediaKindFromMime, n as extensionForMime, o as isGifMedia, p as maxBytesForKind, r as getFileExtension, s as kindFromMime, t as detectMime, u as MAX_DOCUMENT_BYTES } from "../mime-C0asv0HQ.js";
import "../ssrf-BifXvxQR.js";
import "../fetch-guard-BTY_UfUY.js";
import "../provider-web-search-BHcSeZzc.js";
import "../manager-DmufLZK2.js";
import "../config-schema-CKb8N9O1.js";
import "../channel-plugin-common-BX9jRob2.js";
import "../status-helpers-D9rzemcb.js";
import "../conversation-runtime-C56HP9yG.js";
import "../runtime-whatsapp-boundary-BMepWtnk.js";
import "../pairing-store-Bh2YkvRc.js";
import "../json-store-BIglSbRG.js";
import "../secret-file-CEppYIjz.js";
import "../channel-config-schema-CrubAncY.js";
import "../setup-binary-BLFhRd_P.js";
import "../archive-Df6iZdzY.js";
import "../fs-safe-B2bSW_OI.js";
import "../signal-cli-install-UDdZA4-C.js";
import "../setup-wizard-proxy--gFyw5Eh.js";
import "../setup-C0zuBL_6.js";
import "../dm-policy-shared-ChGQzSwJ.js";
import "../hook-runtime-sGC9reX7.js";
import "../diagnostic-BtBb10Ry.js";
import { C as DEFAULT_TIMEOUT_SECONDS, S as DEFAULT_PROMPT, T as MIN_AUDIO_FILE_BYTES, _ as DEFAULT_IMAGE_MODELS, b as DEFAULT_MAX_CHARS_BY_CAPABILITY, f as AUTO_AUDIO_KEY_PROVIDERS, g as DEFAULT_AUDIO_MODELS, h as CLI_OUTPUT_MAX_BUFFER, m as AUTO_VIDEO_KEY_PROVIDERS, p as AUTO_IMAGE_KEY_PROVIDERS, v as DEFAULT_MAX_BYTES, w as DEFAULT_VIDEO_MAX_BASE64_BYTES, x as DEFAULT_MEDIA_CONCURRENCY, y as DEFAULT_MAX_CHARS } from "../templating-Bxm_rtRG.js";
import "../channel-reply-pipeline-DGoktEKi.js";
import "../reply-history-yYXk1ivL.js";
import "../commands-registry.data-_PzNk48W.js";
import "../commands-registry-ABOHrld2.js";
import "../frontmatter-Cd_6iJyw.js";
import "../env-overrides-B5iBORFt.js";
import "../skills-DAP6Yelo.js";
import "../skills-remote-BWQnXFJW.js";
import "../workspace-dirs-DhbPcAGw.js";
import "../pairing-token-CQLjZNmB.js";
import "../skill-commands-I_17sL0j.js";
import "../level-overrides-Ds8ppNbt.js";
import "../config-BnJ6TZJ3.js";
import { A as cleanOldMedia, F as saveMediaSource, I as setMediaStoreNetworkDepsForTest, M as extractOriginalFilename, N as getMediaDir, O as MEDIA_MAX_BYTES, P as saveMediaBuffer, j as ensureMediaDir, k as SaveMediaSourceError } from "../routes-sPyjSxSY.js";
import "../ssh-tunnel-BSFzY8ID.js";
import "../server-middleware-Nz-5LEFG.js";
import "../logging-DFpZCDzJ.js";
import "../config-runtime-gAoemo4L.js";
import "../exec-approvals-CFIa5yg0.js";
import "../webhook-ingress-BBdXea0x.js";
import "../system-events-C2ngfmKm.js";
import "../ssrf-policy-BlgZt63j.js";
import "../provider-auth-ref-Dd8Ctvr9.js";
import "../provider-auth-helpers-CK3uH19P.js";
import "../provider-api-key-auth-CyO5NBeZ.js";
import "../runtime-env-CAlGDNNR.js";
import "../pairing-labels-BAkUxk5K.js";
import "../directory-runtime-BgPR38m_.js";
import "../read-only-account-inspect-B8C32NVU.js";
import "../src-Bc4Fm2Ge.js";
import { a as getDefaultLocalRoots, c as MediaFetchError, i as assertLocalMediaAllowed, l as fetchRemoteMedia, o as getAgentScopedMediaLocalRoots, r as LocalMediaAccessError, s as getDefaultMediaLocalRoots } from "../web-media-Bo_Nxwjv.js";
import "../temp-path-Dgv6Kihk.js";
import { a as describeImagesWithModel, i as describeImageWithModel } from "../media-understanding-Ch_w554d.js";
import { n as normalizePollInput, r as resolvePollMaxSelections, t as normalizePollDurationHours } from "../polls-GziVvR8X.js";
import "../web-media-BvqWsq_l.js";
import "../state-paths-C-WJBTWG.js";
import "../llm-task-Djjqflw_.js";
import "../pi-model-discovery-BNyYuKdC.js";
import "../exec-inline-eval-Di27nCtK.js";
import "../target-registry-CofRQwTE.js";
import "../external-content-Bq0zNXEh.js";
import "../brave-Hgb0hhUH.js";
import "../duckduckgo-B0xgAU7B.js";
import "../exa-Ce69v825.js";
import "../security-runtime-CoZ_sfgg.js";
import "../provider-usage-qQL8PJ0p.js";
import "../provider-onboard-CqAihpGi.js";
import "../perplexity-BNVyHCAP.js";
import "../stagger-DZXzssdr.js";
import "../command-secret-targets-wRPHRRib.js";
import "../delivery-queue-1A7pbB9e.js";
import "../channel-summary-DHY50TCU.js";
import "../follow-redirects-0oww1mdf.js";
import "../tool-policy-match-DnEs8-IP.js";
import "../runtime-CUG7pnyw.js";
import "../channel-status-C_NTcHvs.js";
import "../discord-core-DpMUVJg9.js";
import "../cli-runtime-CYhFvgcO.js";
import "../config-presence-Cqse_N2e.js";
import "../query-expansion-CZbIeYUY.js";
import "../search-manager-DsKEM_vh.js";
import "../acp-runtime-BS3aFfwU.js";
import "../telegram-core-BFUHv9d-.js";
import "../audit--q0z8yGd.js";
import "../gateway-runtime-Czf7LoCY.js";
import "../connection-auth-ZVYO9ARZ.js";
import "../mcp-config-CpCZwCwq.js";
export { AUTO_AUDIO_KEY_PROVIDERS, AUTO_IMAGE_KEY_PROVIDERS, AUTO_VIDEO_KEY_PROVIDERS, CLI_OUTPUT_MAX_BUFFER, DEFAULT_AUDIO_MODELS, DEFAULT_IMAGE_MODELS, DEFAULT_IMESSAGE_ATTACHMENT_ROOTS, DEFAULT_MAX_BYTES, DEFAULT_MAX_CHARS, DEFAULT_MAX_CHARS_BY_CAPABILITY, DEFAULT_MEDIA_CONCURRENCY, DEFAULT_PROMPT, DEFAULT_TIMEOUT_SECONDS, DEFAULT_VIDEO_MAX_BASE64_BYTES, IMAGE_REDUCE_QUALITY_STEPS, LocalMediaAccessError, MAX_AUDIO_BYTES, MAX_DOCUMENT_BYTES, MAX_IMAGE_BYTES, MAX_VIDEO_BYTES, MEDIA_FFMPEG_MAX_AUDIO_DURATION_SECS, MEDIA_FFMPEG_MAX_BUFFER_BYTES, MEDIA_FFMPEG_TIMEOUT_MS, MEDIA_FFPROBE_TIMEOUT_MS, MEDIA_MAX_BYTES, MIN_AUDIO_FILE_BYTES, MediaFetchError, SaveMediaSourceError, TELEGRAM_VOICE_AUDIO_EXTENSIONS, TELEGRAM_VOICE_MIME_TYPES, assertLocalMediaAllowed, buildAgentMediaPayload, buildImageResizeSideGrid, buildOutboundMediaLoadOptions, buildProviderRegistry, cleanOldMedia, clearMediaUnderstandingBinaryCacheForTests, convertHeicToJpeg, crc32, createDirectTextMediaOutbound, createMediaAttachmentCache, createScopedChannelMediaMaxBytesResolver, describeImageWithModel, describeImagesWithModel, detectMime, encodePngRgba, ensureMediaDir, extensionForMime, extractOriginalFilename, fetchRemoteMedia, fillPixel, getAgentScopedMediaLocalRoots, getDefaultLocalRoots, getDefaultMediaLocalRoots, getFileExtension, getImageMetadata, getMediaDir, hasAlphaChannel, imageMimeFromFormat, isAudioFileName, isGifMedia, isInboundPathAllowed, isTelegramVoiceCompatibleAudio, isValidInboundPathRootPattern, isVoiceCompatibleAudio, kindFromMime, maxBytesForKind, mediaKindFromMime, mergeInboundPathRoots, normalizeExifOrientation, normalizeInboundPathRoots, normalizeMediaAttachments, normalizeMimeType, normalizePollDurationHours, normalizePollInput, optimizeImageToPng, parseFfprobeCodecAndSampleRate, parseFfprobeCsvFields, pngChunk, resizeToJpeg, resizeToPng, resolveAutoImageModel, resolveChannelMediaMaxBytes, resolveIMessageAttachmentRoots, resolveIMessageRemoteAttachmentRoots, resolveMediaAttachmentLocalRoots, resolveOutboundAttachmentFromUrl, resolveOutboundMediaLocalRoots, resolvePollMaxSelections, resolveScopedChannelMediaMaxBytes, runCapability, runFfmpeg, runFfprobe, saveMediaBuffer, saveMediaSource, setMediaStoreNetworkDepsForTest, transcribeFirstAudio, unlinkIfExists };
