// openclaw/dist/plugin-sdk/media-understanding.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../logger-BmpSCz93.js";
import "../ip-Ce8EDTBZ.js";
import "../ssrf-BifXvxQR.js";
import "../fetch-guard-BTY_UfUY.js";
import { a as describeImagesWithModel, c as postJsonRequest, i as describeImageWithModel, l as postTranscriptionRequest, n as deepgramMediaUnderstandingProvider, o as assertOkOrThrowHttpError, r as transcribeOpenAiCompatibleAudio, s as normalizeBaseUrl, t as groqMediaUnderstandingProvider, u as requireTranscriptionText } from "../media-understanding-Ch_w554d.js";
export { assertOkOrThrowHttpError, deepgramMediaUnderstandingProvider, describeImageWithModel, describeImagesWithModel, groqMediaUnderstandingProvider, normalizeBaseUrl, postJsonRequest, postTranscriptionRequest, requireTranscriptionText, transcribeOpenAiCompatibleAudio };
