// openclaw/dist/plugin-sdk/provider-auth-api-key.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../boolean-C3GkJetE.js";
import "../env-Dnra1IpT.js";
import "../utils-CS0Ikux6.js";
import "../paths-C4Oy7wjk.js";
import "../auth-profiles-CLtbIyEz.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import "../io-CZZeeo8R.js";
import "../shell-env-BOu7XeT_.js";
import "../safe-text-yavot2qw.js";
import "../version-CD3oP1-d.js";
import "../env-substitution-O6uabUmO.js";
import "../includes-Bj3eLUWH.js";
import "../zod-schema.providers-core-COy5nBQ0.js";
import "../legacy-web-search-BB-ZHEhz.js";
import "../registry-C5UkPpaO.js";
import "../config-state-Br0ucqMb.js";
import "../min-host-version-RMBWtIAR.js";
import "../manifest-registry-B5JNQdOM.js";
import "../runtime-guard-PhQ6PwQa.js";
import "../avatar-policy-B5nOfso_.js";
import "../ip-Ce8EDTBZ.js";
import "../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../zod-schema.core-BuVz8Rk7.js";
import "../config-Bqe8DjNc.js";
import "../file-lock-BJsnXQjL.js";
import "../audit-fs-DTQIjp2r.js";
import "../resolve-B7S4PlnO.js";
import { n as normalizeSecretInput, t as normalizeOptionalSecretInput } from "../normalize-secret-input-BsX1agd_.js";
import { a as upsertAuthProfile } from "../profiles-B_nscSgH.js";
import "../provider-env-vars-BCe4Mf67.js";
import { a as normalizeSecretInputModeInput, i as normalizeApiKeyInput, n as ensureApiKeyFromOptionEnvOrPrompt, r as formatApiKeyPreview, s as validateApiKeyInput } from "../provider-auth-input-WYCWu6tD.js";
import "../anthropic-vertex-provider-7stzcdx2.js";
import { t as resolveSecretInputModeForEnvSelection } from "../provider-auth-mode-8HZLxa_B.js";
import { n as promptSecretRefForSetup } from "../provider-auth-ref-Dd8Ctvr9.js";
import { n as buildApiKeyCredential, t as applyAuthProfileConfig } from "../provider-auth-helpers-CK3uH19P.js";
import { t as createProviderApiKeyAuthMethod } from "../provider-api-key-auth-CyO5NBeZ.js";
import "../provider-auth-api-key-D-GwUZDd.js";
export { applyAuthProfileConfig, buildApiKeyCredential, createProviderApiKeyAuthMethod, ensureApiKeyFromOptionEnvOrPrompt, formatApiKeyPreview, normalizeApiKeyInput, normalizeOptionalSecretInput, normalizeSecretInput, normalizeSecretInputModeInput, promptSecretRefForSetup, resolveSecretInputModeForEnvSelection, upsertAuthProfile, validateApiKeyInput };
