// openclaw/dist/plugin-sdk/reply-history.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../utils-CS0Ikux6.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../registry-C5UkPpaO.js";
import "../runtime-Iz8uZ7EU.js";
import "../plugins-B09-vgme.js";
import { a as buildPendingHistoryContextFromMap, c as evictOldHistoryKeys, i as buildHistoryContextFromMap, l as recordPendingHistoryEntry, n as buildHistoryContext, o as clearHistoryEntries, r as buildHistoryContextFromEntries, s as clearHistoryEntriesIfEnabled, t as DEFAULT_GROUP_HISTORY_LIMIT, u as recordPendingHistoryEntryIfEnabled } from "../reply-history-yYXk1ivL.js";
export { DEFAULT_GROUP_HISTORY_LIMIT, buildHistoryContext, buildHistoryContextFromEntries, buildHistoryContextFromMap, buildPendingHistoryContextFromMap, clearHistoryEntries, clearHistoryEntriesIfEnabled, evictOldHistoryKeys, recordPendingHistoryEntry, recordPendingHistoryEntryIfEnabled };
