// openclaw/dist/plugin-sdk/self-hosted-provider-setup.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../utils-CS0Ikux6.js";
import "../paths-C4Oy7wjk.js";
import "../agent-scope-bjWqU22i.js";
import "../boundary-path-Dm0QJ7-y.js";
import "../boundary-file-read-DcZxlWD8.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import "../workspace-D4K6QX9X.js";
import "../model-selection-BnFtDmP7.js";
import "../file-lock-BJsnXQjL.js";
import "../profiles-B_nscSgH.js";
import "../provider-catalog-BGWDC_kb.js";
import "../provider-env-vars-BCe4Mf67.js";
import "../provider-model-minimax-CYJqVplI.js";
import "../provider-models-ld3k2AX4.js";
import "../anthropic-vertex-provider-7stzcdx2.js";
import { a as SELF_HOSTED_DEFAULT_COST, i as SELF_HOSTED_DEFAULT_CONTEXT_WINDOW, n as buildSglangProvider, o as SELF_HOSTED_DEFAULT_MAX_TOKENS, r as buildVllmProvider } from "../models-config.providers.discovery-DnQmXa4L.js";
import "../provider-auth-helpers-CK3uH19P.js";
import "../upsert-with-lock-BVZ8xSaq.js";
import { a as promptAndConfigureOpenAICompatibleSelfHostedProviderAuth, i as promptAndConfigureOpenAICompatibleSelfHostedProvider, n as configureOpenAICompatibleSelfHostedProviderNonInteractive, r as discoverOpenAICompatibleSelfHostedProvider, t as applyProviderDefaultModel } from "../provider-self-hosted-setup-W3RY4ltD.js";
export { SELF_HOSTED_DEFAULT_CONTEXT_WINDOW, SELF_HOSTED_DEFAULT_COST, SELF_HOSTED_DEFAULT_MAX_TOKENS, applyProviderDefaultModel, buildSglangProvider, buildVllmProvider, configureOpenAICompatibleSelfHostedProviderNonInteractive, discoverOpenAICompatibleSelfHostedProvider, promptAndConfigureOpenAICompatibleSelfHostedProvider, promptAndConfigureOpenAICompatibleSelfHostedProviderAuth };
