// openclaw/dist/plugin-sdk/setup-runtime.js
import "../logger-kwZIqwuw.js";
import "../paths-ViKUYWUK.js";
import "../tmp-openclaw-dir-idKIOMmb.js";
import "../theme-CdOoMzRk.js";
import "../globals-DBUMOBZ8.js";
import "../subsystem-DISldKSB.js";
import "../ansi-BEJF8NKS.js";
import "../utils-CS0Ikux6.js";
import { g as DEFAULT_ACCOUNT_ID } from "../session-key-DAhnzjyr.js";
import "../logger-BmpSCz93.js";
import "../exec-B5_AYfQG.js";
import { r as createEnvPatchedAccountSetupAdapter } from "../setup-helpers-BHFbY0Qo.js";
import "../setup-binary-BLFhRd_P.js";
import { G as resolveEntriesWithOptionalToken, P as patchChannelConfigForAccount, f as createAccountScopedAllowFromSection, h as createLegacyCompatChannelDmPolicy, j as parseMentionOrPrefixedId, p as createAccountScopedGroupAccessSection, rt as setSetupChannelEnabled, t as createAllowlistSetupWizardProxy, x as createStandardChannelSetupStatus, z as promptLegacyChannelAllowFromForAccount } from "../setup-wizard-proxy--gFyw5Eh.js";
import "../setup-runtime-CdMZ9yID.js";
export { DEFAULT_ACCOUNT_ID, createAccountScopedAllowFromSection, createAccountScopedGroupAccessSection, createAllowlistSetupWizardProxy, createEnvPatchedAccountSetupAdapter, createLegacyCompatChannelDmPolicy, createStandardChannelSetupStatus, parseMentionOrPrefixedId, patchChannelConfigForAccount, promptLegacyChannelAllowFromForAccount, resolveEntriesWithOptionalToken, setSetupChannelEnabled };
