// openclaw/dist/extensions/cloudflare-ai-gateway/index.js
import "../../redact-BDinS1q9.js";
import "../../errors-BxyFnvP3.js";
import "../../unhandled-rejections-CDJ8dOVP.js";
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../boolean-C3GkJetE.js";
import "../../env-Dnra1IpT.js";
import "../../warning-filter-CBhOcgHd.js";
import "../../utils-CS0Ikux6.js";
import "../../links-8xRhWBQL.js";
import "../../paths-C4Oy7wjk.js";
import "../../auth-profiles-CLtbIyEz.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../io-CZZeeo8R.js";
import "../../shell-env-BOu7XeT_.js";
import "../../safe-text-yavot2qw.js";
import "../../version-CD3oP1-d.js";
import { i as coerceSecretRef } from "../../types.secrets-DgX397vI.js";
import "../../env-substitution-O6uabUmO.js";
import "../../includes-Bj3eLUWH.js";
import "../../zod-schema.providers-core-COy5nBQ0.js";
import "../../legacy-web-search-BB-ZHEhz.js";
import "../../registry-C5UkPpaO.js";
import "../../config-state-Br0ucqMb.js";
import "../../min-host-version-RMBWtIAR.js";
import "../../manifest-registry-B5JNQdOM.js";
import "../../runtime-guard-PhQ6PwQa.js";
import "../../avatar-policy-B5nOfso_.js";
import "../../ip-Ce8EDTBZ.js";
import "../../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../../zod-schema.core-BuVz8Rk7.js";
import "../../config-Bqe8DjNc.js";
import "../../file-lock-BJsnXQjL.js";
import "../../audit-fs-DTQIjp2r.js";
import "../../resolve-B7S4PlnO.js";
import { t as normalizeOptionalSecretInput } from "../../normalize-secret-input-BsX1agd_.js";
import { a as upsertAuthProfile, c as ensureAuthProfileStore, n as listProfilesForProvider } from "../../profiles-B_nscSgH.js";
import "../../tailscale-CxY-91GZ.js";
import "../../tailnet-DHiKSG7V.js";
import "../../net-JAg9zL0G.js";
import "../../auth-Buq0Niri.js";
import "../../credentials-D8pfRimu.js";
import "../../message-channel-BliByQBl.js";
import "../../store-_QDcIMYk.js";
import "../../runtime-Iz8uZ7EU.js";
import "../../plugins-B09-vgme.js";
import "../../sessions-BwqyiEaq.js";
import "../../paths-rhN9LKM_.js";
import "../../session-write-lock-Db3SWNU3.js";
import "../../method-scopes-B2vIWWxl.js";
import "../../call-BhjE3oIo.js";
import "../../prompt-style-HkzccPBT.js";
import "../../ports-lsof-Dlzvm55V.js";
import "../../restart-stale-pids-OP_Aybe0.js";
import "../../ports-B3yeEL-n.js";
import "../../logging-D4CVa7JL.js";
import "../../commands-2T-1FAo7.js";
import "../../issue-format-C8oDJazD.js";
import "../../identity-D07jLUrU.js";
import "../../heartbeat-B91tt3Lu.js";
import "../../pi-embedded-CswW9luA.js";
import "../../internal-hooks-ClBId-OS.js";
import "../../multimodal-rLHaC7mC.js";
import "../../memory-search-C_KRQGyn.js";
import "../../provider-catalog-BGWDC_kb.js";
import "../../secret-input-BOkqlW2S.js";
import "../../bindings-DFhSj1AD.js";
import "../../resolve-route-jw3TrvbG.js";
import "../../routing-GIGwPHsU.js";
import "../../identity-file-AtA39mSs.js";
import "../../outbound-runtime-QVIwx-7l.js";
import "../../provider-env-vars-BCe4Mf67.js";
import { C as resolveNonEnvSecretRefApiKeyMarker, i as normalizeApiKeyInput, n as ensureApiKeyFromOptionEnvOrPrompt, s as validateApiKeyInput } from "../../provider-auth-input-WYCWu6tD.js";
import "../../provider-model-minimax-CYJqVplI.js";
import { at as buildCloudflareAiGatewayModelDefinition, it as CLOUDFLARE_AI_GATEWAY_DEFAULT_MODEL_REF, ot as resolveCloudflareAiGatewayBaseUrl } from "../../provider-models-ld3k2AX4.js";
import "../../anthropic-vertex-provider-7stzcdx2.js";
import "../../models-config.providers.discovery-DnQmXa4L.js";
import "../../text-runtime-BXjeqyjp.js";
import "../../tool-catalog-DP3BJZz3.js";
import "../../docker-Cq_Zfc99.js";
import "../../sandbox-Dt1ZOIQc.js";
import "../../common-DArxwyX8.js";
import "../../image-ops-339pOknK.js";
import "../../thinking-rAawjv9l.js";
import "../../path-alias-guards-DMcFUz5-.js";
import "../../sandbox-paths-D-zPD6r4.js";
import "../../channel-actions-DZIaeZzh.js";
import "../../mime-C0asv0HQ.js";
import "../../ssrf-BifXvxQR.js";
import "../../fetch-guard-BTY_UfUY.js";
import "../../provider-web-search-BHcSeZzc.js";
import "../../manager-DmufLZK2.js";
import "../../config-schema-CKb8N9O1.js";
import "../../channel-plugin-common-BX9jRob2.js";
import "../../status-helpers-D9rzemcb.js";
import "../../conversation-runtime-C56HP9yG.js";
import "../../runtime-whatsapp-boundary-BMepWtnk.js";
import "../../pairing-store-Bh2YkvRc.js";
import "../../json-store-BIglSbRG.js";
import { t as definePluginEntry } from "../../plugin-entry-BRMgG77c.js";
import "../../secret-file-CEppYIjz.js";
import "../../channel-config-schema-CrubAncY.js";
import "../../setup-binary-BLFhRd_P.js";
import "../../archive-Df6iZdzY.js";
import "../../fs-safe-B2bSW_OI.js";
import "../../signal-cli-install-UDdZA4-C.js";
import "../../setup-wizard-proxy--gFyw5Eh.js";
import "../../setup-C0zuBL_6.js";
import "../../dm-policy-shared-ChGQzSwJ.js";
import "../../hook-runtime-sGC9reX7.js";
import "../../diagnostic-BtBb10Ry.js";
import "../../templating-Bxm_rtRG.js";
import "../../channel-reply-pipeline-DGoktEKi.js";
import "../../reply-history-yYXk1ivL.js";
import "../../commands-registry.data-_PzNk48W.js";
import "../../commands-registry-ABOHrld2.js";
import "../../frontmatter-Cd_6iJyw.js";
import "../../env-overrides-B5iBORFt.js";
import "../../skills-DAP6Yelo.js";
import "../../skills-remote-BWQnXFJW.js";
import "../../workspace-dirs-DhbPcAGw.js";
import "../../pairing-token-CQLjZNmB.js";
import "../../skill-commands-I_17sL0j.js";
import "../../level-overrides-Ds8ppNbt.js";
import "../../config-BnJ6TZJ3.js";
import "../../routes-sPyjSxSY.js";
import "../../ssh-tunnel-BSFzY8ID.js";
import "../../server-middleware-Nz-5LEFG.js";
import "../../logging-DFpZCDzJ.js";
import "../../config-runtime-gAoemo4L.js";
import "../../exec-approvals-CFIa5yg0.js";
import "../../webhook-ingress-BBdXea0x.js";
import "../../system-events-C2ngfmKm.js";
import "../../ssrf-policy-BlgZt63j.js";
import "../../provider-auth-ref-Dd8Ctvr9.js";
import { n as buildApiKeyCredential, t as applyAuthProfileConfig } from "../../provider-auth-helpers-CK3uH19P.js";
import "../../provider-api-key-auth-CyO5NBeZ.js";
import "../../runtime-env-CAlGDNNR.js";
import "../../pairing-labels-BAkUxk5K.js";
import "../../directory-runtime-BgPR38m_.js";
import "../../read-only-account-inspect-B8C32NVU.js";
import "../../src-Bc4Fm2Ge.js";
import "../../web-media-Bo_Nxwjv.js";
import "../../temp-path-Dgv6Kihk.js";
import "../../media-understanding-Ch_w554d.js";
import "../../web-media-BvqWsq_l.js";
import "../../state-paths-C-WJBTWG.js";
import "../../llm-task-Djjqflw_.js";
import "../../pi-model-discovery-BNyYuKdC.js";
import "../../exec-inline-eval-Di27nCtK.js";
import "../../target-registry-CofRQwTE.js";
import "../../external-content-Bq0zNXEh.js";
import "../../brave-Hgb0hhUH.js";
import "../../duckduckgo-B0xgAU7B.js";
import "../../exa-Ce69v825.js";
import "../../security-runtime-CoZ_sfgg.js";
import "../../provider-usage-qQL8PJ0p.js";
import { i as applyCloudflareAiGatewayConfig, o as buildCloudflareAiGatewayConfigPatch } from "../../provider-onboard-CqAihpGi.js";
import "../../perplexity-BNVyHCAP.js";
import "../../stagger-DZXzssdr.js";
import "../../command-secret-targets-wRPHRRib.js";
import "../../delivery-queue-1A7pbB9e.js";
import "../../channel-summary-DHY50TCU.js";
import "../../follow-redirects-0oww1mdf.js";
import "../../tool-policy-match-DnEs8-IP.js";
import "../../runtime-CUG7pnyw.js";
import "../../channel-status-C_NTcHvs.js";
import "../../discord-core-DpMUVJg9.js";
import "../../cli-runtime-CYhFvgcO.js";
import "../../config-presence-Cqse_N2e.js";
import "../../query-expansion-CZbIeYUY.js";
import "../../search-manager-DsKEM_vh.js";
import "../../acp-runtime-BS3aFfwU.js";
import "../../telegram-core-BFUHv9d-.js";
import "../../audit--q0z8yGd.js";
import "../../gateway-runtime-Czf7LoCY.js";
import "../../connection-auth-ZVYO9ARZ.js";
import "../../mcp-config-CpCZwCwq.js";
//#region extensions/cloudflare-ai-gateway/index.ts
const PROVIDER_ID = "cloudflare-ai-gateway";
const PROVIDER_ENV_VAR = "CLOUDFLARE_AI_GATEWAY_API_KEY";
const PROFILE_ID = "cloudflare-ai-gateway:default";
function resolveApiKeyFromCredential(cred) {
	if (!cred || cred.type !== "api_key") return;
	const keyRef = coerceSecretRef(cred.keyRef);
	if (keyRef && keyRef.id.trim()) return keyRef.source === "env" ? keyRef.id.trim() : resolveNonEnvSecretRefApiKeyMarker(keyRef.source);
	return cred.key?.trim() || void 0;
}
function resolveMetadataFromCredential(cred) {
	if (!cred || cred.type !== "api_key") return {};
	return {
		accountId: cred?.metadata?.accountId?.trim() || void 0,
		gatewayId: cred?.metadata?.gatewayId?.trim() || void 0
	};
}
async function resolveCloudflareGatewayMetadataInteractive(ctx) {
	let accountId = ctx.accountId?.trim() ?? "";
	let gatewayId = ctx.gatewayId?.trim() ?? "";
	if (!accountId) {
		const value = await ctx.prompter.text({
			message: "Enter Cloudflare Account ID",
			validate: (val) => String(val ?? "").trim() ? void 0 : "Account ID is required"
		});
		accountId = String(value ?? "").trim();
	}
	if (!gatewayId) {
		const value = await ctx.prompter.text({
			message: "Enter Cloudflare AI Gateway ID",
			validate: (val) => String(val ?? "").trim() ? void 0 : "Gateway ID is required"
		});
		gatewayId = String(value ?? "").trim();
	}
	return {
		accountId,
		gatewayId
	};
}
var cloudflare_ai_gateway_default = definePluginEntry({
	id: PROVIDER_ID,
	name: "Cloudflare AI Gateway Provider",
	description: "Bundled Cloudflare AI Gateway provider plugin",
	register(api) {
		api.registerProvider({
			id: PROVIDER_ID,
			label: "Cloudflare AI Gateway",
			docsPath: "/providers/cloudflare-ai-gateway",
			envVars: ["CLOUDFLARE_AI_GATEWAY_API_KEY"],
			auth: [{
				id: "api-key",
				label: "Cloudflare AI Gateway",
				hint: "Account ID + Gateway ID + API key",
				kind: "api_key",
				wizard: {
					choiceId: "cloudflare-ai-gateway-api-key",
					choiceLabel: "Cloudflare AI Gateway",
					choiceHint: "Account ID + Gateway ID + API key",
					groupId: "cloudflare-ai-gateway",
					groupLabel: "Cloudflare AI Gateway",
					groupHint: "Account ID + Gateway ID + API key"
				},
				run: async (ctx) => {
					const metadata = await resolveCloudflareGatewayMetadataInteractive({
						accountId: normalizeOptionalSecretInput(ctx.opts?.cloudflareAiGatewayAccountId),
						gatewayId: normalizeOptionalSecretInput(ctx.opts?.cloudflareAiGatewayGatewayId),
						prompter: ctx.prompter
					});
					let capturedSecretInput;
					let capturedCredential = false;
					let capturedMode;
					await ensureApiKeyFromOptionEnvOrPrompt({
						token: normalizeOptionalSecretInput(ctx.opts?.cloudflareAiGatewayApiKey),
						tokenProvider: "cloudflare-ai-gateway",
						secretInputMode: ctx.allowSecretRefPrompt === false ? ctx.secretInputMode ?? "plaintext" : ctx.secretInputMode,
						config: ctx.config,
						expectedProviders: [PROVIDER_ID],
						provider: PROVIDER_ID,
						envLabel: PROVIDER_ENV_VAR,
						promptMessage: "Enter Cloudflare AI Gateway API key",
						normalize: normalizeApiKeyInput,
						validate: validateApiKeyInput,
						prompter: ctx.prompter,
						setCredential: async (apiKey, mode) => {
							capturedSecretInput = apiKey;
							capturedCredential = true;
							capturedMode = mode;
						}
					});
					if (!capturedCredential) throw new Error("Missing Cloudflare AI Gateway API key.");
					return {
						profiles: [{
							profileId: PROFILE_ID,
							credential: buildApiKeyCredential(PROVIDER_ID, capturedSecretInput ?? "", {
								accountId: metadata.accountId,
								gatewayId: metadata.gatewayId
							}, capturedMode ? { secretInputMode: capturedMode } : void 0)
						}],
						configPatch: buildCloudflareAiGatewayConfigPatch(metadata),
						defaultModel: CLOUDFLARE_AI_GATEWAY_DEFAULT_MODEL_REF
					};
				},
				runNonInteractive: async (ctx) => {
					const storedMetadata = resolveMetadataFromCredential(ensureAuthProfileStore(ctx.agentDir, { allowKeychainPrompt: false }).profiles[PROFILE_ID]);
					const accountId = normalizeOptionalSecretInput(ctx.opts.cloudflareAiGatewayAccountId) ?? storedMetadata.accountId;
					const gatewayId = normalizeOptionalSecretInput(ctx.opts.cloudflareAiGatewayGatewayId) ?? storedMetadata.gatewayId;
					if (!accountId || !gatewayId) {
						ctx.runtime.error("Cloudflare AI Gateway setup requires --cloudflare-ai-gateway-account-id and --cloudflare-ai-gateway-gateway-id.");
						ctx.runtime.exit(1);
						return null;
					}
					const resolved = await ctx.resolveApiKey({
						provider: PROVIDER_ID,
						flagValue: normalizeOptionalSecretInput(ctx.opts.cloudflareAiGatewayApiKey),
						flagName: "--cloudflare-ai-gateway-api-key",
						envVar: PROVIDER_ENV_VAR
					});
					if (!resolved) return null;
					if (resolved.source !== "profile") {
						const credential = ctx.toApiKeyCredential({
							provider: PROVIDER_ID,
							resolved,
							metadata: {
								accountId,
								gatewayId
							}
						});
						if (!credential) return null;
						upsertAuthProfile({
							profileId: PROFILE_ID,
							credential,
							agentDir: ctx.agentDir
						});
					}
					return applyCloudflareAiGatewayConfig(applyAuthProfileConfig(ctx.config, {
						profileId: PROFILE_ID,
						provider: PROVIDER_ID,
						mode: "api_key"
					}), {
						accountId,
						gatewayId
					});
				}
			}],
			catalog: {
				order: "late",
				run: async (ctx) => {
					const authStore = ensureAuthProfileStore(ctx.agentDir, { allowKeychainPrompt: false });
					const envManagedApiKey = ctx.env[PROVIDER_ENV_VAR]?.trim() ? PROVIDER_ENV_VAR : void 0;
					for (const profileId of listProfilesForProvider(authStore, PROVIDER_ID)) {
						const cred = authStore.profiles[profileId];
						if (!cred || cred.type !== "api_key") continue;
						const apiKey = envManagedApiKey ?? resolveApiKeyFromCredential(cred);
						if (!apiKey) continue;
						const accountId = cred.metadata?.accountId?.trim();
						const gatewayId = cred.metadata?.gatewayId?.trim();
						if (!accountId || !gatewayId) continue;
						const baseUrl = resolveCloudflareAiGatewayBaseUrl({
							accountId,
							gatewayId
						});
						if (!baseUrl) continue;
						return { provider: {
							baseUrl,
							api: "anthropic-messages",
							apiKey,
							models: [buildCloudflareAiGatewayModelDefinition()]
						} };
					}
					return null;
				}
			}
		});
	}
});
//#endregion
export { cloudflare_ai_gateway_default as default };
