// openclaw/dist/extensions/lobster/index.js
import "../../redact-BDinS1q9.js";
import "../../errors-BxyFnvP3.js";
import "../../unhandled-rejections-CDJ8dOVP.js";
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../boolean-C3GkJetE.js";
import "../../env-Dnra1IpT.js";
import "../../warning-filter-CBhOcgHd.js";
import "../../utils-CS0Ikux6.js";
import "../../links-8xRhWBQL.js";
import "../../paths-C4Oy7wjk.js";
import "../../auth-profiles-CLtbIyEz.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../io-CZZeeo8R.js";
import "../../shell-env-BOu7XeT_.js";
import "../../safe-text-yavot2qw.js";
import "../../version-CD3oP1-d.js";
import "../../env-substitution-O6uabUmO.js";
import "../../includes-Bj3eLUWH.js";
import "../../zod-schema.providers-core-COy5nBQ0.js";
import "../../legacy-web-search-BB-ZHEhz.js";
import "../../registry-C5UkPpaO.js";
import "../../config-state-Br0ucqMb.js";
import "../../min-host-version-RMBWtIAR.js";
import "../../manifest-registry-B5JNQdOM.js";
import "../../runtime-guard-PhQ6PwQa.js";
import "../../avatar-policy-B5nOfso_.js";
import "../../ip-Ce8EDTBZ.js";
import "../../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../../zod-schema.core-BuVz8Rk7.js";
import "../../config-Bqe8DjNc.js";
import "../../file-lock-BJsnXQjL.js";
import "../../audit-fs-DTQIjp2r.js";
import "../../resolve-B7S4PlnO.js";
import "../../profiles-B_nscSgH.js";
import "../../tailscale-CxY-91GZ.js";
import "../../tailnet-DHiKSG7V.js";
import "../../net-JAg9zL0G.js";
import "../../auth-Buq0Niri.js";
import "../../credentials-D8pfRimu.js";
import "../../message-channel-BliByQBl.js";
import "../../store-_QDcIMYk.js";
import "../../runtime-Iz8uZ7EU.js";
import "../../plugins-B09-vgme.js";
import "../../sessions-BwqyiEaq.js";
import "../../paths-rhN9LKM_.js";
import "../../session-write-lock-Db3SWNU3.js";
import "../../method-scopes-B2vIWWxl.js";
import "../../call-BhjE3oIo.js";
import "../../prompt-style-HkzccPBT.js";
import "../../ports-lsof-Dlzvm55V.js";
import "../../restart-stale-pids-OP_Aybe0.js";
import "../../ports-B3yeEL-n.js";
import "../../logging-D4CVa7JL.js";
import "../../commands-2T-1FAo7.js";
import "../../issue-format-C8oDJazD.js";
import "../../identity-D07jLUrU.js";
import "../../heartbeat-B91tt3Lu.js";
import "../../pi-embedded-CswW9luA.js";
import "../../internal-hooks-ClBId-OS.js";
import "../../multimodal-rLHaC7mC.js";
import "../../memory-search-C_KRQGyn.js";
import "../../provider-catalog-BGWDC_kb.js";
import "../../secret-input-BOkqlW2S.js";
import "../../bindings-DFhSj1AD.js";
import "../../resolve-route-jw3TrvbG.js";
import "../../routing-GIGwPHsU.js";
import "../../identity-file-AtA39mSs.js";
import "../../outbound-runtime-QVIwx-7l.js";
import "../../provider-env-vars-BCe4Mf67.js";
import "../../provider-auth-input-WYCWu6tD.js";
import "../../provider-model-minimax-CYJqVplI.js";
import "../../provider-models-ld3k2AX4.js";
import "../../anthropic-vertex-provider-7stzcdx2.js";
import "../../models-config.providers.discovery-DnQmXa4L.js";
import "../../text-runtime-BXjeqyjp.js";
import "../../tool-catalog-DP3BJZz3.js";
import "../../docker-Cq_Zfc99.js";
import "../../sandbox-Dt1ZOIQc.js";
import "../../common-DArxwyX8.js";
import "../../image-ops-339pOknK.js";
import "../../thinking-rAawjv9l.js";
import "../../path-alias-guards-DMcFUz5-.js";
import "../../sandbox-paths-D-zPD6r4.js";
import "../../channel-actions-DZIaeZzh.js";
import "../../mime-C0asv0HQ.js";
import "../../ssrf-BifXvxQR.js";
import "../../fetch-guard-BTY_UfUY.js";
import "../../provider-web-search-BHcSeZzc.js";
import "../../manager-DmufLZK2.js";
import "../../config-schema-CKb8N9O1.js";
import "../../channel-plugin-common-BX9jRob2.js";
import "../../status-helpers-D9rzemcb.js";
import "../../conversation-runtime-C56HP9yG.js";
import "../../runtime-whatsapp-boundary-BMepWtnk.js";
import "../../pairing-store-Bh2YkvRc.js";
import "../../json-store-BIglSbRG.js";
import { t as definePluginEntry } from "../../plugin-entry-BRMgG77c.js";
import "../../secret-file-CEppYIjz.js";
import "../../channel-config-schema-CrubAncY.js";
import "../../setup-binary-BLFhRd_P.js";
import "../../archive-Df6iZdzY.js";
import "../../fs-safe-B2bSW_OI.js";
import "../../signal-cli-install-UDdZA4-C.js";
import "../../setup-wizard-proxy--gFyw5Eh.js";
import "../../setup-C0zuBL_6.js";
import "../../dm-policy-shared-ChGQzSwJ.js";
import "../../hook-runtime-sGC9reX7.js";
import "../../diagnostic-BtBb10Ry.js";
import "../../templating-Bxm_rtRG.js";
import "../../channel-reply-pipeline-DGoktEKi.js";
import "../../reply-history-yYXk1ivL.js";
import "../../commands-registry.data-_PzNk48W.js";
import "../../commands-registry-ABOHrld2.js";
import "../../frontmatter-Cd_6iJyw.js";
import "../../env-overrides-B5iBORFt.js";
import "../../skills-DAP6Yelo.js";
import "../../skills-remote-BWQnXFJW.js";
import "../../workspace-dirs-DhbPcAGw.js";
import "../../pairing-token-CQLjZNmB.js";
import "../../skill-commands-I_17sL0j.js";
import "../../level-overrides-Ds8ppNbt.js";
import "../../config-BnJ6TZJ3.js";
import "../../routes-sPyjSxSY.js";
import { a as resolveWindowsSpawnProgramCandidate, n as materializeWindowsSpawnProgram, t as applyWindowsSpawnProgramPolicy } from "../../windows-spawn-D5U7V5RW.js";
import "../../ssh-tunnel-BSFzY8ID.js";
import "../../server-middleware-Nz-5LEFG.js";
import "../../logging-DFpZCDzJ.js";
import "../../config-runtime-gAoemo4L.js";
import "../../exec-approvals-CFIa5yg0.js";
import "../../webhook-ingress-BBdXea0x.js";
import "../../system-events-C2ngfmKm.js";
import "../../ssrf-policy-BlgZt63j.js";
import "../../provider-auth-ref-Dd8Ctvr9.js";
import "../../provider-auth-helpers-CK3uH19P.js";
import "../../provider-api-key-auth-CyO5NBeZ.js";
import "../../runtime-env-CAlGDNNR.js";
import "../../pairing-labels-BAkUxk5K.js";
import "../../directory-runtime-BgPR38m_.js";
import "../../read-only-account-inspect-B8C32NVU.js";
import "../../src-Bc4Fm2Ge.js";
import "../../web-media-Bo_Nxwjv.js";
import "../../temp-path-Dgv6Kihk.js";
import "../../media-understanding-Ch_w554d.js";
import "../../web-media-BvqWsq_l.js";
import "../../state-paths-C-WJBTWG.js";
import "../../llm-task-Djjqflw_.js";
import "../../pi-model-discovery-BNyYuKdC.js";
import "../../exec-inline-eval-Di27nCtK.js";
import "../../target-registry-CofRQwTE.js";
import "../../external-content-Bq0zNXEh.js";
import "../../brave-Hgb0hhUH.js";
import "../../duckduckgo-B0xgAU7B.js";
import "../../exa-Ce69v825.js";
import "../../security-runtime-CoZ_sfgg.js";
import "../../provider-usage-qQL8PJ0p.js";
import "../../provider-onboard-CqAihpGi.js";
import "../../perplexity-BNVyHCAP.js";
import "../../stagger-DZXzssdr.js";
import "../../command-secret-targets-wRPHRRib.js";
import "../../delivery-queue-1A7pbB9e.js";
import "../../channel-summary-DHY50TCU.js";
import "../../follow-redirects-0oww1mdf.js";
import "../../tool-policy-match-DnEs8-IP.js";
import "../../runtime-CUG7pnyw.js";
import "../../channel-status-C_NTcHvs.js";
import "../../discord-core-DpMUVJg9.js";
import "../../cli-runtime-CYhFvgcO.js";
import "../../config-presence-Cqse_N2e.js";
import "../../query-expansion-CZbIeYUY.js";
import "../../search-manager-DsKEM_vh.js";
import "../../acp-runtime-BS3aFfwU.js";
import "../../telegram-core-BFUHv9d-.js";
import "../../audit--q0z8yGd.js";
import "../../gateway-runtime-Czf7LoCY.js";
import "../../connection-auth-ZVYO9ARZ.js";
import "../../mcp-config-CpCZwCwq.js";
import path from "node:path";
import { spawn } from "node:child_process";
import { Type } from "@sinclair/typebox";
//#region extensions/lobster/src/windows-spawn.ts
function resolveWindowsLobsterSpawn(execPath, argv, env) {
	const resolved = materializeWindowsSpawnProgram(applyWindowsSpawnProgramPolicy({
		candidate: resolveWindowsSpawnProgramCandidate({
			command: execPath,
			env,
			packageName: "lobster"
		}),
		allowShellFallback: false
	}), argv);
	if (resolved.shell) throw new Error("lobster wrapper resolved to shell fallback unexpectedly");
	return {
		command: resolved.command,
		argv: resolved.argv,
		windowsHide: resolved.windowsHide
	};
}
//#endregion
//#region extensions/lobster/src/lobster-tool.ts
function normalizeForCwdSandbox(p) {
	const normalized = path.normalize(p);
	return process.platform === "win32" ? normalized.toLowerCase() : normalized;
}
function resolveCwd(cwdRaw) {
	if (typeof cwdRaw !== "string" || !cwdRaw.trim()) return process.cwd();
	const cwd = cwdRaw.trim();
	if (path.isAbsolute(cwd)) throw new Error("cwd must be a relative path");
	const base = process.cwd();
	const resolved = path.resolve(base, cwd);
	const rel = path.relative(normalizeForCwdSandbox(base), normalizeForCwdSandbox(resolved));
	if (rel === "" || rel === ".") return resolved;
	if (rel.startsWith("..") || path.isAbsolute(rel)) throw new Error("cwd must stay within the gateway working directory");
	return resolved;
}
async function runLobsterSubprocessOnce(params) {
	const { execPath, argv, cwd } = params;
	const timeoutMs = Math.max(200, params.timeoutMs);
	const maxStdoutBytes = Math.max(1024, params.maxStdoutBytes);
	const env = {
		...process.env,
		LOBSTER_MODE: "tool"
	};
	if ((env.NODE_OPTIONS ?? "").includes("--inspect")) delete env.NODE_OPTIONS;
	const spawnTarget = process.platform === "win32" ? resolveWindowsLobsterSpawn(execPath, argv, env) : {
		command: execPath,
		argv
	};
	return await new Promise((resolve, reject) => {
		const child = spawn(spawnTarget.command, spawnTarget.argv, {
			cwd,
			stdio: [
				"ignore",
				"pipe",
				"pipe"
			],
			env,
			windowsHide: spawnTarget.windowsHide
		});
		let stdout = "";
		let stdoutBytes = 0;
		let stderr = "";
		let settled = false;
		const settle = (result) => {
			if (settled) return;
			settled = true;
			clearTimeout(timer);
			if (result.ok) resolve(result.value);
			else reject(result.error);
		};
		const failAndTerminate = (message) => {
			try {
				child.kill("SIGKILL");
			} finally {
				settle({
					ok: false,
					error: new Error(message)
				});
			}
		};
		child.stdout?.setEncoding("utf8");
		child.stderr?.setEncoding("utf8");
		child.stdout?.on("data", (chunk) => {
			const str = String(chunk);
			stdoutBytes += Buffer.byteLength(str, "utf8");
			if (stdoutBytes > maxStdoutBytes) {
				failAndTerminate("lobster output exceeded maxStdoutBytes");
				return;
			}
			stdout += str;
		});
		child.stderr?.on("data", (chunk) => {
			stderr += String(chunk);
		});
		const timer = setTimeout(() => {
			failAndTerminate("lobster subprocess timed out");
		}, timeoutMs);
		child.once("error", (err) => {
			settle({
				ok: false,
				error: err
			});
		});
		child.once("exit", (code) => {
			if (code !== 0) {
				settle({
					ok: false,
					error: /* @__PURE__ */ new Error(`lobster failed (${code ?? "?"}): ${stderr.trim() || stdout.trim()}`)
				});
				return;
			}
			settle({
				ok: true,
				value: { stdout }
			});
		});
	});
}
function parseEnvelope(stdout) {
	const trimmed = stdout.trim();
	const tryParse = (input) => {
		try {
			return JSON.parse(input);
		} catch {
			return;
		}
	};
	let parsed = tryParse(trimmed);
	if (parsed === void 0) {
		const suffixMatch = trimmed.match(/({[\s\S]*}|\[[\s\S]*])\s*$/);
		if (suffixMatch?.[1]) parsed = tryParse(suffixMatch[1]);
	}
	if (parsed === void 0) throw new Error("lobster returned invalid JSON");
	if (!parsed || typeof parsed !== "object") throw new Error("lobster returned invalid JSON envelope");
	const ok = parsed.ok;
	if (ok === true || ok === false) return parsed;
	throw new Error("lobster returned invalid JSON envelope");
}
function buildLobsterArgv(action, params) {
	if (action === "run") {
		const pipeline = typeof params.pipeline === "string" ? params.pipeline : "";
		if (!pipeline.trim()) throw new Error("pipeline required");
		const argv = [
			"run",
			"--mode",
			"tool",
			pipeline
		];
		const argsJson = typeof params.argsJson === "string" ? params.argsJson : "";
		if (argsJson.trim()) argv.push("--args-json", argsJson);
		return argv;
	}
	if (action === "resume") {
		const token = typeof params.token === "string" ? params.token : "";
		if (!token.trim()) throw new Error("token required");
		const approve = params.approve;
		if (typeof approve !== "boolean") throw new Error("approve required");
		return [
			"resume",
			"--token",
			token,
			"--approve",
			approve ? "yes" : "no"
		];
	}
	throw new Error(`Unknown action: ${action}`);
}
function createLobsterTool(api) {
	return {
		name: "lobster",
		label: "Lobster Workflow",
		description: "Run Lobster pipelines as a local-first workflow runtime (typed JSON envelope + resumable approvals).",
		parameters: Type.Object({
			action: Type.Unsafe({
				type: "string",
				enum: ["run", "resume"]
			}),
			pipeline: Type.Optional(Type.String()),
			argsJson: Type.Optional(Type.String()),
			token: Type.Optional(Type.String()),
			approve: Type.Optional(Type.Boolean()),
			cwd: Type.Optional(Type.String({ description: "Relative working directory (optional). Must stay within the gateway working directory." })),
			timeoutMs: Type.Optional(Type.Number()),
			maxStdoutBytes: Type.Optional(Type.Number())
		}),
		async execute(_id, params) {
			const action = typeof params.action === "string" ? params.action.trim() : "";
			if (!action) throw new Error("action required");
			const execPath = "lobster";
			const cwd = resolveCwd(params.cwd);
			const timeoutMs = typeof params.timeoutMs === "number" ? params.timeoutMs : 2e4;
			const maxStdoutBytes = typeof params.maxStdoutBytes === "number" ? params.maxStdoutBytes : 512e3;
			const argv = buildLobsterArgv(action, params);
			if (api.runtime?.version && api.logger?.debug) api.logger.debug(`lobster plugin runtime=${api.runtime.version}`);
			const { stdout } = await runLobsterSubprocessOnce({
				execPath,
				argv,
				cwd,
				timeoutMs,
				maxStdoutBytes
			});
			const envelope = parseEnvelope(stdout);
			return {
				content: [{
					type: "text",
					text: JSON.stringify(envelope, null, 2)
				}],
				details: envelope
			};
		}
	};
}
//#endregion
//#region extensions/lobster/index.ts
var lobster_default = definePluginEntry({
	id: "lobster",
	name: "Lobster",
	description: "Optional local shell helper tools",
	register(api) {
		api.registerTool(((ctx) => {
			if (ctx.sandboxed) return null;
			return createLobsterTool(api);
		}), { optional: true });
	}
});
//#endregion
export { lobster_default as default };
