// openclaw/dist/extensions/minimax/index.js
import "../../redact-BDinS1q9.js";
import "../../errors-BxyFnvP3.js";
import "../../unhandled-rejections-CDJ8dOVP.js";
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../boolean-C3GkJetE.js";
import "../../env-Dnra1IpT.js";
import "../../warning-filter-CBhOcgHd.js";
import "../../utils-CS0Ikux6.js";
import "../../links-8xRhWBQL.js";
import "../../paths-C4Oy7wjk.js";
import "../../auth-profiles-CLtbIyEz.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../io-CZZeeo8R.js";
import "../../shell-env-BOu7XeT_.js";
import "../../safe-text-yavot2qw.js";
import "../../version-CD3oP1-d.js";
import "../../env-substitution-O6uabUmO.js";
import "../../includes-Bj3eLUWH.js";
import "../../zod-schema.providers-core-COy5nBQ0.js";
import "../../legacy-web-search-BB-ZHEhz.js";
import "../../registry-C5UkPpaO.js";
import "../../config-state-Br0ucqMb.js";
import "../../min-host-version-RMBWtIAR.js";
import "../../manifest-registry-B5JNQdOM.js";
import "../../runtime-guard-PhQ6PwQa.js";
import "../../avatar-policy-B5nOfso_.js";
import "../../ip-Ce8EDTBZ.js";
import "../../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../../zod-schema.core-BuVz8Rk7.js";
import "../../config-Bqe8DjNc.js";
import "../../file-lock-BJsnXQjL.js";
import "../../audit-fs-DTQIjp2r.js";
import "../../resolve-B7S4PlnO.js";
import { c as ensureAuthProfileStore, n as listProfilesForProvider } from "../../profiles-B_nscSgH.js";
import "../../tailscale-CxY-91GZ.js";
import "../../tailnet-DHiKSG7V.js";
import "../../net-JAg9zL0G.js";
import "../../auth-Buq0Niri.js";
import "../../credentials-D8pfRimu.js";
import "../../message-channel-BliByQBl.js";
import "../../store-_QDcIMYk.js";
import "../../runtime-Iz8uZ7EU.js";
import "../../plugins-B09-vgme.js";
import "../../sessions-BwqyiEaq.js";
import "../../paths-rhN9LKM_.js";
import "../../session-write-lock-Db3SWNU3.js";
import "../../method-scopes-B2vIWWxl.js";
import "../../call-BhjE3oIo.js";
import "../../prompt-style-HkzccPBT.js";
import "../../ports-lsof-Dlzvm55V.js";
import "../../restart-stale-pids-OP_Aybe0.js";
import "../../ports-B3yeEL-n.js";
import "../../logging-D4CVa7JL.js";
import "../../commands-2T-1FAo7.js";
import "../../issue-format-C8oDJazD.js";
import "../../identity-D07jLUrU.js";
import "../../heartbeat-B91tt3Lu.js";
import { Hh as buildOauthProviderAuthResult } from "../../pi-embedded-CswW9luA.js";
import "../../internal-hooks-ClBId-OS.js";
import "../../multimodal-rLHaC7mC.js";
import "../../memory-search-C_KRQGyn.js";
import { C as buildMinimaxPortalProvider, w as buildMinimaxProvider } from "../../provider-catalog-BGWDC_kb.js";
import "../../secret-input-BOkqlW2S.js";
import "../../bindings-DFhSj1AD.js";
import "../../resolve-route-jw3TrvbG.js";
import "../../routing-GIGwPHsU.js";
import "../../identity-file-AtA39mSs.js";
import "../../outbound-runtime-QVIwx-7l.js";
import "../../provider-env-vars-BCe4Mf67.js";
import "../../provider-auth-input-WYCWu6tD.js";
import { o as isMiniMaxModernModelId, r as MINIMAX_TEXT_MODEL_CATALOG, t as MINIMAX_DEFAULT_MODEL_ID } from "../../provider-model-minimax-CYJqVplI.js";
import "../../provider-models-ld3k2AX4.js";
import "../../anthropic-vertex-provider-7stzcdx2.js";
import "../../models-config.providers.discovery-DnQmXa4L.js";
import "../../text-runtime-BXjeqyjp.js";
import "../../tool-catalog-DP3BJZz3.js";
import "../../docker-Cq_Zfc99.js";
import "../../sandbox-Dt1ZOIQc.js";
import "../../common-DArxwyX8.js";
import "../../image-ops-339pOknK.js";
import "../../thinking-rAawjv9l.js";
import "../../path-alias-guards-DMcFUz5-.js";
import "../../sandbox-paths-D-zPD6r4.js";
import "../../channel-actions-DZIaeZzh.js";
import "../../mime-C0asv0HQ.js";
import "../../ssrf-BifXvxQR.js";
import "../../fetch-guard-BTY_UfUY.js";
import "../../provider-web-search-BHcSeZzc.js";
import "../../manager-DmufLZK2.js";
import "../../config-schema-CKb8N9O1.js";
import "../../channel-plugin-common-BX9jRob2.js";
import "../../status-helpers-D9rzemcb.js";
import "../../conversation-runtime-C56HP9yG.js";
import "../../runtime-whatsapp-boundary-BMepWtnk.js";
import "../../pairing-store-Bh2YkvRc.js";
import "../../json-store-BIglSbRG.js";
import { t as definePluginEntry } from "../../plugin-entry-BRMgG77c.js";
import "../../secret-file-CEppYIjz.js";
import "../../channel-config-schema-CrubAncY.js";
import "../../setup-binary-BLFhRd_P.js";
import "../../archive-Df6iZdzY.js";
import "../../fs-safe-B2bSW_OI.js";
import "../../signal-cli-install-UDdZA4-C.js";
import "../../setup-wizard-proxy--gFyw5Eh.js";
import "../../setup-C0zuBL_6.js";
import "../../dm-policy-shared-ChGQzSwJ.js";
import "../../hook-runtime-sGC9reX7.js";
import "../../diagnostic-BtBb10Ry.js";
import "../../templating-Bxm_rtRG.js";
import "../../channel-reply-pipeline-DGoktEKi.js";
import "../../reply-history-yYXk1ivL.js";
import "../../commands-registry.data-_PzNk48W.js";
import "../../commands-registry-ABOHrld2.js";
import "../../frontmatter-Cd_6iJyw.js";
import "../../env-overrides-B5iBORFt.js";
import "../../skills-DAP6Yelo.js";
import "../../skills-remote-BWQnXFJW.js";
import "../../workspace-dirs-DhbPcAGw.js";
import "../../pairing-token-CQLjZNmB.js";
import "../../skill-commands-I_17sL0j.js";
import "../../level-overrides-Ds8ppNbt.js";
import "../../config-BnJ6TZJ3.js";
import "../../routes-sPyjSxSY.js";
import "../../ssh-tunnel-BSFzY8ID.js";
import "../../server-middleware-Nz-5LEFG.js";
import "../../logging-DFpZCDzJ.js";
import "../../config-runtime-gAoemo4L.js";
import "../../exec-approvals-CFIa5yg0.js";
import "../../webhook-ingress-BBdXea0x.js";
import "../../system-events-C2ngfmKm.js";
import "../../ssrf-policy-BlgZt63j.js";
import "../../provider-auth-ref-Dd8Ctvr9.js";
import "../../provider-auth-helpers-CK3uH19P.js";
import { t as createProviderApiKeyAuthMethod } from "../../provider-api-key-auth-CyO5NBeZ.js";
import "../../runtime-env-CAlGDNNR.js";
import "../../pairing-labels-BAkUxk5K.js";
import "../../directory-runtime-BgPR38m_.js";
import "../../read-only-account-inspect-B8C32NVU.js";
import "../../src-Bc4Fm2Ge.js";
import "../../web-media-Bo_Nxwjv.js";
import "../../temp-path-Dgv6Kihk.js";
import { a as describeImagesWithModel, i as describeImageWithModel } from "../../media-understanding-Ch_w554d.js";
import "../../web-media-BvqWsq_l.js";
import "../../state-paths-C-WJBTWG.js";
import "../../llm-task-Djjqflw_.js";
import "../../pi-model-discovery-BNyYuKdC.js";
import "../../exec-inline-eval-Di27nCtK.js";
import "../../target-registry-CofRQwTE.js";
import "../../external-content-Bq0zNXEh.js";
import "../../brave-Hgb0hhUH.js";
import "../../duckduckgo-B0xgAU7B.js";
import "../../exa-Ce69v825.js";
import "../../security-runtime-CoZ_sfgg.js";
import { n as fetchMinimaxUsage } from "../../provider-usage-qQL8PJ0p.js";
import { n as applyOnboardAuthAgentModelsAndProviders, t as applyAgentDefaultModelPrimary } from "../../provider-onboarding-config-ICH5iU24.js";
import "../../provider-onboard-CqAihpGi.js";
import "../../perplexity-BNVyHCAP.js";
import "../../stagger-DZXzssdr.js";
import "../../command-secret-targets-wRPHRRib.js";
import "../../delivery-queue-1A7pbB9e.js";
import "../../channel-summary-DHY50TCU.js";
import "../../follow-redirects-0oww1mdf.js";
import "../../tool-policy-match-DnEs8-IP.js";
import "../../runtime-CUG7pnyw.js";
import "../../channel-status-C_NTcHvs.js";
import "../../discord-core-DpMUVJg9.js";
import "../../cli-runtime-CYhFvgcO.js";
import "../../config-presence-Cqse_N2e.js";
import "../../query-expansion-CZbIeYUY.js";
import "../../search-manager-DsKEM_vh.js";
import "../../acp-runtime-BS3aFfwU.js";
import "../../telegram-core-BFUHv9d-.js";
import "../../audit--q0z8yGd.js";
import "../../gateway-runtime-Czf7LoCY.js";
import "../../connection-auth-ZVYO9ARZ.js";
import "../../mcp-config-CpCZwCwq.js";
//#region extensions/minimax/media-understanding-provider.ts
const minimaxMediaUnderstandingProvider = {
	id: "minimax",
	capabilities: ["image"],
	describeImage: describeImageWithModel,
	describeImages: describeImagesWithModel
};
const minimaxPortalMediaUnderstandingProvider = {
	id: "minimax-portal",
	capabilities: ["image"],
	describeImage: describeImageWithModel,
	describeImages: describeImagesWithModel
};
//#endregion
//#region extensions/minimax/model-definitions.ts
const MINIMAX_API_BASE_URL = "https://api.minimax.io/anthropic";
const MINIMAX_CN_API_BASE_URL = "https://api.minimaxi.com/anthropic";
`${MINIMAX_DEFAULT_MODEL_ID}`;
const DEFAULT_MINIMAX_CONTEXT_WINDOW = 204800;
const DEFAULT_MINIMAX_MAX_TOKENS = 131072;
const MINIMAX_API_COST = {
	input: .3,
	output: 1.2,
	cacheRead: .06,
	cacheWrite: .375
};
function buildMinimaxModelDefinition(params) {
	const catalog = MINIMAX_TEXT_MODEL_CATALOG[params.id];
	return {
		id: params.id,
		name: params.name ?? catalog?.name ?? `MiniMax ${params.id}`,
		reasoning: params.reasoning ?? catalog?.reasoning ?? false,
		input: ["text"],
		cost: params.cost,
		contextWindow: params.contextWindow,
		maxTokens: params.maxTokens
	};
}
function buildMinimaxApiModelDefinition(modelId) {
	return buildMinimaxModelDefinition({
		id: modelId,
		cost: MINIMAX_API_COST,
		contextWindow: DEFAULT_MINIMAX_CONTEXT_WINDOW,
		maxTokens: DEFAULT_MINIMAX_MAX_TOKENS
	});
}
//#endregion
//#region extensions/minimax/onboard.ts
function applyMinimaxApiProviderConfigWithBaseUrl(cfg, params) {
	const providers = { ...cfg.models?.providers };
	const existingProvider = providers[params.providerId];
	const existingModels = existingProvider?.models ?? [];
	const apiModel = buildMinimaxApiModelDefinition(params.modelId);
	const mergedModels = existingModels.some((model) => model.id === params.modelId) ? existingModels : [...existingModels, apiModel];
	const { apiKey: existingApiKey, ...existingProviderRest } = existingProvider ?? {
		baseUrl: params.baseUrl,
		models: []
	};
	const resolvedApiKey = typeof existingApiKey === "string" ? existingApiKey : void 0;
	const normalizedApiKey = resolvedApiKey?.trim() === "minimax" ? "" : resolvedApiKey;
	providers[params.providerId] = {
		...existingProviderRest,
		baseUrl: params.baseUrl,
		api: "anthropic-messages",
		authHeader: true,
		...normalizedApiKey?.trim() ? { apiKey: normalizedApiKey } : {},
		models: mergedModels.length > 0 ? mergedModels : [apiModel]
	};
	const models = { ...cfg.agents?.defaults?.models };
	const modelRef = `${params.providerId}/${params.modelId}`;
	models[modelRef] = {
		...models[modelRef],
		alias: "Minimax"
	};
	return applyOnboardAuthAgentModelsAndProviders(cfg, {
		agentModels: models,
		providers
	});
}
function applyMinimaxApiConfigWithBaseUrl(cfg, params) {
	return applyAgentDefaultModelPrimary(applyMinimaxApiProviderConfigWithBaseUrl(cfg, params), `${params.providerId}/${params.modelId}`);
}
function applyMinimaxApiConfig(cfg, modelId = "MiniMax-M2.7") {
	return applyMinimaxApiConfigWithBaseUrl(cfg, {
		providerId: "minimax",
		modelId,
		baseUrl: MINIMAX_API_BASE_URL
	});
}
function applyMinimaxApiConfigCn(cfg, modelId = "MiniMax-M2.7") {
	return applyMinimaxApiConfigWithBaseUrl(cfg, {
		providerId: "minimax",
		modelId,
		baseUrl: MINIMAX_CN_API_BASE_URL
	});
}
//#endregion
//#region extensions/minimax/index.ts
const API_PROVIDER_ID = "minimax";
const PORTAL_PROVIDER_ID = "minimax-portal";
const PROVIDER_LABEL = "MiniMax";
const DEFAULT_MODEL = MINIMAX_DEFAULT_MODEL_ID;
const DEFAULT_BASE_URL_CN = "https://api.minimaxi.com/anthropic";
const DEFAULT_BASE_URL_GLOBAL = "https://api.minimax.io/anthropic";
function getDefaultBaseUrl(region) {
	return region === "cn" ? DEFAULT_BASE_URL_CN : DEFAULT_BASE_URL_GLOBAL;
}
function apiModelRef(modelId) {
	return `${API_PROVIDER_ID}/${modelId}`;
}
function portalModelRef(modelId) {
	return `${PORTAL_PROVIDER_ID}/${modelId}`;
}
function buildPortalProviderCatalog(params) {
	return {
		...buildMinimaxPortalProvider(),
		baseUrl: params.baseUrl,
		apiKey: params.apiKey
	};
}
function resolveApiCatalog(ctx) {
	const apiKey = ctx.resolveProviderApiKey(API_PROVIDER_ID).apiKey;
	if (!apiKey) return null;
	return { provider: {
		...buildMinimaxProvider(),
		apiKey
	} };
}
function resolvePortalCatalog(ctx) {
	const explicitProvider = ctx.config.models?.providers?.[PORTAL_PROVIDER_ID];
	const envApiKey = ctx.resolveProviderApiKey(PORTAL_PROVIDER_ID).apiKey;
	const hasProfiles = listProfilesForProvider(ensureAuthProfileStore(ctx.agentDir, { allowKeychainPrompt: false }), PORTAL_PROVIDER_ID).length > 0;
	const explicitApiKey = typeof explicitProvider?.apiKey === "string" ? explicitProvider.apiKey.trim() : void 0;
	const apiKey = envApiKey ?? explicitApiKey ?? (hasProfiles ? "minimax-oauth" : void 0);
	if (!apiKey) return null;
	return { provider: buildPortalProviderCatalog({
		baseUrl: (typeof explicitProvider?.baseUrl === "string" ? explicitProvider.baseUrl.trim() : void 0) || DEFAULT_BASE_URL_GLOBAL,
		apiKey
	}) };
}
function createOAuthHandler(region) {
	const defaultBaseUrl = getDefaultBaseUrl(region);
	const regionLabel = region === "cn" ? "CN" : "Global";
	return async (ctx) => {
		const progress = ctx.prompter.progress(`Starting MiniMax OAuth (${regionLabel})…`);
		try {
			const { loginMiniMaxPortalOAuth } = await import("../../oauth.runtime-CGpHc4-s.js");
			const result = await loginMiniMaxPortalOAuth({
				openUrl: ctx.openUrl,
				note: ctx.prompter.note,
				progress,
				region
			});
			progress.stop("MiniMax OAuth complete");
			if (result.notification_message) await ctx.prompter.note(result.notification_message, "MiniMax OAuth");
			const baseUrl = result.resourceUrl || defaultBaseUrl;
			return buildOauthProviderAuthResult({
				providerId: PORTAL_PROVIDER_ID,
				defaultModel: portalModelRef(DEFAULT_MODEL),
				access: result.access,
				refresh: result.refresh,
				expires: result.expires,
				configPatch: {
					models: { providers: { [PORTAL_PROVIDER_ID]: {
						baseUrl,
						models: []
					} } },
					agents: { defaults: { models: {
						[portalModelRef("MiniMax-M2")]: { alias: "minimax-m2" },
						[portalModelRef("MiniMax-M2.1")]: { alias: "minimax-m2.1" },
						[portalModelRef("MiniMax-M2.1-highspeed")]: { alias: "minimax-m2.1-highspeed" },
						[portalModelRef("MiniMax-M2.7")]: { alias: "minimax-m2.7" },
						[portalModelRef("MiniMax-M2.7-highspeed")]: { alias: "minimax-m2.7-highspeed" },
						[portalModelRef("MiniMax-M2.5")]: { alias: "minimax-m2.5" },
						[portalModelRef("MiniMax-M2.5-highspeed")]: { alias: "minimax-m2.5-highspeed" },
						[portalModelRef("MiniMax-M2.5-Lightning")]: { alias: "minimax-m2.5-lightning" }
					} } }
				},
				notes: [
					"MiniMax OAuth tokens auto-refresh. Re-run login if refresh fails or access is revoked.",
					`Base URL defaults to ${defaultBaseUrl}. Override models.providers.${PORTAL_PROVIDER_ID}.baseUrl if needed.`,
					...result.notification_message ? [result.notification_message] : []
				]
			});
		} catch (err) {
			const errorMsg = err instanceof Error ? err.message : String(err);
			progress.stop(`MiniMax OAuth failed: ${errorMsg}`);
			await ctx.prompter.note("If OAuth fails, verify your MiniMax account has portal access and try again.", "MiniMax OAuth");
			throw err;
		}
	};
}
var minimax_default = definePluginEntry({
	id: API_PROVIDER_ID,
	name: "MiniMax",
	description: "Bundled MiniMax API-key and OAuth provider plugin",
	register(api) {
		api.registerProvider({
			id: API_PROVIDER_ID,
			label: PROVIDER_LABEL,
			docsPath: "/providers/minimax",
			envVars: ["MINIMAX_API_KEY"],
			auth: [createProviderApiKeyAuthMethod({
				providerId: API_PROVIDER_ID,
				methodId: "api-global",
				label: "MiniMax API key (Global)",
				hint: "Global endpoint - api.minimax.io",
				optionKey: "minimaxApiKey",
				flagName: "--minimax-api-key",
				envVar: "MINIMAX_API_KEY",
				promptMessage: "Enter MiniMax API key (sk-api- or sk-cp-)\nhttps://platform.minimax.io/user-center/basic-information/interface-key",
				profileId: "minimax:global",
				allowProfile: false,
				defaultModel: apiModelRef(DEFAULT_MODEL),
				expectedProviders: ["minimax"],
				applyConfig: (cfg) => applyMinimaxApiConfig(cfg),
				wizard: {
					choiceId: "minimax-global-api",
					choiceLabel: "MiniMax API key (Global)",
					choiceHint: "Global endpoint - api.minimax.io",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				}
			}), createProviderApiKeyAuthMethod({
				providerId: API_PROVIDER_ID,
				methodId: "api-cn",
				label: "MiniMax API key (CN)",
				hint: "CN endpoint - api.minimaxi.com",
				optionKey: "minimaxApiKey",
				flagName: "--minimax-api-key",
				envVar: "MINIMAX_API_KEY",
				promptMessage: "Enter MiniMax CN API key (sk-api- or sk-cp-)\nhttps://platform.minimaxi.com/user-center/basic-information/interface-key",
				profileId: "minimax:cn",
				allowProfile: false,
				defaultModel: apiModelRef(DEFAULT_MODEL),
				expectedProviders: ["minimax", "minimax-cn"],
				applyConfig: (cfg) => applyMinimaxApiConfigCn(cfg),
				wizard: {
					choiceId: "minimax-cn-api",
					choiceLabel: "MiniMax API key (CN)",
					choiceHint: "CN endpoint - api.minimaxi.com",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				}
			})],
			catalog: {
				order: "simple",
				run: async (ctx) => resolveApiCatalog(ctx)
			},
			resolveUsageAuth: async (ctx) => {
				const apiKey = ctx.resolveApiKeyFromConfigAndStore({ envDirect: [ctx.env.MINIMAX_CODE_PLAN_KEY, ctx.env.MINIMAX_API_KEY] });
				return apiKey ? { token: apiKey } : null;
			},
			isModernModelRef: ({ modelId }) => isMiniMaxModernModelId(modelId),
			fetchUsageSnapshot: async (ctx) => await fetchMinimaxUsage(ctx.token, ctx.timeoutMs, ctx.fetchFn)
		});
		api.registerProvider({
			id: PORTAL_PROVIDER_ID,
			label: PROVIDER_LABEL,
			docsPath: "/providers/minimax",
			envVars: ["MINIMAX_OAUTH_TOKEN", "MINIMAX_API_KEY"],
			catalog: { run: async (ctx) => resolvePortalCatalog(ctx) },
			auth: [{
				id: "oauth",
				label: "MiniMax OAuth (Global)",
				hint: "Global endpoint - api.minimax.io",
				kind: "device_code",
				wizard: {
					choiceId: "minimax-global-oauth",
					choiceLabel: "MiniMax OAuth (Global)",
					choiceHint: "Global endpoint - api.minimax.io",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				},
				run: createOAuthHandler("global")
			}, {
				id: "oauth-cn",
				label: "MiniMax OAuth (CN)",
				hint: "CN endpoint - api.minimaxi.com",
				kind: "device_code",
				wizard: {
					choiceId: "minimax-cn-oauth",
					choiceLabel: "MiniMax OAuth (CN)",
					choiceHint: "CN endpoint - api.minimaxi.com",
					groupId: "minimax",
					groupLabel: "MiniMax",
					groupHint: "M2.7 (recommended)"
				},
				run: createOAuthHandler("cn")
			}],
			isModernModelRef: ({ modelId }) => isMiniMaxModernModelId(modelId)
		});
		api.registerMediaUnderstandingProvider(minimaxMediaUnderstandingProvider);
		api.registerMediaUnderstandingProvider(minimaxPortalMediaUnderstandingProvider);
	}
});
//#endregion
export { minimax_default as default };
