// openclaw/dist/channel.runtime-EWHvJFoY.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import "./links-8xRhWBQL.js";
import "./paths-C4Oy7wjk.js";
import "./auth-profiles-CLtbIyEz.js";
import "./agent-scope-bjWqU22i.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import "./model-selection-BnFtDmP7.js";
import "./io-CZZeeo8R.js";
import "./shell-env-BOu7XeT_.js";
import "./safe-text-yavot2qw.js";
import "./version-CD3oP1-d.js";
import "./env-substitution-O6uabUmO.js";
import "./includes-Bj3eLUWH.js";
import "./zod-schema.providers-core-COy5nBQ0.js";
import "./legacy-web-search-BB-ZHEhz.js";
import "./registry-C5UkPpaO.js";
import "./config-state-Br0ucqMb.js";
import "./min-host-version-RMBWtIAR.js";
import "./manifest-registry-B5JNQdOM.js";
import "./runtime-guard-PhQ6PwQa.js";
import "./avatar-policy-B5nOfso_.js";
import "./ip-Ce8EDTBZ.js";
import "./zod-schema.agent-runtime-Dtg4Jy6G.js";
import "./zod-schema.core-BuVz8Rk7.js";
import "./config-Bqe8DjNc.js";
import "./file-lock-BJsnXQjL.js";
import "./audit-fs-DTQIjp2r.js";
import "./resolve-B7S4PlnO.js";
import "./profiles-B_nscSgH.js";
import "./tailscale-CxY-91GZ.js";
import "./tailnet-DHiKSG7V.js";
import "./net-JAg9zL0G.js";
import "./auth-Buq0Niri.js";
import "./credentials-D8pfRimu.js";
import "./message-channel-BliByQBl.js";
import "./store-_QDcIMYk.js";
import "./runtime-Iz8uZ7EU.js";
import "./plugins-B09-vgme.js";
import "./sessions-BwqyiEaq.js";
import "./paths-rhN9LKM_.js";
import "./session-write-lock-Db3SWNU3.js";
import "./method-scopes-B2vIWWxl.js";
import "./call-BhjE3oIo.js";
import "./prompt-style-HkzccPBT.js";
import "./ports-lsof-Dlzvm55V.js";
import "./restart-stale-pids-OP_Aybe0.js";
import "./ports-B3yeEL-n.js";
import "./logging-D4CVa7JL.js";
import "./commands-2T-1FAo7.js";
import "./issue-format-C8oDJazD.js";
import "./identity-D07jLUrU.js";
import "./heartbeat-B91tt3Lu.js";
import "./pi-embedded-CswW9luA.js";
import "./internal-hooks-ClBId-OS.js";
import "./multimodal-rLHaC7mC.js";
import "./memory-search-C_KRQGyn.js";
import "./provider-catalog-BGWDC_kb.js";
import "./secret-input-BOkqlW2S.js";
import "./bindings-DFhSj1AD.js";
import "./resolve-route-jw3TrvbG.js";
import "./routing-GIGwPHsU.js";
import "./identity-file-AtA39mSs.js";
import "./outbound-runtime-QVIwx-7l.js";
import "./provider-env-vars-BCe4Mf67.js";
import "./provider-auth-input-WYCWu6tD.js";
import "./provider-model-minimax-CYJqVplI.js";
import "./provider-models-ld3k2AX4.js";
import "./anthropic-vertex-provider-7stzcdx2.js";
import "./models-config.providers.discovery-DnQmXa4L.js";
import "./text-runtime-BXjeqyjp.js";
import "./tool-catalog-DP3BJZz3.js";
import "./docker-Cq_Zfc99.js";
import "./sandbox-Dt1ZOIQc.js";
import "./common-DArxwyX8.js";
import "./image-ops-339pOknK.js";
import "./thinking-rAawjv9l.js";
import "./path-alias-guards-DMcFUz5-.js";
import "./sandbox-paths-D-zPD6r4.js";
import "./channel-actions-DZIaeZzh.js";
import "./mime-C0asv0HQ.js";
import "./ssrf-BifXvxQR.js";
import "./fetch-guard-BTY_UfUY.js";
import "./provider-web-search-BHcSeZzc.js";
import "./manager-DmufLZK2.js";
import "./config-schema-CKb8N9O1.js";
import "./channel-plugin-common-BX9jRob2.js";
import "./status-helpers-D9rzemcb.js";
import "./conversation-runtime-C56HP9yG.js";
import { i as createAttachedChannelResultAdapter } from "./channel-send-result-DXTOPdrt.js";
import "./runtime-whatsapp-boundary-BMepWtnk.js";
import "./pairing-store-Bh2YkvRc.js";
import "./json-store-BIglSbRG.js";
import "./secret-file-CEppYIjz.js";
import "./channel-config-schema-CrubAncY.js";
import "./setup-binary-BLFhRd_P.js";
import "./archive-Df6iZdzY.js";
import "./fs-safe-B2bSW_OI.js";
import "./signal-cli-install-UDdZA4-C.js";
import "./setup-wizard-proxy--gFyw5Eh.js";
import "./setup-C0zuBL_6.js";
import "./dm-policy-shared-ChGQzSwJ.js";
import "./hook-runtime-sGC9reX7.js";
import "./diagnostic-BtBb10Ry.js";
import "./templating-Bxm_rtRG.js";
import "./channel-reply-pipeline-DGoktEKi.js";
import "./reply-history-yYXk1ivL.js";
import "./commands-registry.data-_PzNk48W.js";
import "./commands-registry-ABOHrld2.js";
import "./frontmatter-Cd_6iJyw.js";
import "./env-overrides-B5iBORFt.js";
import "./skills-DAP6Yelo.js";
import "./skills-remote-BWQnXFJW.js";
import "./workspace-dirs-DhbPcAGw.js";
import "./pairing-token-CQLjZNmB.js";
import "./skill-commands-I_17sL0j.js";
import "./level-overrides-Ds8ppNbt.js";
import "./config-BnJ6TZJ3.js";
import "./routes-sPyjSxSY.js";
import "./ssh-tunnel-BSFzY8ID.js";
import "./server-middleware-Nz-5LEFG.js";
import "./logging-DFpZCDzJ.js";
import "./config-runtime-gAoemo4L.js";
import "./exec-approvals-CFIa5yg0.js";
import "./webhook-ingress-BBdXea0x.js";
import "./system-events-C2ngfmKm.js";
import "./ssrf-policy-BlgZt63j.js";
import "./provider-auth-ref-Dd8Ctvr9.js";
import "./provider-auth-helpers-CK3uH19P.js";
import "./provider-api-key-auth-CyO5NBeZ.js";
import "./runtime-env-CAlGDNNR.js";
import "./pairing-labels-BAkUxk5K.js";
import "./directory-runtime-BgPR38m_.js";
import "./read-only-account-inspect-B8C32NVU.js";
import "./src-Bc4Fm2Ge.js";
import "./web-media-Bo_Nxwjv.js";
import "./temp-path-Dgv6Kihk.js";
import "./media-understanding-Ch_w554d.js";
import "./web-media-BvqWsq_l.js";
import "./state-paths-C-WJBTWG.js";
import "./llm-task-Djjqflw_.js";
import "./pi-model-discovery-BNyYuKdC.js";
import "./exec-inline-eval-Di27nCtK.js";
import "./target-registry-CofRQwTE.js";
import "./external-content-Bq0zNXEh.js";
import "./brave-Hgb0hhUH.js";
import "./duckduckgo-B0xgAU7B.js";
import "./exa-Ce69v825.js";
import "./security-runtime-CoZ_sfgg.js";
import "./provider-usage-qQL8PJ0p.js";
import "./provider-onboard-CqAihpGi.js";
import "./perplexity-BNVyHCAP.js";
import "./stagger-DZXzssdr.js";
import "./command-secret-targets-wRPHRRib.js";
import "./delivery-queue-1A7pbB9e.js";
import "./channel-summary-DHY50TCU.js";
import "./follow-redirects-0oww1mdf.js";
import "./tool-policy-match-DnEs8-IP.js";
import "./runtime-CUG7pnyw.js";
import "./channel-status-C_NTcHvs.js";
import "./discord-core-DpMUVJg9.js";
import "./cli-runtime-CYhFvgcO.js";
import "./config-presence-Cqse_N2e.js";
import "./query-expansion-CZbIeYUY.js";
import "./search-manager-DsKEM_vh.js";
import "./acp-runtime-BS3aFfwU.js";
import "./telegram-core-BFUHv9d-.js";
import "./audit--q0z8yGd.js";
import "./gateway-runtime-Czf7LoCY.js";
import "./connection-auth-ZVYO9ARZ.js";
import "./mcp-config-CpCZwCwq.js";
import "./src-CiYLBx7o.js";
import { c as createFeishuClient, h as resolveFeishuAccount, o as probeFeishu } from "./feishu-B3toYTOi.js";
import "./minimal-mVF52Eoc.js";
import "./persistent-dedupe-D3_qFQJb.js";
import { a as getChatInfo, i as removeReactionFeishu, n as addReactionFeishu, o as getChatMembers, r as listReactionsFeishu, s as getFeishuMemberInfo } from "./reactions-BOJKVinB.js";
import { t as getFeishuRuntime } from "./runtime-Dlxd4mPr.js";
import { n as listFeishuDirectoryPeers, t as listFeishuDirectoryGroups } from "./directory.static-BAWMOvWW.js";
import { C as sendMediaFeishu, a as sendCardFeishu, c as sendStructuredCardFeishu, n as getMessageFeishu, o as sendMarkdownCardFeishu, s as sendMessageFeishu, t as editMessageFeishu } from "./send-UOyXcfxW.js";
import fs from "fs";
import path from "path";
//#region extensions/feishu/src/directory.ts
async function listFeishuDirectoryPeersLive(params) {
	const account = resolveFeishuAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	if (!account.configured) return listFeishuDirectoryPeers(params);
	try {
		const client = createFeishuClient(account);
		const peers = [];
		const limit = params.limit ?? 50;
		const response = await client.contact.user.list({ params: { page_size: Math.min(limit, 50) } });
		if (response.code !== 0) throw new Error(response.msg || `code ${response.code}`);
		for (const user of response.data?.items ?? []) {
			if (user.open_id) {
				const q = params.query?.trim().toLowerCase() || "";
				const name = user.name || "";
				if (!q || user.open_id.toLowerCase().includes(q) || name.toLowerCase().includes(q)) peers.push({
					kind: "user",
					id: user.open_id,
					name: name || void 0
				});
			}
			if (peers.length >= limit) break;
		}
		return peers;
	} catch (err) {
		if (params.fallbackToStatic === false) throw err instanceof Error ? err : /* @__PURE__ */ new Error("Feishu live peer lookup failed");
		return listFeishuDirectoryPeers(params);
	}
}
async function listFeishuDirectoryGroupsLive(params) {
	const account = resolveFeishuAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	if (!account.configured) return listFeishuDirectoryGroups(params);
	try {
		const client = createFeishuClient(account);
		const groups = [];
		const limit = params.limit ?? 50;
		const response = await client.im.chat.list({ params: { page_size: Math.min(limit, 100) } });
		if (response.code !== 0) throw new Error(response.msg || `code ${response.code}`);
		for (const chat of response.data?.items ?? []) {
			if (chat.chat_id) {
				const q = params.query?.trim().toLowerCase() || "";
				const name = chat.name || "";
				if (!q || chat.chat_id.toLowerCase().includes(q) || name.toLowerCase().includes(q)) groups.push({
					kind: "group",
					id: chat.chat_id,
					name: name || void 0
				});
			}
			if (groups.length >= limit) break;
		}
		return groups;
	} catch (err) {
		if (params.fallbackToStatic === false) throw err instanceof Error ? err : /* @__PURE__ */ new Error("Feishu live group lookup failed");
		return listFeishuDirectoryGroups(params);
	}
}
//#endregion
//#region extensions/feishu/src/outbound.ts
function normalizePossibleLocalImagePath(text) {
	const raw = text?.trim();
	if (!raw) return null;
	if (/\s/.test(raw)) return null;
	if (/^(https?:\/\/|data:|file:\/\/)/i.test(raw)) return null;
	const ext = path.extname(raw).toLowerCase();
	if (![
		".jpg",
		".jpeg",
		".png",
		".gif",
		".webp",
		".bmp",
		".ico",
		".tiff"
	].includes(ext)) return null;
	if (!path.isAbsolute(raw)) return null;
	if (!fs.existsSync(raw)) return null;
	try {
		if (!fs.statSync(raw).isFile()) return null;
	} catch {
		return null;
	}
	return raw;
}
function shouldUseCard(text) {
	return /```[\s\S]*?```/.test(text) || /\|.+\|[\r\n]+\|[-:| ]+\|/.test(text);
}
function resolveReplyToMessageId(params) {
	const replyToId = params.replyToId?.trim();
	if (replyToId) return replyToId;
	if (params.threadId == null) return;
	return String(params.threadId).trim() || void 0;
}
async function sendOutboundText(params) {
	const { cfg, to, text, accountId, replyToMessageId } = params;
	const renderMode = resolveFeishuAccount({
		cfg,
		accountId
	}).config?.renderMode ?? "auto";
	if (renderMode === "card" || renderMode === "auto" && shouldUseCard(text)) return sendMarkdownCardFeishu({
		cfg,
		to,
		text,
		accountId,
		replyToMessageId
	});
	return sendMessageFeishu({
		cfg,
		to,
		text,
		accountId,
		replyToMessageId
	});
}
const feishuOutbound = {
	deliveryMode: "direct",
	chunker: (text, limit) => getFeishuRuntime().channel.text.chunkMarkdownText(text, limit),
	chunkerMode: "markdown",
	textChunkLimit: 4e3,
	...createAttachedChannelResultAdapter({
		channel: "feishu",
		sendText: async ({ cfg, to, text, accountId, replyToId, threadId, mediaLocalRoots, identity }) => {
			const replyToMessageId = resolveReplyToMessageId({
				replyToId,
				threadId
			});
			const localImagePath = normalizePossibleLocalImagePath(text);
			if (localImagePath) try {
				return await sendMediaFeishu({
					cfg,
					to,
					mediaUrl: localImagePath,
					accountId: accountId ?? void 0,
					replyToMessageId,
					mediaLocalRoots
				});
			} catch (err) {
				console.error(`[feishu] local image path auto-send failed:`, err);
			}
			const renderMode = resolveFeishuAccount({
				cfg,
				accountId: accountId ?? void 0
			}).config?.renderMode ?? "auto";
			if (renderMode === "card" || renderMode === "auto" && shouldUseCard(text)) {
				const header = identity ? {
					title: identity.emoji ? `${identity.emoji} ${identity.name ?? ""}`.trim() : identity.name ?? "",
					template: "blue"
				} : void 0;
				return await sendStructuredCardFeishu({
					cfg,
					to,
					text,
					replyToMessageId,
					replyInThread: threadId != null && !replyToId,
					accountId: accountId ?? void 0,
					header: header?.title ? header : void 0
				});
			}
			return await sendOutboundText({
				cfg,
				to,
				text,
				accountId: accountId ?? void 0,
				replyToMessageId
			});
		},
		sendMedia: async ({ cfg, to, text, mediaUrl, accountId, mediaLocalRoots, replyToId, threadId }) => {
			const replyToMessageId = resolveReplyToMessageId({
				replyToId,
				threadId
			});
			if (text?.trim()) await sendOutboundText({
				cfg,
				to,
				text,
				accountId: accountId ?? void 0,
				replyToMessageId
			});
			if (mediaUrl) try {
				return await sendMediaFeishu({
					cfg,
					to,
					mediaUrl,
					accountId: accountId ?? void 0,
					mediaLocalRoots,
					replyToMessageId
				});
			} catch (err) {
				console.error(`[feishu] sendMediaFeishu failed:`, err);
				return await sendOutboundText({
					cfg,
					to,
					text: `📎 ${mediaUrl}`,
					accountId: accountId ?? void 0,
					replyToMessageId
				});
			}
			return await sendOutboundText({
				cfg,
				to,
				text: text ?? "",
				accountId: accountId ?? void 0,
				replyToMessageId
			});
		}
	})
};
//#endregion
//#region extensions/feishu/src/pins.ts
function assertFeishuPinApiSuccess(response, action) {
	if (response.code !== 0) throw new Error(`Feishu ${action} failed: ${response.msg || `code ${response.code}`}`);
}
function normalizePin(pin) {
	return {
		messageId: pin.message_id,
		chatId: pin.chat_id,
		operatorId: pin.operator_id,
		operatorIdType: pin.operator_id_type,
		createTime: pin.create_time
	};
}
async function createPinFeishu(params) {
	const account = resolveFeishuAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	if (!account.configured) throw new Error(`Feishu account "${account.accountId}" not configured`);
	const response = await createFeishuClient(account).im.pin.create({ data: { message_id: params.messageId } });
	assertFeishuPinApiSuccess(response, "pin create");
	return response.data?.pin ? normalizePin(response.data.pin) : null;
}
async function removePinFeishu(params) {
	const account = resolveFeishuAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	if (!account.configured) throw new Error(`Feishu account "${account.accountId}" not configured`);
	assertFeishuPinApiSuccess(await createFeishuClient(account).im.pin.delete({ path: { message_id: params.messageId } }), "pin delete");
}
async function listPinsFeishu(params) {
	const account = resolveFeishuAccount({
		cfg: params.cfg,
		accountId: params.accountId
	});
	if (!account.configured) throw new Error(`Feishu account "${account.accountId}" not configured`);
	const response = await createFeishuClient(account).im.pin.list({ params: {
		chat_id: params.chatId,
		...params.startTime ? { start_time: params.startTime } : {},
		...params.endTime ? { end_time: params.endTime } : {},
		...typeof params.pageSize === "number" ? { page_size: Math.max(1, Math.min(100, Math.floor(params.pageSize))) } : {},
		...params.pageToken ? { page_token: params.pageToken } : {}
	} });
	assertFeishuPinApiSuccess(response, "pin list");
	return {
		chatId: params.chatId,
		pins: (response.data?.items ?? []).map(normalizePin),
		hasMore: response.data?.has_more === true,
		pageToken: response.data?.page_token
	};
}
//#endregion
//#region extensions/feishu/src/channel.runtime.ts
const feishuChannelRuntime = {
	listFeishuDirectoryGroupsLive,
	listFeishuDirectoryPeersLive,
	feishuOutbound: { ...feishuOutbound },
	createPinFeishu,
	listPinsFeishu,
	removePinFeishu,
	probeFeishu,
	addReactionFeishu,
	listReactionsFeishu,
	removeReactionFeishu,
	getChatInfo,
	getChatMembers,
	getFeishuMemberInfo,
	editMessageFeishu,
	getMessageFeishu,
	sendCardFeishu,
	sendMessageFeishu
};
//#endregion
export { feishuChannelRuntime };
