// openclaw/dist/channel.runtime-DVZwGdas.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import "./links-8xRhWBQL.js";
import "./paths-C4Oy7wjk.js";
import "./auth-profiles-CLtbIyEz.js";
import "./agent-scope-bjWqU22i.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import "./model-selection-BnFtDmP7.js";
import "./io-CZZeeo8R.js";
import "./shell-env-BOu7XeT_.js";
import "./safe-text-yavot2qw.js";
import "./version-CD3oP1-d.js";
import "./env-substitution-O6uabUmO.js";
import "./includes-Bj3eLUWH.js";
import "./zod-schema.providers-core-COy5nBQ0.js";
import "./legacy-web-search-BB-ZHEhz.js";
import "./registry-C5UkPpaO.js";
import "./config-state-Br0ucqMb.js";
import "./min-host-version-RMBWtIAR.js";
import "./manifest-registry-B5JNQdOM.js";
import "./runtime-guard-PhQ6PwQa.js";
import "./avatar-policy-B5nOfso_.js";
import "./ip-Ce8EDTBZ.js";
import "./zod-schema.agent-runtime-Dtg4Jy6G.js";
import "./zod-schema.core-BuVz8Rk7.js";
import "./config-Bqe8DjNc.js";
import "./file-lock-BJsnXQjL.js";
import "./audit-fs-DTQIjp2r.js";
import "./resolve-B7S4PlnO.js";
import "./profiles-B_nscSgH.js";
import "./tailscale-CxY-91GZ.js";
import "./tailnet-DHiKSG7V.js";
import "./net-JAg9zL0G.js";
import "./auth-Buq0Niri.js";
import "./credentials-D8pfRimu.js";
import "./message-channel-BliByQBl.js";
import "./store-_QDcIMYk.js";
import "./runtime-Iz8uZ7EU.js";
import "./plugins-B09-vgme.js";
import "./sessions-BwqyiEaq.js";
import "./paths-rhN9LKM_.js";
import "./session-write-lock-Db3SWNU3.js";
import "./method-scopes-B2vIWWxl.js";
import "./call-BhjE3oIo.js";
import "./prompt-style-HkzccPBT.js";
import "./ports-lsof-Dlzvm55V.js";
import "./restart-stale-pids-OP_Aybe0.js";
import "./ports-B3yeEL-n.js";
import "./logging-D4CVa7JL.js";
import "./commands-2T-1FAo7.js";
import "./issue-format-C8oDJazD.js";
import "./identity-D07jLUrU.js";
import "./heartbeat-B91tt3Lu.js";
import { Bg as resolveMentionGatingWithBypass, C_ as resolveInboundRouteEnvelopeBuilderWithRuntime } from "./pi-embedded-CswW9luA.js";
import "./internal-hooks-ClBId-OS.js";
import "./multimodal-rLHaC7mC.js";
import "./memory-search-C_KRQGyn.js";
import "./provider-catalog-BGWDC_kb.js";
import "./secret-input-BOkqlW2S.js";
import "./bindings-DFhSj1AD.js";
import "./resolve-route-jw3TrvbG.js";
import "./routing-GIGwPHsU.js";
import "./identity-file-AtA39mSs.js";
import "./outbound-runtime-QVIwx-7l.js";
import "./provider-env-vars-BCe4Mf67.js";
import "./provider-auth-input-WYCWu6tD.js";
import "./provider-model-minimax-CYJqVplI.js";
import "./provider-models-ld3k2AX4.js";
import "./anthropic-vertex-provider-7stzcdx2.js";
import "./models-config.providers.discovery-DnQmXa4L.js";
import "./text-runtime-BXjeqyjp.js";
import "./tool-catalog-DP3BJZz3.js";
import "./docker-Cq_Zfc99.js";
import "./sandbox-Dt1ZOIQc.js";
import "./common-DArxwyX8.js";
import "./image-ops-339pOknK.js";
import "./thinking-rAawjv9l.js";
import "./path-alias-guards-DMcFUz5-.js";
import "./sandbox-paths-D-zPD6r4.js";
import "./channel-actions-DZIaeZzh.js";
import "./mime-C0asv0HQ.js";
import "./ssrf-BifXvxQR.js";
import "./fetch-guard-BTY_UfUY.js";
import "./provider-web-search-BHcSeZzc.js";
import "./manager-DmufLZK2.js";
import "./config-schema-CKb8N9O1.js";
import "./channel-plugin-common-BX9jRob2.js";
import { a as warnMissingProviderGroupPolicyFallbackOnce, n as resolveAllowlistProviderRuntimeGroupPolicy, r as resolveDefaultGroupPolicy, t as GROUP_POLICY_BLOCKED_LABEL } from "./runtime-group-policy-DqDSPQSR.js";
import "./status-helpers-D9rzemcb.js";
import "./conversation-runtime-C56HP9yG.js";
import "./runtime-whatsapp-boundary-BMepWtnk.js";
import "./pairing-store-Bh2YkvRc.js";
import "./json-store-BIglSbRG.js";
import { n as createChannelPairingController } from "./channel-pairing-CA9BsqQV.js";
import "./secret-file-CEppYIjz.js";
import "./channel-config-schema-CrubAncY.js";
import "./setup-binary-BLFhRd_P.js";
import "./archive-Df6iZdzY.js";
import "./fs-safe-B2bSW_OI.js";
import "./signal-cli-install-UDdZA4-C.js";
import "./setup-wizard-proxy--gFyw5Eh.js";
import "./setup-C0zuBL_6.js";
import { o as resolveDmGroupAccessWithLists } from "./dm-policy-shared-ChGQzSwJ.js";
import { a as resolveSenderScopedGroupPolicy, t as evaluateGroupRouteAccessForPolicy } from "./group-access-DbAQwRuM.js";
import { i as deliverTextOrMediaReply, p as resolveSendableOutboundReplyParts } from "./reply-payload-BTv3KfjW.js";
import "./hook-runtime-sGC9reX7.js";
import "./diagnostic-BtBb10Ry.js";
import "./templating-Bxm_rtRG.js";
import { t as createChannelReplyPipeline } from "./channel-reply-pipeline-DGoktEKi.js";
import "./reply-history-yYXk1ivL.js";
import "./commands-registry.data-_PzNk48W.js";
import "./commands-registry-ABOHrld2.js";
import "./frontmatter-Cd_6iJyw.js";
import "./env-overrides-B5iBORFt.js";
import "./skills-DAP6Yelo.js";
import "./skills-remote-BWQnXFJW.js";
import "./workspace-dirs-DhbPcAGw.js";
import "./pairing-token-CQLjZNmB.js";
import "./skill-commands-I_17sL0j.js";
import "./level-overrides-Ds8ppNbt.js";
import "./config-BnJ6TZJ3.js";
import "./routes-sPyjSxSY.js";
import "./ssh-tunnel-BSFzY8ID.js";
import "./server-middleware-Nz-5LEFG.js";
import "./logging-DFpZCDzJ.js";
import "./config-runtime-gAoemo4L.js";
import { n as isDangerousNameMatchingEnabled } from "./dangerous-name-matching-TMkiuJ0K.js";
import "./exec-approvals-CFIa5yg0.js";
import { a as resolveWebhookTargetWithAuthOrReject, c as withResolvedWebhookRequestPipeline, n as registerWebhookTargetWithPluginRoute, x as readJsonWebhookBodyOrReject, y as createWebhookInFlightLimiter } from "./webhook-ingress-BBdXea0x.js";
import "./system-events-C2ngfmKm.js";
import "./ssrf-policy-BlgZt63j.js";
import "./provider-auth-ref-Dd8Ctvr9.js";
import "./provider-auth-helpers-CK3uH19P.js";
import "./provider-api-key-auth-CyO5NBeZ.js";
import "./runtime-env-CAlGDNNR.js";
import "./pairing-labels-BAkUxk5K.js";
import "./directory-runtime-BgPR38m_.js";
import "./read-only-account-inspect-B8C32NVU.js";
import "./src-Bc4Fm2Ge.js";
import "./web-media-Bo_Nxwjv.js";
import "./temp-path-Dgv6Kihk.js";
import "./media-understanding-Ch_w554d.js";
import "./web-media-BvqWsq_l.js";
import "./state-paths-C-WJBTWG.js";
import "./llm-task-Djjqflw_.js";
import "./pi-model-discovery-BNyYuKdC.js";
import "./exec-inline-eval-Di27nCtK.js";
import "./target-registry-CofRQwTE.js";
import "./external-content-Bq0zNXEh.js";
import "./brave-Hgb0hhUH.js";
import "./duckduckgo-B0xgAU7B.js";
import "./exa-Ce69v825.js";
import "./security-runtime-CoZ_sfgg.js";
import "./provider-usage-qQL8PJ0p.js";
import "./provider-onboard-CqAihpGi.js";
import "./perplexity-BNVyHCAP.js";
import "./stagger-DZXzssdr.js";
import "./command-secret-targets-wRPHRRib.js";
import "./delivery-queue-1A7pbB9e.js";
import "./channel-summary-DHY50TCU.js";
import "./follow-redirects-0oww1mdf.js";
import "./tool-policy-match-DnEs8-IP.js";
import "./runtime-CUG7pnyw.js";
import { n as resolveWebhookPath } from "./webhook-path-DXVsrFbg.js";
import "./channel-status-C_NTcHvs.js";
import "./discord-core-DpMUVJg9.js";
import "./cli-runtime-CYhFvgcO.js";
import "./config-presence-Cqse_N2e.js";
import "./query-expansion-CZbIeYUY.js";
import "./search-manager-DsKEM_vh.js";
import "./acp-runtime-BS3aFfwU.js";
import "./telegram-core-BFUHv9d-.js";
import "./audit--q0z8yGd.js";
import "./gateway-runtime-Czf7LoCY.js";
import "./connection-auth-ZVYO9ARZ.js";
import "./mcp-config-CpCZwCwq.js";
import "./channel-setup-B0zShv38.js";
import "./googlechat--DBsOlsq.js";
import "./jws-D3zx7ylK.js";
import { c as sendGoogleChatMessage, d as verifyGoogleChatRequest, i as downloadGoogleChatMedia, l as updateGoogleChatMessage, n as deleteGoogleChatMessage, s as probeGoogleChat, u as uploadGoogleChatAttachment } from "./api-BS7lTvAw.js";
import "./json-bigint-CDcZo0Kq.js";
import { t as getGoogleChatRuntime } from "./runtime-B2AIp8U0.js";
//#region extensions/googlechat/src/monitor-access.ts
function normalizeUserId(raw) {
	const trimmed = raw?.trim() ?? "";
	if (!trimmed) return "";
	return trimmed.replace(/^users\//i, "").toLowerCase();
}
function isEmailLike(value) {
	return value.includes("@");
}
function isSenderAllowed(senderId, senderEmail, allowFrom, allowNameMatching = false) {
	if (allowFrom.includes("*")) return true;
	const normalizedSenderId = normalizeUserId(senderId);
	const normalizedEmail = senderEmail?.trim().toLowerCase() ?? "";
	return allowFrom.some((entry) => {
		const normalized = String(entry).trim().toLowerCase();
		if (!normalized) return false;
		const withoutPrefix = normalized.replace(/^(googlechat|google-chat|gchat):/i, "");
		if (withoutPrefix.startsWith("users/")) return normalizeUserId(withoutPrefix) === normalizedSenderId;
		if (allowNameMatching && normalizedEmail && isEmailLike(withoutPrefix)) return withoutPrefix === normalizedEmail;
		return withoutPrefix.replace(/^users\//i, "") === normalizedSenderId;
	});
}
function resolveGroupConfig(params) {
	const { groupId, groupName, groups } = params;
	const entries = groups ?? {};
	if (Object.keys(entries).length === 0) return {
		entry: void 0,
		allowlistConfigured: false
	};
	const normalizedName = groupName?.trim().toLowerCase();
	let entry = [
		groupId,
		groupName ?? "",
		normalizedName ?? ""
	].filter(Boolean).map((candidate) => entries[candidate]).find(Boolean);
	if (!entry && normalizedName) entry = entries[normalizedName];
	const fallback = entries["*"];
	return {
		entry: entry ?? fallback,
		allowlistConfigured: true,
		fallback
	};
}
function extractMentionInfo(annotations, botUser) {
	const mentionAnnotations = annotations.filter((entry) => entry.type === "USER_MENTION");
	const hasAnyMention = mentionAnnotations.length > 0;
	const botTargets = new Set(["users/app", botUser?.trim()].filter(Boolean));
	return {
		hasAnyMention,
		wasMentioned: mentionAnnotations.some((entry) => {
			const userName = entry.userMention?.user?.name;
			if (!userName) return false;
			if (botTargets.has(userName)) return true;
			return normalizeUserId(userName) === "app";
		})
	};
}
const warnedDeprecatedUsersEmailAllowFrom = /* @__PURE__ */ new Set();
function warnDeprecatedUsersEmailEntries(logVerbose, entries) {
	const deprecated = entries.map((v) => String(v).trim()).filter((v) => /^users\/.+@.+/i.test(v));
	if (deprecated.length === 0) return;
	const key = deprecated.map((v) => v.toLowerCase()).sort().join(",");
	if (warnedDeprecatedUsersEmailAllowFrom.has(key)) return;
	warnedDeprecatedUsersEmailAllowFrom.add(key);
	logVerbose(`Deprecated allowFrom entry detected: "users/<email>" is no longer treated as an email allowlist. Use raw email (alice@example.com) or immutable user id (users/<id>). entries=${deprecated.join(", ")}`);
}
async function applyGoogleChatInboundAccessPolicy(params) {
	const { account, config, core, space, message, isGroup, senderId, senderName, senderEmail, rawBody, statusSink, logVerbose } = params;
	const allowNameMatching = isDangerousNameMatchingEnabled(account.config);
	const spaceId = space.name ?? "";
	const pairing = createChannelPairingController({
		core,
		channel: "googlechat",
		accountId: account.accountId
	});
	const defaultGroupPolicy = resolveDefaultGroupPolicy(config);
	const { groupPolicy, providerMissingFallbackApplied } = resolveAllowlistProviderRuntimeGroupPolicy({
		providerConfigPresent: config.channels?.googlechat !== void 0,
		groupPolicy: account.config.groupPolicy,
		defaultGroupPolicy
	});
	warnMissingProviderGroupPolicyFallbackOnce({
		providerMissingFallbackApplied,
		providerKey: "googlechat",
		accountId: account.accountId,
		blockedLabel: GROUP_POLICY_BLOCKED_LABEL.space,
		log: logVerbose
	});
	const groupConfigResolved = resolveGroupConfig({
		groupId: spaceId,
		groupName: space.displayName ?? null,
		groups: account.config.groups ?? void 0
	});
	const groupEntry = groupConfigResolved.entry;
	const groupUsers = groupEntry?.users ?? account.config.groupAllowFrom ?? [];
	let effectiveWasMentioned;
	if (isGroup) {
		const groupAllowlistConfigured = groupConfigResolved.allowlistConfigured;
		const routeAccess = evaluateGroupRouteAccessForPolicy({
			groupPolicy,
			routeAllowlistConfigured: groupAllowlistConfigured,
			routeMatched: Boolean(groupEntry),
			routeEnabled: groupEntry?.enabled !== false && groupEntry?.allow !== false
		});
		if (!routeAccess.allowed) {
			if (routeAccess.reason === "disabled") logVerbose(`drop group message (groupPolicy=disabled, space=${spaceId})`);
			else if (routeAccess.reason === "empty_allowlist") logVerbose(`drop group message (groupPolicy=allowlist, no allowlist, space=${spaceId})`);
			else if (routeAccess.reason === "route_not_allowlisted") logVerbose(`drop group message (not allowlisted, space=${spaceId})`);
			else if (routeAccess.reason === "route_disabled") logVerbose(`drop group message (space disabled, space=${spaceId})`);
			return { ok: false };
		}
		if (groupUsers.length > 0) {
			const normalizedGroupUsers = groupUsers.map((v) => String(v));
			warnDeprecatedUsersEmailEntries(logVerbose, normalizedGroupUsers);
			if (!isSenderAllowed(senderId, senderEmail, normalizedGroupUsers, allowNameMatching)) {
				logVerbose(`drop group message (sender not allowed, ${senderId})`);
				return { ok: false };
			}
		}
	}
	const dmPolicy = account.config.dm?.policy ?? "pairing";
	const configAllowFrom = (account.config.dm?.allowFrom ?? []).map((v) => String(v));
	const normalizedGroupUsers = groupUsers.map((v) => String(v));
	const senderGroupPolicy = resolveSenderScopedGroupPolicy({
		groupPolicy,
		groupAllowFrom: normalizedGroupUsers
	});
	const shouldComputeAuth = core.channel.commands.shouldComputeCommandAuthorized(rawBody, config);
	const access = resolveDmGroupAccessWithLists({
		isGroup,
		dmPolicy,
		groupPolicy: senderGroupPolicy,
		allowFrom: configAllowFrom,
		groupAllowFrom: normalizedGroupUsers,
		storeAllowFrom: !isGroup && dmPolicy !== "allowlist" && (dmPolicy !== "open" || shouldComputeAuth) ? await pairing.readAllowFromStore().catch(() => []) : [],
		groupAllowFromFallbackToAllowFrom: false,
		isSenderAllowed: (allowFrom) => isSenderAllowed(senderId, senderEmail, allowFrom, allowNameMatching)
	});
	const effectiveAllowFrom = access.effectiveAllowFrom;
	const effectiveGroupAllowFrom = access.effectiveGroupAllowFrom;
	warnDeprecatedUsersEmailEntries(logVerbose, effectiveAllowFrom);
	const commandAllowFrom = isGroup ? effectiveGroupAllowFrom : effectiveAllowFrom;
	const useAccessGroups = config.commands?.useAccessGroups !== false;
	const senderAllowedForCommands = isSenderAllowed(senderId, senderEmail, commandAllowFrom, allowNameMatching);
	const commandAuthorized = shouldComputeAuth ? core.channel.commands.resolveCommandAuthorizedFromAuthorizers({
		useAccessGroups,
		authorizers: [{
			configured: commandAllowFrom.length > 0,
			allowed: senderAllowedForCommands
		}]
	}) : void 0;
	if (isGroup) {
		const requireMention = groupEntry?.requireMention ?? account.config.requireMention ?? true;
		const mentionInfo = extractMentionInfo(message.annotations ?? [], account.config.botUser);
		const allowTextCommands = core.channel.commands.shouldHandleTextCommands({
			cfg: config,
			surface: "googlechat"
		});
		const mentionGate = resolveMentionGatingWithBypass({
			isGroup: true,
			requireMention,
			canDetectMention: true,
			wasMentioned: mentionInfo.wasMentioned,
			implicitMention: false,
			hasAnyMention: mentionInfo.hasAnyMention,
			allowTextCommands,
			hasControlCommand: core.channel.text.hasControlCommand(rawBody, config),
			commandAuthorized: commandAuthorized === true
		});
		effectiveWasMentioned = mentionGate.effectiveWasMentioned;
		if (mentionGate.shouldSkip) {
			logVerbose(`drop group message (mention required, space=${spaceId})`);
			return { ok: false };
		}
	}
	if (isGroup && access.decision !== "allow") {
		logVerbose(`drop group message (sender policy blocked, reason=${access.reason}, space=${spaceId})`);
		return { ok: false };
	}
	if (!isGroup) {
		if (account.config.dm?.enabled === false) {
			logVerbose(`Blocked Google Chat DM from ${senderId} (dmPolicy=disabled)`);
			return { ok: false };
		}
		if (access.decision !== "allow") {
			if (access.decision === "pairing") await pairing.issueChallenge({
				senderId,
				senderIdLine: `Your Google Chat user id: ${senderId}`,
				meta: {
					name: senderName || void 0,
					email: senderEmail
				},
				onCreated: () => {
					logVerbose(`googlechat pairing request sender=${senderId}`);
				},
				sendPairingReply: async (text) => {
					await sendGoogleChatMessage({
						account,
						space: spaceId,
						text
					});
					statusSink?.({ lastOutboundAt: Date.now() });
				},
				onReplyError: (err) => {
					logVerbose(`pairing reply failed for ${senderId}: ${String(err)}`);
				}
			});
			else logVerbose(`Blocked unauthorized Google Chat sender ${senderId} (dmPolicy=${dmPolicy})`);
			return { ok: false };
		}
	}
	if (isGroup && core.channel.commands.isControlCommandMessage(rawBody, config) && commandAuthorized !== true) {
		logVerbose(`googlechat: drop control command from ${senderId}`);
		return { ok: false };
	}
	return {
		ok: true,
		commandAuthorized,
		effectiveWasMentioned,
		groupSystemPrompt: groupEntry?.systemPrompt?.trim() || void 0
	};
}
//#endregion
//#region extensions/googlechat/src/monitor-webhook.ts
function extractBearerToken(header) {
	const authHeader = Array.isArray(header) ? String(header[0] ?? "") : String(header ?? "");
	return authHeader.toLowerCase().startsWith("bearer ") ? authHeader.slice(7).trim() : "";
}
const ADD_ON_PREAUTH_MAX_BYTES = 16 * 1024;
const ADD_ON_PREAUTH_TIMEOUT_MS = 3e3;
function parseGoogleChatInboundPayload(raw, res) {
	if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
		res.statusCode = 400;
		res.end("invalid payload");
		return { ok: false };
	}
	let eventPayload = raw;
	let addOnBearerToken = "";
	const rawObj = raw;
	if (rawObj.commonEventObject?.hostApp === "CHAT" && rawObj.chat?.messagePayload) {
		const chat = rawObj.chat;
		const messagePayload = chat.messagePayload;
		eventPayload = {
			type: "MESSAGE",
			space: messagePayload?.space,
			message: messagePayload?.message,
			user: chat.user,
			eventTime: chat.eventTime
		};
		addOnBearerToken = String(rawObj.authorizationEventObject?.systemIdToken ?? "").trim();
	}
	const event = eventPayload;
	const eventType = event.type ?? eventPayload.eventType;
	if (typeof eventType !== "string") {
		res.statusCode = 400;
		res.end("invalid payload");
		return { ok: false };
	}
	if (!event.space || typeof event.space !== "object" || Array.isArray(event.space)) {
		res.statusCode = 400;
		res.end("invalid payload");
		return { ok: false };
	}
	if (eventType === "MESSAGE") {
		if (!event.message || typeof event.message !== "object" || Array.isArray(event.message)) {
			res.statusCode = 400;
			res.end("invalid payload");
			return { ok: false };
		}
	}
	return {
		ok: true,
		event,
		addOnBearerToken
	};
}
function createGoogleChatWebhookRequestHandler(params) {
	return async (req, res) => {
		return await withResolvedWebhookRequestPipeline({
			req,
			res,
			targetsByPath: params.webhookTargets,
			allowMethods: ["POST"],
			requireJsonContentType: true,
			inFlightLimiter: params.webhookInFlightLimiter,
			handle: async ({ targets }) => {
				const headerBearer = extractBearerToken(req.headers.authorization);
				let selectedTarget = null;
				let parsedEvent = null;
				const readAndParseEvent = async (profile) => {
					const body = await readJsonWebhookBodyOrReject({
						req,
						res,
						profile,
						...profile === "pre-auth" ? {
							maxBytes: ADD_ON_PREAUTH_MAX_BYTES,
							timeoutMs: ADD_ON_PREAUTH_TIMEOUT_MS
						} : {},
						emptyObjectOnEmpty: false,
						invalidJsonMessage: "invalid payload"
					});
					if (!body.ok) return null;
					const parsed = parseGoogleChatInboundPayload(body.value, res);
					return parsed.ok ? parsed : null;
				};
				if (headerBearer) {
					selectedTarget = await resolveWebhookTargetWithAuthOrReject({
						targets,
						res,
						isMatch: async (target) => {
							return (await verifyGoogleChatRequest({
								bearer: headerBearer,
								audienceType: target.audienceType,
								audience: target.audience,
								expectedAddOnPrincipal: target.account.config.appPrincipal
							})).ok;
						}
					});
					if (!selectedTarget) return true;
					const parsed = await readAndParseEvent("post-auth");
					if (!parsed) return true;
					parsedEvent = parsed.event;
				} else {
					const parsed = await readAndParseEvent("pre-auth");
					if (!parsed) return true;
					parsedEvent = parsed.event;
					if (!parsed.addOnBearerToken) {
						res.statusCode = 401;
						res.end("unauthorized");
						return true;
					}
					selectedTarget = await resolveWebhookTargetWithAuthOrReject({
						targets,
						res,
						isMatch: async (target) => {
							return (await verifyGoogleChatRequest({
								bearer: parsed.addOnBearerToken,
								audienceType: target.audienceType,
								audience: target.audience,
								expectedAddOnPrincipal: target.account.config.appPrincipal
							})).ok;
						}
					});
					if (!selectedTarget) return true;
				}
				if (!selectedTarget || !parsedEvent) {
					res.statusCode = 401;
					res.end("unauthorized");
					return true;
				}
				const dispatchTarget = selectedTarget;
				dispatchTarget.statusSink?.({ lastInboundAt: Date.now() });
				params.processEvent(parsedEvent, dispatchTarget).catch((err) => {
					dispatchTarget.runtime.error?.(`[${dispatchTarget.account.accountId}] Google Chat webhook failed: ${String(err)}`);
				});
				res.statusCode = 200;
				res.setHeader("Content-Type", "application/json");
				res.end("{}");
				return true;
			}
		});
	};
}
//#endregion
//#region extensions/googlechat/src/monitor.ts
const webhookTargets = /* @__PURE__ */ new Map();
const googleChatWebhookRequestHandler = createGoogleChatWebhookRequestHandler({
	webhookTargets,
	webhookInFlightLimiter: createWebhookInFlightLimiter(),
	processEvent: async (event, target) => {
		await processGoogleChatEvent(event, target);
	}
});
function logVerbose(core, runtime, message) {
	if (core.logging.shouldLogVerbose()) runtime.log?.(`[googlechat] ${message}`);
}
function registerGoogleChatWebhookTarget(target) {
	return registerWebhookTargetWithPluginRoute({
		targetsByPath: webhookTargets,
		target,
		route: {
			auth: "plugin",
			match: "exact",
			pluginId: "googlechat",
			source: "googlechat-webhook",
			accountId: target.account.accountId,
			log: target.runtime.log,
			handler: async (req, res) => {
				if (!await handleGoogleChatWebhookRequest(req, res) && !res.headersSent) {
					res.statusCode = 404;
					res.setHeader("Content-Type", "text/plain; charset=utf-8");
					res.end("Not Found");
				}
			}
		}
	}).unregister;
}
function normalizeAudienceType(value) {
	const normalized = value?.trim().toLowerCase();
	if (normalized === "app-url" || normalized === "app_url" || normalized === "app") return "app-url";
	if (normalized === "project-number" || normalized === "project_number" || normalized === "project") return "project-number";
}
async function handleGoogleChatWebhookRequest(req, res) {
	return await googleChatWebhookRequestHandler(req, res);
}
async function processGoogleChatEvent(event, target) {
	if ((event.type ?? event.eventType) !== "MESSAGE") return;
	if (!event.message || !event.space) return;
	await processMessageWithPipeline({
		event,
		account: target.account,
		config: target.config,
		runtime: target.runtime,
		core: target.core,
		statusSink: target.statusSink,
		mediaMaxMb: target.mediaMaxMb
	});
}
/**
* Resolve bot display name with fallback chain:
* 1. Account config name
* 2. Agent name from config
* 3. "OpenClaw" as generic fallback
*/
function resolveBotDisplayName(params) {
	const { accountName, agentId, config } = params;
	if (accountName?.trim()) return accountName.trim();
	const agent = config.agents?.list?.find((a) => a.id === agentId);
	if (agent?.name?.trim()) return agent.name.trim();
	return "OpenClaw";
}
async function processMessageWithPipeline(params) {
	const { event, account, config, runtime, core, statusSink, mediaMaxMb } = params;
	const space = event.space;
	const message = event.message;
	if (!space || !message) return;
	const spaceId = space.name ?? "";
	if (!spaceId) return;
	const isGroup = (space.type ?? "").toUpperCase() !== "DM";
	const sender = message.sender ?? event.user;
	const senderId = sender?.name ?? "";
	const senderName = sender?.displayName ?? "";
	const senderEmail = sender?.email ?? void 0;
	if (!(account.config.allowBots === true)) {
		if (sender?.type?.toUpperCase() === "BOT") {
			logVerbose(core, runtime, `skip bot-authored message (${senderId || "unknown"})`);
			return;
		}
		if (senderId === "users/app") {
			logVerbose(core, runtime, "skip app-authored message");
			return;
		}
	}
	const messageText = (message.argumentText ?? message.text ?? "").trim();
	const attachments = message.attachment ?? [];
	const hasMedia = attachments.length > 0;
	const rawBody = messageText || (hasMedia ? "<media:attachment>" : "");
	if (!rawBody) return;
	const access = await applyGoogleChatInboundAccessPolicy({
		account,
		config,
		core,
		space,
		message,
		isGroup,
		senderId,
		senderName,
		senderEmail,
		rawBody,
		statusSink,
		logVerbose: (message) => logVerbose(core, runtime, message)
	});
	if (!access.ok) return;
	const { commandAuthorized, effectiveWasMentioned, groupSystemPrompt } = access;
	const { route, buildEnvelope } = resolveInboundRouteEnvelopeBuilderWithRuntime({
		cfg: config,
		channel: "googlechat",
		accountId: account.accountId,
		peer: {
			kind: isGroup ? "group" : "direct",
			id: spaceId
		},
		runtime: core.channel,
		sessionStore: config.session?.store
	});
	let mediaPath;
	let mediaType;
	if (attachments.length > 0) {
		const first = attachments[0];
		const attachmentData = await downloadAttachment(first, account, mediaMaxMb, core);
		if (attachmentData) {
			mediaPath = attachmentData.path;
			mediaType = attachmentData.contentType;
		}
	}
	const fromLabel = isGroup ? space.displayName || `space:${spaceId}` : senderName || `user:${senderId}`;
	const { storePath, body } = buildEnvelope({
		channel: "Google Chat",
		from: fromLabel,
		timestamp: event.eventTime ? Date.parse(event.eventTime) : void 0,
		body: rawBody
	});
	const ctxPayload = core.channel.reply.finalizeInboundContext({
		Body: body,
		BodyForAgent: rawBody,
		RawBody: rawBody,
		CommandBody: rawBody,
		From: `googlechat:${senderId}`,
		To: `googlechat:${spaceId}`,
		SessionKey: route.sessionKey,
		AccountId: route.accountId,
		ChatType: isGroup ? "channel" : "direct",
		ConversationLabel: fromLabel,
		SenderName: senderName || void 0,
		SenderId: senderId,
		SenderUsername: senderEmail,
		WasMentioned: isGroup ? effectiveWasMentioned : void 0,
		CommandAuthorized: commandAuthorized,
		Provider: "googlechat",
		Surface: "googlechat",
		MessageSid: message.name,
		MessageSidFull: message.name,
		ReplyToId: message.thread?.name,
		ReplyToIdFull: message.thread?.name,
		MediaPath: mediaPath,
		MediaType: mediaType,
		MediaUrl: mediaPath,
		GroupSpace: isGroup ? space.displayName ?? void 0 : void 0,
		GroupSystemPrompt: isGroup ? groupSystemPrompt : void 0,
		OriginatingChannel: "googlechat",
		OriginatingTo: `googlechat:${spaceId}`
	});
	core.channel.session.recordSessionMetaFromInbound({
		storePath,
		sessionKey: ctxPayload.SessionKey ?? route.sessionKey,
		ctx: ctxPayload
	}).catch((err) => {
		runtime.error?.(`googlechat: failed updating session meta: ${String(err)}`);
	});
	let typingIndicator = account.config.typingIndicator ?? "message";
	if (typingIndicator === "reaction") {
		runtime.error?.(`[${account.accountId}] typingIndicator="reaction" requires user OAuth (not supported with service account). Falling back to "message" mode.`);
		typingIndicator = "message";
	}
	let typingMessageName;
	if (typingIndicator === "message") try {
		typingMessageName = (await sendGoogleChatMessage({
			account,
			space: spaceId,
			text: `_${resolveBotDisplayName({
				accountName: account.config.name,
				agentId: route.agentId,
				config
			})} is typing..._`,
			thread: message.thread?.name
		}))?.messageName;
	} catch (err) {
		runtime.error?.(`Failed sending typing message: ${String(err)}`);
	}
	const { onModelSelected, ...replyPipeline } = createChannelReplyPipeline({
		cfg: config,
		agentId: route.agentId,
		channel: "googlechat",
		accountId: route.accountId
	});
	await core.channel.reply.dispatchReplyWithBufferedBlockDispatcher({
		ctx: ctxPayload,
		cfg: config,
		dispatcherOptions: {
			...replyPipeline,
			deliver: async (payload) => {
				await deliverGoogleChatReply({
					payload,
					account,
					spaceId,
					runtime,
					core,
					config,
					statusSink,
					typingMessageName
				});
				typingMessageName = void 0;
			},
			onError: (err, info) => {
				runtime.error?.(`[${account.accountId}] Google Chat ${info.kind} reply failed: ${String(err)}`);
			}
		},
		replyOptions: { onModelSelected }
	});
}
async function downloadAttachment(attachment, account, mediaMaxMb, core) {
	const resourceName = attachment.attachmentDataRef?.resourceName;
	if (!resourceName) return null;
	const maxBytes = Math.max(1, mediaMaxMb) * 1024 * 1024;
	const downloaded = await downloadGoogleChatMedia({
		account,
		resourceName,
		maxBytes
	});
	const saved = await core.channel.media.saveMediaBuffer(downloaded.buffer, downloaded.contentType ?? attachment.contentType, "inbound", maxBytes, attachment.contentName);
	return {
		path: saved.path,
		contentType: saved.contentType
	};
}
async function deliverGoogleChatReply(params) {
	const { payload, account, spaceId, runtime, core, config, statusSink, typingMessageName } = params;
	const reply = resolveSendableOutboundReplyParts(payload);
	const mediaCount = reply.mediaCount;
	const hasMedia = reply.hasMedia;
	const text = reply.text;
	let firstTextChunk = true;
	let suppressCaption = false;
	if (hasMedia) {
		if (typingMessageName) try {
			await deleteGoogleChatMessage({
				account,
				messageName: typingMessageName
			});
		} catch (err) {
			runtime.error?.(`Google Chat typing cleanup failed: ${String(err)}`);
			const fallbackText = reply.hasText ? text : mediaCount > 1 ? "Sent attachments." : "Sent attachment.";
			try {
				await updateGoogleChatMessage({
					account,
					messageName: typingMessageName,
					text: fallbackText
				});
				suppressCaption = Boolean(text.trim());
			} catch (updateErr) {
				runtime.error?.(`Google Chat typing update failed: ${String(updateErr)}`);
			}
		}
	}
	const chunkLimit = account.config.textChunkLimit ?? 4e3;
	const chunkMode = core.channel.text.resolveChunkMode(config, "googlechat", account.accountId);
	await deliverTextOrMediaReply({
		payload,
		text: suppressCaption ? "" : reply.text,
		chunkText: (value) => core.channel.text.chunkMarkdownTextWithMode(value, chunkLimit, chunkMode),
		sendText: async (chunk) => {
			try {
				if (firstTextChunk && typingMessageName) await updateGoogleChatMessage({
					account,
					messageName: typingMessageName,
					text: chunk
				});
				else await sendGoogleChatMessage({
					account,
					space: spaceId,
					text: chunk,
					thread: payload.replyToId
				});
				firstTextChunk = false;
				statusSink?.({ lastOutboundAt: Date.now() });
			} catch (err) {
				runtime.error?.(`Google Chat message send failed: ${String(err)}`);
			}
		},
		sendMedia: async ({ mediaUrl, caption }) => {
			try {
				const loaded = await core.channel.media.fetchRemoteMedia({
					url: mediaUrl,
					maxBytes: (account.config.mediaMaxMb ?? 20) * 1024 * 1024
				});
				const upload = await uploadAttachmentForReply({
					account,
					spaceId,
					buffer: loaded.buffer,
					contentType: loaded.contentType,
					filename: loaded.fileName ?? "attachment"
				});
				if (!upload.attachmentUploadToken) throw new Error("missing attachment upload token");
				await sendGoogleChatMessage({
					account,
					space: spaceId,
					text: caption,
					thread: payload.replyToId,
					attachments: [{
						attachmentUploadToken: upload.attachmentUploadToken,
						contentName: loaded.fileName
					}]
				});
				statusSink?.({ lastOutboundAt: Date.now() });
			} catch (err) {
				runtime.error?.(`Google Chat attachment send failed: ${String(err)}`);
			}
		}
	});
}
async function uploadAttachmentForReply(params) {
	const { account, spaceId, buffer, contentType, filename } = params;
	const { uploadGoogleChatAttachment } = await import("./api-BFn4BZKr.js");
	return await uploadGoogleChatAttachment({
		account,
		space: spaceId,
		filename,
		buffer,
		contentType
	});
}
function monitorGoogleChatProvider(options) {
	const core = getGoogleChatRuntime();
	const webhookPath = resolveWebhookPath({
		webhookPath: options.webhookPath,
		webhookUrl: options.webhookUrl,
		defaultPath: "/googlechat"
	});
	if (!webhookPath) {
		options.runtime.error?.(`[${options.account.accountId}] invalid webhook path`);
		return () => {};
	}
	const audienceType = normalizeAudienceType(options.account.config.audienceType);
	const audience = options.account.config.audience?.trim();
	const mediaMaxMb = options.account.config.mediaMaxMb ?? 20;
	const unregisterTarget = registerGoogleChatWebhookTarget({
		account: options.account,
		config: options.config,
		runtime: options.runtime,
		core,
		path: webhookPath,
		audienceType,
		audience,
		statusSink: options.statusSink,
		mediaMaxMb
	});
	return () => {
		unregisterTarget();
	};
}
async function startGoogleChatMonitor(params) {
	return monitorGoogleChatProvider(params);
}
function resolveGoogleChatWebhookPath(params) {
	return resolveWebhookPath({
		webhookPath: params.account.config.webhookPath,
		webhookUrl: params.account.config.webhookUrl,
		defaultPath: "/googlechat"
	}) ?? "/googlechat";
}
//#endregion
//#region extensions/googlechat/src/channel.runtime.ts
const googleChatChannelRuntime = {
	probeGoogleChat,
	sendGoogleChatMessage,
	uploadGoogleChatAttachment,
	resolveGoogleChatWebhookPath,
	startGoogleChatMonitor
};
//#endregion
export { googleChatChannelRuntime };
