// openclaw/dist/models-cli-BwofDMyk.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import { r as theme } from "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import { m as defaultRuntime } from "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import { t as formatDocsLink } from "./links-8xRhWBQL.js";
import "./paths-C4Oy7wjk.js";
import "./auth-profiles-CLtbIyEz.js";
import "./agent-scope-bjWqU22i.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import "./model-selection-BnFtDmP7.js";
import "./io-CZZeeo8R.js";
import "./shell-env-BOu7XeT_.js";
import "./safe-text-yavot2qw.js";
import "./version-CD3oP1-d.js";
import "./env-substitution-O6uabUmO.js";
import "./includes-Bj3eLUWH.js";
import "./zod-schema.providers-core-COy5nBQ0.js";
import "./legacy-web-search-BB-ZHEhz.js";
import "./registry-C5UkPpaO.js";
import "./config-state-Br0ucqMb.js";
import "./min-host-version-RMBWtIAR.js";
import "./manifest-registry-B5JNQdOM.js";
import "./runtime-guard-PhQ6PwQa.js";
import "./avatar-policy-B5nOfso_.js";
import "./ip-Ce8EDTBZ.js";
import "./zod-schema.agent-runtime-Dtg4Jy6G.js";
import "./zod-schema.core-BuVz8Rk7.js";
import "./config-Bqe8DjNc.js";
import "./file-lock-BJsnXQjL.js";
import "./audit-fs-DTQIjp2r.js";
import "./resolve-B7S4PlnO.js";
import "./profiles-B_nscSgH.js";
import "./tailscale-CxY-91GZ.js";
import "./tailnet-DHiKSG7V.js";
import "./net-JAg9zL0G.js";
import "./auth-Buq0Niri.js";
import "./credentials-D8pfRimu.js";
import "./message-channel-BliByQBl.js";
import "./store-_QDcIMYk.js";
import "./runtime-Iz8uZ7EU.js";
import "./plugins-B09-vgme.js";
import "./sessions-BwqyiEaq.js";
import "./paths-rhN9LKM_.js";
import "./session-write-lock-Db3SWNU3.js";
import "./method-scopes-B2vIWWxl.js";
import "./call-BhjE3oIo.js";
import "./control-ui-shared-pNEAj5Ay.js";
import "./onboard-helpers-Bv9rFhNK.js";
import "./prompt-style-HkzccPBT.js";
import "./ports-lsof-Dlzvm55V.js";
import "./restart-stale-pids-OP_Aybe0.js";
import "./ports-B3yeEL-n.js";
import "./logging-D4CVa7JL.js";
import "./commands-2T-1FAo7.js";
import "./issue-format-C8oDJazD.js";
import "./progress-DlclNKc3.js";
import "./identity-D07jLUrU.js";
import "./heartbeat-B91tt3Lu.js";
import "./pi-embedded-CswW9luA.js";
import "./internal-hooks-ClBId-OS.js";
import "./multimodal-rLHaC7mC.js";
import "./memory-search-C_KRQGyn.js";
import "./provider-catalog-BGWDC_kb.js";
import "./secret-input-BOkqlW2S.js";
import "./bindings-DFhSj1AD.js";
import "./resolve-route-jw3TrvbG.js";
import "./routing-GIGwPHsU.js";
import "./identity-file-AtA39mSs.js";
import "./outbound-runtime-QVIwx-7l.js";
import "./provider-env-vars-BCe4Mf67.js";
import "./provider-auth-input-WYCWu6tD.js";
import "./provider-model-minimax-CYJqVplI.js";
import "./provider-models-ld3k2AX4.js";
import "./anthropic-vertex-provider-7stzcdx2.js";
import "./models-config.providers.discovery-DnQmXa4L.js";
import "./text-runtime-BXjeqyjp.js";
import "./tool-catalog-DP3BJZz3.js";
import "./docker-Cq_Zfc99.js";
import "./sandbox-Dt1ZOIQc.js";
import "./common-DArxwyX8.js";
import "./image-ops-339pOknK.js";
import "./thinking-rAawjv9l.js";
import "./path-alias-guards-DMcFUz5-.js";
import "./sandbox-paths-D-zPD6r4.js";
import "./channel-actions-DZIaeZzh.js";
import "./mime-C0asv0HQ.js";
import "./ssrf-BifXvxQR.js";
import "./fetch-guard-BTY_UfUY.js";
import "./provider-web-search-BHcSeZzc.js";
import "./manager-DmufLZK2.js";
import "./config-schema-CKb8N9O1.js";
import "./channel-plugin-common-BX9jRob2.js";
import "./status-helpers-D9rzemcb.js";
import "./conversation-runtime-C56HP9yG.js";
import "./runtime-whatsapp-boundary-BMepWtnk.js";
import "./pairing-store-Bh2YkvRc.js";
import "./json-store-BIglSbRG.js";
import "./secret-file-CEppYIjz.js";
import "./channel-config-schema-CrubAncY.js";
import "./setup-binary-BLFhRd_P.js";
import "./archive-Df6iZdzY.js";
import "./fs-safe-B2bSW_OI.js";
import "./signal-cli-install-UDdZA4-C.js";
import "./setup-wizard-proxy--gFyw5Eh.js";
import "./setup-C0zuBL_6.js";
import "./dm-policy-shared-ChGQzSwJ.js";
import "./hook-runtime-sGC9reX7.js";
import "./diagnostic-BtBb10Ry.js";
import "./templating-Bxm_rtRG.js";
import "./channel-reply-pipeline-DGoktEKi.js";
import "./reply-history-yYXk1ivL.js";
import "./commands-registry.data-_PzNk48W.js";
import "./commands-registry-ABOHrld2.js";
import "./frontmatter-Cd_6iJyw.js";
import "./env-overrides-B5iBORFt.js";
import "./skills-DAP6Yelo.js";
import "./skills-remote-BWQnXFJW.js";
import "./workspace-dirs-DhbPcAGw.js";
import "./pairing-token-CQLjZNmB.js";
import "./skill-commands-I_17sL0j.js";
import "./level-overrides-Ds8ppNbt.js";
import "./config-BnJ6TZJ3.js";
import "./routes-sPyjSxSY.js";
import "./ssh-tunnel-BSFzY8ID.js";
import "./server-middleware-Nz-5LEFG.js";
import "./logging-DFpZCDzJ.js";
import "./config-runtime-gAoemo4L.js";
import "./exec-approvals-CFIa5yg0.js";
import "./webhook-ingress-BBdXea0x.js";
import "./system-events-C2ngfmKm.js";
import "./ssrf-policy-BlgZt63j.js";
import "./provider-auth-ref-Dd8Ctvr9.js";
import "./provider-auth-helpers-CK3uH19P.js";
import "./provider-api-key-auth-CyO5NBeZ.js";
import "./runtime-env-CAlGDNNR.js";
import "./pairing-labels-BAkUxk5K.js";
import "./directory-runtime-BgPR38m_.js";
import "./read-only-account-inspect-B8C32NVU.js";
import "./src-Bc4Fm2Ge.js";
import "./web-media-Bo_Nxwjv.js";
import "./temp-path-Dgv6Kihk.js";
import "./media-understanding-Ch_w554d.js";
import "./web-media-BvqWsq_l.js";
import "./state-paths-C-WJBTWG.js";
import "./llm-task-Djjqflw_.js";
import "./pi-model-discovery-BNyYuKdC.js";
import "./exec-inline-eval-Di27nCtK.js";
import "./target-registry-CofRQwTE.js";
import "./external-content-Bq0zNXEh.js";
import "./brave-Hgb0hhUH.js";
import "./duckduckgo-B0xgAU7B.js";
import "./exa-Ce69v825.js";
import "./security-runtime-CoZ_sfgg.js";
import "./provider-usage-qQL8PJ0p.js";
import "./provider-onboard-CqAihpGi.js";
import "./perplexity-BNVyHCAP.js";
import "./stagger-DZXzssdr.js";
import "./command-secret-targets-wRPHRRib.js";
import "./delivery-queue-1A7pbB9e.js";
import "./channel-summary-DHY50TCU.js";
import "./follow-redirects-0oww1mdf.js";
import "./tool-policy-match-DnEs8-IP.js";
import "./runtime-CUG7pnyw.js";
import "./channel-status-C_NTcHvs.js";
import "./discord-core-DpMUVJg9.js";
import "./cli-runtime-CYhFvgcO.js";
import "./config-presence-Cqse_N2e.js";
import "./query-expansion-CZbIeYUY.js";
import "./search-manager-DsKEM_vh.js";
import "./acp-runtime-BS3aFfwU.js";
import "./telegram-core-BFUHv9d-.js";
import "./audit--q0z8yGd.js";
import "./gateway-runtime-Czf7LoCY.js";
import "./connection-auth-ZVYO9ARZ.js";
import "./mcp-config-CpCZwCwq.js";
import "./provider-auth-login-DYrAsJtB.js";
import "./auth-health-5xlVc9LU.js";
import { n as runCommandWithRuntime, t as resolveOptionFromCommand } from "./cli-utils-BqhOq4Xi.js";
import "./note-DnEr021d.js";
import "./clack-prompter-C-6ovu45.js";
import "./table-Dg7Qnbxp.js";
import { C as modelsAliasesRemoveCommand, S as modelsAliasesListCommand, _ as modelsAuthAddCommand, a as modelsListCommand, b as modelsAuthSetupTokenCommand, c as modelsImageFallbacksListCommand, d as modelsFallbacksClearCommand, f as modelsFallbacksListCommand, g as modelsAuthOrderSetCommand, h as modelsAuthOrderGetCommand, i as modelsStatusCommand, l as modelsImageFallbacksRemoveCommand, m as modelsAuthOrderClearCommand, n as modelsSetCommand, o as modelsImageFallbacksAddCommand, p as modelsFallbacksRemoveCommand, r as modelsScanCommand, s as modelsImageFallbacksClearCommand, t as modelsSetImageCommand, u as modelsFallbacksAddCommand, v as modelsAuthLoginCommand, x as modelsAliasesAddCommand, y as modelsAuthPasteTokenCommand } from "./models-CGaEg8OQ.js";
import "./provider-auth-choice-helpers-JT-z_39f.js";
import "./provider-auth-guidance-DaSGySRC.js";
//#region src/cli/models-cli.ts
function runModelsCommand(action) {
	return runCommandWithRuntime(defaultRuntime, action);
}
function registerModelsCli(program) {
	const models = program.command("models").description("Model discovery, scanning, and configuration").option("--status-json", "Output JSON (alias for `models status --json`)", false).option("--status-plain", "Plain output (alias for `models status --plain`)", false).option("--agent <id>", "Agent id to inspect (overrides OPENCLAW_AGENT_DIR/PI_CODING_AGENT_DIR)").addHelpText("after", () => `\n${theme.muted("Docs:")} ${formatDocsLink("/cli/models", "docs.openclaw.ai/cli/models")}\n`);
	models.command("list").description("List models (configured by default)").option("--all", "Show full model catalog", false).option("--local", "Filter to local models", false).option("--provider <name>", "Filter by provider").option("--json", "Output JSON", false).option("--plain", "Plain line output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsListCommand(opts, defaultRuntime);
		});
	});
	models.command("status").description("Show configured model state").option("--json", "Output JSON", false).option("--plain", "Plain output", false).option("--check", "Exit non-zero if auth is expiring/expired (1=expired/missing, 2=expiring)", false).option("--probe", "Probe configured provider auth (live)", false).option("--probe-provider <name>", "Only probe a single provider").option("--probe-profile <id>", "Only probe specific auth profile ids (repeat or comma-separated)", (value, previous) => {
		const next = Array.isArray(previous) ? previous : previous ? [previous] : [];
		next.push(value);
		return next;
	}).option("--probe-timeout <ms>", "Per-probe timeout in ms").option("--probe-concurrency <n>", "Concurrent probes").option("--probe-max-tokens <n>", "Probe max tokens (best-effort)").option("--agent <id>", "Agent id to inspect (overrides OPENCLAW_AGENT_DIR/PI_CODING_AGENT_DIR)").action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsStatusCommand({
				json: Boolean(opts.json),
				plain: Boolean(opts.plain),
				check: Boolean(opts.check),
				probe: Boolean(opts.probe),
				probeProvider: opts.probeProvider,
				probeProfile: opts.probeProfile,
				probeTimeout: opts.probeTimeout,
				probeConcurrency: opts.probeConcurrency,
				probeMaxTokens: opts.probeMaxTokens,
				agent
			}, defaultRuntime);
		});
	});
	models.command("set").description("Set the default model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsSetCommand(model, defaultRuntime);
		});
	});
	models.command("set-image").description("Set the image model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsSetImageCommand(model, defaultRuntime);
		});
	});
	const aliases = models.command("aliases").description("Manage model aliases");
	aliases.command("list").description("List model aliases").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAliasesListCommand(opts, defaultRuntime);
		});
	});
	aliases.command("add").description("Add or update a model alias").argument("<alias>", "Alias name").argument("<model>", "Model id or alias").action(async (alias, model) => {
		await runModelsCommand(async () => {
			await modelsAliasesAddCommand(alias, model, defaultRuntime);
		});
	});
	aliases.command("remove").description("Remove a model alias").argument("<alias>", "Alias name").action(async (alias) => {
		await runModelsCommand(async () => {
			await modelsAliasesRemoveCommand(alias, defaultRuntime);
		});
	});
	const fallbacks = models.command("fallbacks").description("Manage model fallback list");
	fallbacks.command("list").description("List fallback models").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsFallbacksListCommand(opts, defaultRuntime);
		});
	});
	fallbacks.command("add").description("Add a fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsFallbacksAddCommand(model, defaultRuntime);
		});
	});
	fallbacks.command("remove").description("Remove a fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsFallbacksRemoveCommand(model, defaultRuntime);
		});
	});
	fallbacks.command("clear").description("Clear all fallback models").action(async () => {
		await runModelsCommand(async () => {
			await modelsFallbacksClearCommand(defaultRuntime);
		});
	});
	const imageFallbacks = models.command("image-fallbacks").description("Manage image model fallback list");
	imageFallbacks.command("list").description("List image fallback models").option("--json", "Output JSON", false).option("--plain", "Plain output", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksListCommand(opts, defaultRuntime);
		});
	});
	imageFallbacks.command("add").description("Add an image fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksAddCommand(model, defaultRuntime);
		});
	});
	imageFallbacks.command("remove").description("Remove an image fallback model").argument("<model>", "Model id or alias").action(async (model) => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksRemoveCommand(model, defaultRuntime);
		});
	});
	imageFallbacks.command("clear").description("Clear all image fallback models").action(async () => {
		await runModelsCommand(async () => {
			await modelsImageFallbacksClearCommand(defaultRuntime);
		});
	});
	models.command("scan").description("Scan OpenRouter free models for tools + images").option("--min-params <b>", "Minimum parameter size (billions)").option("--max-age-days <days>", "Skip models older than N days").option("--provider <name>", "Filter by provider prefix").option("--max-candidates <n>", "Max fallback candidates", "6").option("--timeout <ms>", "Per-probe timeout in ms").option("--concurrency <n>", "Probe concurrency").option("--no-probe", "Skip live probes; list free candidates only").option("--yes", "Accept defaults without prompting", false).option("--no-input", "Disable prompts (use defaults)").option("--set-default", "Set agents.defaults.model to the first selection", false).option("--set-image", "Set agents.defaults.imageModel to the first image selection", false).option("--json", "Output JSON", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsScanCommand(opts, defaultRuntime);
		});
	});
	models.action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsStatusCommand({
				json: Boolean(opts?.statusJson),
				plain: Boolean(opts?.statusPlain),
				agent: opts?.agent
			}, defaultRuntime);
		});
	});
	const auth = models.command("auth").description("Manage model auth profiles");
	auth.option("--agent <id>", "Agent id for auth order get/set/clear");
	auth.action(() => {
		auth.help();
	});
	auth.command("add").description("Interactive auth helper (setup-token or paste token)").action(async () => {
		await runModelsCommand(async () => {
			await modelsAuthAddCommand({}, defaultRuntime);
		});
	});
	auth.command("login").description("Run a provider plugin auth flow (OAuth/API key)").option("--provider <id>", "Provider id registered by a plugin").option("--method <id>", "Provider auth method id").option("--set-default", "Apply the provider's default model recommendation", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthLoginCommand({
				provider: opts.provider,
				method: opts.method,
				setDefault: Boolean(opts.setDefault)
			}, defaultRuntime);
		});
	});
	auth.command("setup-token").description("Run a provider CLI to create/sync a token (TTY required)").option("--provider <name>", "Provider id (default: anthropic)").option("--yes", "Skip confirmation", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthSetupTokenCommand({
				provider: opts.provider,
				yes: Boolean(opts.yes)
			}, defaultRuntime);
		});
	});
	auth.command("paste-token").description("Paste a token into auth-profiles.json and update config").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--profile-id <id>", "Auth profile id (default: <provider>:manual)").option("--expires-in <duration>", "Optional expiry duration (e.g. 365d, 12h). Stored as absolute expiresAt.").action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthPasteTokenCommand({
				provider: opts.provider,
				profileId: opts.profileId,
				expiresIn: opts.expiresIn
			}, defaultRuntime);
		});
	});
	auth.command("login-github-copilot").description("Login to GitHub Copilot via GitHub device flow (TTY required)").option("--yes", "Overwrite existing profile without prompting", false).action(async (opts) => {
		await runModelsCommand(async () => {
			await modelsAuthLoginCommand({
				provider: "github-copilot",
				method: "device",
				yes: Boolean(opts.yes)
			}, defaultRuntime);
		});
	});
	const order = auth.command("order").description("Manage per-agent auth profile order overrides");
	order.command("get").description("Show per-agent auth order override (from auth-profiles.json)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").option("--json", "Output JSON", false).action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderGetCommand({
				provider: opts.provider,
				agent,
				json: Boolean(opts.json)
			}, defaultRuntime);
		});
	});
	order.command("set").description("Set per-agent auth order override (locks rotation to this list)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").argument("<profileIds...>", "Auth profile ids (e.g. anthropic:default)").action(async (profileIds, opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderSetCommand({
				provider: opts.provider,
				agent,
				order: profileIds
			}, defaultRuntime);
		});
	});
	order.command("clear").description("Clear per-agent auth order override (fall back to config/round-robin)").requiredOption("--provider <name>", "Provider id (e.g. anthropic)").option("--agent <id>", "Agent id (default: configured default agent)").action(async (opts, command) => {
		const agent = resolveOptionFromCommand(command, "agent") ?? opts.agent;
		await runModelsCommand(async () => {
			await modelsAuthOrderClearCommand({
				provider: opts.provider,
				agent
			}, defaultRuntime);
		});
	});
}
//#endregion
export { registerModelsCli };
