// openclaw/dist/subagent-orphan-recovery-BDybRJvH.js
import "./redact-BDinS1q9.js";
import "./errors-BxyFnvP3.js";
import "./unhandled-rejections-CDJ8dOVP.js";
import "./logger-kwZIqwuw.js";
import "./paths-ViKUYWUK.js";
import "./tmp-openclaw-dir-idKIOMmb.js";
import "./theme-CdOoMzRk.js";
import "./globals-DBUMOBZ8.js";
import { t as createSubsystemLogger } from "./subsystem-DISldKSB.js";
import "./ansi-BEJF8NKS.js";
import "./boolean-C3GkJetE.js";
import "./env-Dnra1IpT.js";
import "./warning-filter-CBhOcgHd.js";
import "./utils-CS0Ikux6.js";
import "./links-8xRhWBQL.js";
import "./paths-C4Oy7wjk.js";
import "./auth-profiles-CLtbIyEz.js";
import { u as resolveAgentIdFromSessionKey } from "./session-key-DAhnzjyr.js";
import "./agent-scope-bjWqU22i.js";
import "./boundary-path-Dm0QJ7-y.js";
import "./boundary-file-read-DcZxlWD8.js";
import "./logger-BmpSCz93.js";
import "./exec-B5_AYfQG.js";
import "./workspace-D4K6QX9X.js";
import "./model-selection-BnFtDmP7.js";
import { s as loadConfig } from "./io-CZZeeo8R.js";
import "./shell-env-BOu7XeT_.js";
import "./safe-text-yavot2qw.js";
import "./version-CD3oP1-d.js";
import "./env-substitution-O6uabUmO.js";
import "./includes-Bj3eLUWH.js";
import "./zod-schema.providers-core-COy5nBQ0.js";
import "./legacy-web-search-BB-ZHEhz.js";
import "./registry-C5UkPpaO.js";
import "./config-state-Br0ucqMb.js";
import "./min-host-version-RMBWtIAR.js";
import "./manifest-registry-B5JNQdOM.js";
import "./runtime-guard-PhQ6PwQa.js";
import "./avatar-policy-B5nOfso_.js";
import "./ip-Ce8EDTBZ.js";
import "./zod-schema.agent-runtime-Dtg4Jy6G.js";
import "./zod-schema.core-BuVz8Rk7.js";
import "./config-Bqe8DjNc.js";
import "./file-lock-BJsnXQjL.js";
import "./audit-fs-DTQIjp2r.js";
import "./resolve-B7S4PlnO.js";
import "./profiles-B_nscSgH.js";
import "./tailscale-CxY-91GZ.js";
import "./tailnet-DHiKSG7V.js";
import "./net-JAg9zL0G.js";
import "./auth-Buq0Niri.js";
import "./credentials-D8pfRimu.js";
import "./message-channel-BliByQBl.js";
import { l as updateSessionStore, n as loadSessionStore } from "./store-_QDcIMYk.js";
import "./runtime-Iz8uZ7EU.js";
import "./plugins-B09-vgme.js";
import "./sessions-BwqyiEaq.js";
import { l as resolveStorePath } from "./paths-rhN9LKM_.js";
import "./session-write-lock-Db3SWNU3.js";
import "./method-scopes-B2vIWWxl.js";
import { r as callGateway } from "./call-BhjE3oIo.js";
import "./prompt-style-HkzccPBT.js";
import "./ports-lsof-Dlzvm55V.js";
import "./restart-stale-pids-OP_Aybe0.js";
import "./ports-B3yeEL-n.js";
import "./logging-D4CVa7JL.js";
import "./commands-2T-1FAo7.js";
import "./issue-format-C8oDJazD.js";
import "./identity-D07jLUrU.js";
import "./heartbeat-B91tt3Lu.js";
import { Fo as readSessionMessages, Mc as replaceSubagentRunAfterSteer } from "./pi-embedded-CswW9luA.js";
import "./internal-hooks-ClBId-OS.js";
import "./multimodal-rLHaC7mC.js";
import "./memory-search-C_KRQGyn.js";
import "./provider-catalog-BGWDC_kb.js";
import "./secret-input-BOkqlW2S.js";
import "./bindings-DFhSj1AD.js";
import "./resolve-route-jw3TrvbG.js";
import "./routing-GIGwPHsU.js";
import "./identity-file-AtA39mSs.js";
import "./outbound-runtime-QVIwx-7l.js";
import "./provider-env-vars-BCe4Mf67.js";
import "./provider-auth-input-WYCWu6tD.js";
import "./provider-model-minimax-CYJqVplI.js";
import "./provider-models-ld3k2AX4.js";
import "./anthropic-vertex-provider-7stzcdx2.js";
import "./models-config.providers.discovery-DnQmXa4L.js";
import "./text-runtime-BXjeqyjp.js";
import "./tool-catalog-DP3BJZz3.js";
import "./docker-Cq_Zfc99.js";
import "./sandbox-Dt1ZOIQc.js";
import "./common-DArxwyX8.js";
import "./image-ops-339pOknK.js";
import "./thinking-rAawjv9l.js";
import "./path-alias-guards-DMcFUz5-.js";
import "./sandbox-paths-D-zPD6r4.js";
import "./channel-actions-DZIaeZzh.js";
import "./mime-C0asv0HQ.js";
import "./ssrf-BifXvxQR.js";
import "./fetch-guard-BTY_UfUY.js";
import "./provider-web-search-BHcSeZzc.js";
import "./manager-DmufLZK2.js";
import "./config-schema-CKb8N9O1.js";
import "./channel-plugin-common-BX9jRob2.js";
import "./status-helpers-D9rzemcb.js";
import "./conversation-runtime-C56HP9yG.js";
import "./runtime-whatsapp-boundary-BMepWtnk.js";
import "./pairing-store-Bh2YkvRc.js";
import "./json-store-BIglSbRG.js";
import "./secret-file-CEppYIjz.js";
import "./channel-config-schema-CrubAncY.js";
import "./setup-binary-BLFhRd_P.js";
import "./archive-Df6iZdzY.js";
import "./fs-safe-B2bSW_OI.js";
import "./signal-cli-install-UDdZA4-C.js";
import "./setup-wizard-proxy--gFyw5Eh.js";
import "./setup-C0zuBL_6.js";
import "./dm-policy-shared-ChGQzSwJ.js";
import "./hook-runtime-sGC9reX7.js";
import "./diagnostic-BtBb10Ry.js";
import "./templating-Bxm_rtRG.js";
import "./channel-reply-pipeline-DGoktEKi.js";
import "./reply-history-yYXk1ivL.js";
import "./commands-registry.data-_PzNk48W.js";
import "./commands-registry-ABOHrld2.js";
import "./frontmatter-Cd_6iJyw.js";
import "./env-overrides-B5iBORFt.js";
import "./skills-DAP6Yelo.js";
import "./skills-remote-BWQnXFJW.js";
import "./workspace-dirs-DhbPcAGw.js";
import "./pairing-token-CQLjZNmB.js";
import "./skill-commands-I_17sL0j.js";
import "./level-overrides-Ds8ppNbt.js";
import "./config-BnJ6TZJ3.js";
import "./routes-sPyjSxSY.js";
import "./ssh-tunnel-BSFzY8ID.js";
import "./server-middleware-Nz-5LEFG.js";
import "./logging-DFpZCDzJ.js";
import "./config-runtime-gAoemo4L.js";
import "./exec-approvals-CFIa5yg0.js";
import "./webhook-ingress-BBdXea0x.js";
import "./system-events-C2ngfmKm.js";
import "./ssrf-policy-BlgZt63j.js";
import "./provider-auth-ref-Dd8Ctvr9.js";
import "./provider-auth-helpers-CK3uH19P.js";
import "./provider-api-key-auth-CyO5NBeZ.js";
import "./runtime-env-CAlGDNNR.js";
import "./pairing-labels-BAkUxk5K.js";
import "./directory-runtime-BgPR38m_.js";
import "./read-only-account-inspect-B8C32NVU.js";
import "./src-Bc4Fm2Ge.js";
import "./web-media-Bo_Nxwjv.js";
import "./temp-path-Dgv6Kihk.js";
import "./media-understanding-Ch_w554d.js";
import "./web-media-BvqWsq_l.js";
import "./state-paths-C-WJBTWG.js";
import "./llm-task-Djjqflw_.js";
import "./pi-model-discovery-BNyYuKdC.js";
import "./exec-inline-eval-Di27nCtK.js";
import "./target-registry-CofRQwTE.js";
import "./external-content-Bq0zNXEh.js";
import "./brave-Hgb0hhUH.js";
import "./duckduckgo-B0xgAU7B.js";
import "./exa-Ce69v825.js";
import "./security-runtime-CoZ_sfgg.js";
import "./provider-usage-qQL8PJ0p.js";
import "./provider-onboard-CqAihpGi.js";
import "./perplexity-BNVyHCAP.js";
import "./stagger-DZXzssdr.js";
import "./command-secret-targets-wRPHRRib.js";
import "./delivery-queue-1A7pbB9e.js";
import "./channel-summary-DHY50TCU.js";
import "./follow-redirects-0oww1mdf.js";
import "./tool-policy-match-DnEs8-IP.js";
import "./runtime-CUG7pnyw.js";
import "./channel-status-C_NTcHvs.js";
import "./discord-core-DpMUVJg9.js";
import "./cli-runtime-CYhFvgcO.js";
import "./config-presence-Cqse_N2e.js";
import "./query-expansion-CZbIeYUY.js";
import "./search-manager-DsKEM_vh.js";
import "./acp-runtime-BS3aFfwU.js";
import "./telegram-core-BFUHv9d-.js";
import "./audit--q0z8yGd.js";
import "./gateway-runtime-Czf7LoCY.js";
import "./connection-auth-ZVYO9ARZ.js";
import "./mcp-config-CpCZwCwq.js";
import crypto from "node:crypto";
//#region src/agents/subagent-orphan-recovery.ts
/**
* Post-restart orphan recovery for subagent sessions.
*
* After a SIGUSR1 gateway reload aborts in-flight subagent LLM calls,
* this module scans for orphaned sessions (those with `abortedLastRun: true`
* that are still tracked as active in the subagent registry) and sends a
* synthetic resume message to restart their work.
*
* @see https://github.com/openclaw/openclaw/issues/47711
*/
const log = createSubsystemLogger("subagent-orphan-recovery");
/** Delay before attempting recovery to let the gateway finish bootstrapping. */
const DEFAULT_RECOVERY_DELAY_MS = 5e3;
/**
* Build the resume message for an orphaned subagent.
*/
function buildResumeMessage(task, lastHumanMessage) {
	const maxTaskLen = 2e3;
	let message = `[System] Your previous turn was interrupted by a gateway reload. Your original task was:\n\n${task.length > maxTaskLen ? `${task.slice(0, maxTaskLen)}...` : task}\n\n`;
	if (lastHumanMessage) message += `The last message from the user before the interruption was:\n\n${lastHumanMessage}\n\n`;
	message += `Please continue where you left off.`;
	return message;
}
function extractMessageText(msg) {
	if (!msg || typeof msg !== "object") return;
	const m = msg;
	if (typeof m.content === "string") return m.content;
	if (Array.isArray(m.content)) return m.content.filter((c) => typeof c === "object" && c !== null && c.type === "text" && typeof c.text === "string").map((c) => c.text).filter(Boolean).join("\n") || void 0;
}
/**
* Send a resume message to an orphaned subagent session via the gateway agent method.
*/
async function resumeOrphanedSession(params) {
	let resumeMessage = buildResumeMessage(params.task, params.lastHumanMessage);
	if (params.configChangeHint) resumeMessage += params.configChangeHint;
	try {
		const result = await callGateway({
			method: "agent",
			params: {
				message: resumeMessage,
				sessionKey: params.sessionKey,
				idempotencyKey: crypto.randomUUID(),
				deliver: false,
				lane: "subagent"
			},
			timeoutMs: 1e4
		});
		if (!replaceSubagentRunAfterSteer({
			previousRunId: params.originalRunId,
			nextRunId: result.runId,
			fallback: params.originalRun
		})) {
			log.warn(`resumed orphaned session ${params.sessionKey} but remap failed (old run already removed); treating as failure`);
			return false;
		}
		log.info(`resumed orphaned session: ${params.sessionKey}`);
		return true;
	} catch (err) {
		log.warn(`failed to resume orphaned session ${params.sessionKey}: ${String(err)}`);
		return false;
	}
}
/**
* Scan for and resume orphaned subagent sessions after a gateway restart.
*
* An orphaned session is one where:
* 1. It has an active (not ended) entry in the subagent run registry
* 2. Its session store entry has `abortedLastRun: true`
*
* For each orphaned session found, we:
* 1. Clear the `abortedLastRun` flag
* 2. Send a synthetic resume message to trigger a new LLM turn
*/
async function recoverOrphanedSubagentSessions(params) {
	const result = {
		recovered: 0,
		failed: 0,
		skipped: 0
	};
	const resumedSessionKeys = params.resumedSessionKeys ?? /* @__PURE__ */ new Set();
	const configChangePattern = /openclaw\.json|openclaw gateway restart|config\.patch/i;
	try {
		const activeRuns = params.getActiveRuns();
		if (activeRuns.size === 0) return result;
		const cfg = loadConfig();
		const storeCache = /* @__PURE__ */ new Map();
		for (const [runId, runRecord] of activeRuns.entries()) {
			if (typeof runRecord.endedAt === "number" && runRecord.endedAt > 0) continue;
			const childSessionKey = runRecord.childSessionKey?.trim();
			if (!childSessionKey) continue;
			if (resumedSessionKeys.has(childSessionKey)) {
				result.skipped++;
				continue;
			}
			try {
				const agentId = resolveAgentIdFromSessionKey(childSessionKey);
				const storePath = resolveStorePath(cfg.session?.store, { agentId });
				let store = storeCache.get(storePath);
				if (!store) {
					store = loadSessionStore(storePath);
					storeCache.set(storePath, store);
				}
				const entry = store[childSessionKey];
				if (!entry) {
					result.skipped++;
					continue;
				}
				if (!entry.abortedLastRun) {
					result.skipped++;
					continue;
				}
				log.info(`found orphaned subagent session: ${childSessionKey} (run=${runId})`);
				const messages = readSessionMessages(entry.sessionId, storePath, entry.sessionFile);
				const lastHumanMessage = [...messages].toReversed().find((msg) => msg?.role === "user");
				const configChangeDetected = messages.some((msg) => {
					if (msg?.role !== "assistant") return false;
					const text = extractMessageText(msg);
					return typeof text === "string" && configChangePattern.test(text);
				});
				if (await resumeOrphanedSession({
					sessionKey: childSessionKey,
					task: runRecord.task,
					lastHumanMessage: extractMessageText(lastHumanMessage),
					configChangeHint: configChangeDetected ? "\n\n[config changes from your previous run were already applied — do not re-modify openclaw.json or restart the gateway]" : void 0,
					originalRunId: runId,
					originalRun: runRecord
				})) {
					resumedSessionKeys.add(childSessionKey);
					try {
						await updateSessionStore(storePath, (currentStore) => {
							const current = currentStore[childSessionKey];
							if (current) {
								current.abortedLastRun = false;
								current.updatedAt = Date.now();
								currentStore[childSessionKey] = current;
							}
						});
					} catch (err) {
						log.warn(`resume succeeded but failed to update session store for ${childSessionKey}: ${String(err)}`);
					}
					result.recovered++;
				} else {
					log.warn(`resume failed for ${childSessionKey}; abortedLastRun flag preserved for retry on next restart`);
					result.failed++;
				}
			} catch (err) {
				log.warn(`error processing orphaned session ${childSessionKey}: ${String(err)}`);
				result.failed++;
			}
		}
	} catch (err) {
		log.warn(`orphan recovery scan failed: ${String(err)}`);
		if (result.failed === 0) result.failed = 1;
	}
	if (result.recovered > 0 || result.failed > 0) log.info(`orphan recovery complete: recovered=${result.recovered} failed=${result.failed} skipped=${result.skipped}`);
	return result;
}
/** Maximum number of retry attempts for orphan recovery. */
const MAX_RECOVERY_RETRIES = 3;
/** Backoff multiplier between retries (exponential). */
const RETRY_BACKOFF_MULTIPLIER = 2;
/**
* Schedule orphan recovery after a delay, with retry logic.
* The delay gives the gateway time to fully bootstrap after restart.
* If recovery fails (e.g. gateway not yet ready), retries with exponential backoff.
*/
function scheduleOrphanRecovery(params) {
	const initialDelay = params.delayMs ?? DEFAULT_RECOVERY_DELAY_MS;
	const maxRetries = params.maxRetries ?? MAX_RECOVERY_RETRIES;
	const resumedSessionKeys = /* @__PURE__ */ new Set();
	const attemptRecovery = (attempt, delay) => {
		setTimeout(() => {
			recoverOrphanedSubagentSessions({
				...params,
				resumedSessionKeys
			}).then((result) => {
				if (result.failed > 0 && attempt < maxRetries) {
					const nextDelay = delay * RETRY_BACKOFF_MULTIPLIER;
					log.info(`orphan recovery had ${result.failed} failure(s); retrying in ${nextDelay}ms (attempt ${attempt + 1}/${maxRetries})`);
					attemptRecovery(attempt + 1, nextDelay);
				}
			}).catch((err) => {
				if (attempt < maxRetries) {
					const nextDelay = delay * RETRY_BACKOFF_MULTIPLIER;
					log.warn(`scheduled orphan recovery failed: ${String(err)}; retrying in ${nextDelay}ms (attempt ${attempt + 1}/${maxRetries})`);
					attemptRecovery(attempt + 1, nextDelay);
				} else log.warn(`scheduled orphan recovery failed after ${maxRetries} retries: ${String(err)}`);
			});
		}, delay).unref?.();
	};
	attemptRecovery(0, initialDelay);
}
//#endregion
export { scheduleOrphanRecovery };
