// openclaw/dist/extensions/anthropic/index.js
import "../../redact-BDinS1q9.js";
import "../../errors-BxyFnvP3.js";
import "../../unhandled-rejections-CDJ8dOVP.js";
import "../../logger-kwZIqwuw.js";
import "../../paths-ViKUYWUK.js";
import "../../tmp-openclaw-dir-idKIOMmb.js";
import "../../theme-CdOoMzRk.js";
import "../../globals-DBUMOBZ8.js";
import "../../subsystem-DISldKSB.js";
import "../../ansi-BEJF8NKS.js";
import "../../boolean-C3GkJetE.js";
import "../../env-Dnra1IpT.js";
import "../../warning-filter-CBhOcgHd.js";
import "../../utils-CS0Ikux6.js";
import "../../links-8xRhWBQL.js";
import { s as CLAUDE_CLI_PROFILE_ID } from "../../paths-C4Oy7wjk.js";
import { g as suggestOAuthProfileIdForLegacyDefault } from "../../auth-profiles-CLtbIyEz.js";
import "../../agent-scope-bjWqU22i.js";
import "../../boundary-path-Dm0QJ7-y.js";
import "../../boundary-file-read-DcZxlWD8.js";
import "../../logger-BmpSCz93.js";
import "../../exec-B5_AYfQG.js";
import "../../workspace-D4K6QX9X.js";
import "../../model-selection-BnFtDmP7.js";
import "../../io-CZZeeo8R.js";
import "../../shell-env-BOu7XeT_.js";
import "../../safe-text-yavot2qw.js";
import "../../version-CD3oP1-d.js";
import "../../env-substitution-O6uabUmO.js";
import "../../includes-Bj3eLUWH.js";
import "../../zod-schema.providers-core-COy5nBQ0.js";
import "../../legacy-web-search-BB-ZHEhz.js";
import "../../registry-C5UkPpaO.js";
import "../../config-state-Br0ucqMb.js";
import "../../min-host-version-RMBWtIAR.js";
import "../../manifest-registry-B5JNQdOM.js";
import "../../runtime-guard-PhQ6PwQa.js";
import "../../avatar-policy-B5nOfso_.js";
import "../../ip-Ce8EDTBZ.js";
import { t as parseDurationMs } from "../../parse-duration-D6oKXEgx.js";
import "../../zod-schema.agent-runtime-Dtg4Jy6G.js";
import "../../zod-schema.core-BuVz8Rk7.js";
import "../../config-Bqe8DjNc.js";
import "../../file-lock-BJsnXQjL.js";
import "../../audit-fs-DTQIjp2r.js";
import "../../resolve-B7S4PlnO.js";
import { n as normalizeSecretInput } from "../../normalize-secret-input-BsX1agd_.js";
import { a as upsertAuthProfile, n as listProfilesForProvider } from "../../profiles-B_nscSgH.js";
import { t as formatCliCommand } from "../../command-format-DBQUhADm.js";
import "../../tailscale-CxY-91GZ.js";
import "../../tailnet-DHiKSG7V.js";
import "../../net-JAg9zL0G.js";
import "../../auth-Buq0Niri.js";
import "../../credentials-D8pfRimu.js";
import "../../message-channel-BliByQBl.js";
import "../../store-_QDcIMYk.js";
import "../../runtime-Iz8uZ7EU.js";
import "../../plugins-B09-vgme.js";
import "../../sessions-BwqyiEaq.js";
import "../../paths-rhN9LKM_.js";
import "../../session-write-lock-Db3SWNU3.js";
import "../../method-scopes-B2vIWWxl.js";
import "../../call-BhjE3oIo.js";
import "../../prompt-style-HkzccPBT.js";
import "../../ports-lsof-Dlzvm55V.js";
import "../../restart-stale-pids-OP_Aybe0.js";
import "../../ports-B3yeEL-n.js";
import "../../logging-D4CVa7JL.js";
import "../../commands-2T-1FAo7.js";
import "../../issue-format-C8oDJazD.js";
import "../../identity-D07jLUrU.js";
import "../../heartbeat-B91tt3Lu.js";
import { Uh as buildTokenProfileId, Wh as validateAnthropicSetupToken } from "../../pi-embedded-CswW9luA.js";
import "../../internal-hooks-ClBId-OS.js";
import "../../multimodal-rLHaC7mC.js";
import "../../memory-search-C_KRQGyn.js";
import "../../provider-catalog-BGWDC_kb.js";
import "../../secret-input-BOkqlW2S.js";
import "../../bindings-DFhSj1AD.js";
import "../../resolve-route-jw3TrvbG.js";
import "../../routing-GIGwPHsU.js";
import "../../identity-file-AtA39mSs.js";
import "../../outbound-runtime-QVIwx-7l.js";
import "../../provider-env-vars-BCe4Mf67.js";
import { a as normalizeSecretInputModeInput } from "../../provider-auth-input-WYCWu6tD.js";
import { p as normalizeModelCompat } from "../../provider-model-minimax-CYJqVplI.js";
import "../../provider-models-ld3k2AX4.js";
import "../../anthropic-vertex-provider-7stzcdx2.js";
import "../../models-config.providers.discovery-DnQmXa4L.js";
import "../../text-runtime-BXjeqyjp.js";
import "../../tool-catalog-DP3BJZz3.js";
import "../../docker-Cq_Zfc99.js";
import "../../sandbox-Dt1ZOIQc.js";
import "../../common-DArxwyX8.js";
import "../../image-ops-339pOknK.js";
import "../../thinking-rAawjv9l.js";
import "../../path-alias-guards-DMcFUz5-.js";
import "../../sandbox-paths-D-zPD6r4.js";
import "../../channel-actions-DZIaeZzh.js";
import "../../mime-C0asv0HQ.js";
import "../../ssrf-BifXvxQR.js";
import "../../fetch-guard-BTY_UfUY.js";
import "../../provider-web-search-BHcSeZzc.js";
import "../../manager-DmufLZK2.js";
import "../../config-schema-CKb8N9O1.js";
import "../../channel-plugin-common-BX9jRob2.js";
import "../../status-helpers-D9rzemcb.js";
import "../../conversation-runtime-C56HP9yG.js";
import "../../runtime-whatsapp-boundary-BMepWtnk.js";
import "../../pairing-store-Bh2YkvRc.js";
import "../../json-store-BIglSbRG.js";
import { t as definePluginEntry } from "../../plugin-entry-BRMgG77c.js";
import "../../secret-file-CEppYIjz.js";
import "../../channel-config-schema-CrubAncY.js";
import "../../setup-binary-BLFhRd_P.js";
import "../../archive-Df6iZdzY.js";
import "../../fs-safe-B2bSW_OI.js";
import "../../signal-cli-install-UDdZA4-C.js";
import { t as resolveSecretInputModeForEnvSelection } from "../../provider-auth-mode-8HZLxa_B.js";
import "../../setup-wizard-proxy--gFyw5Eh.js";
import "../../setup-C0zuBL_6.js";
import "../../dm-policy-shared-ChGQzSwJ.js";
import "../../hook-runtime-sGC9reX7.js";
import "../../diagnostic-BtBb10Ry.js";
import "../../templating-Bxm_rtRG.js";
import "../../channel-reply-pipeline-DGoktEKi.js";
import "../../reply-history-yYXk1ivL.js";
import "../../commands-registry.data-_PzNk48W.js";
import "../../commands-registry-ABOHrld2.js";
import "../../frontmatter-Cd_6iJyw.js";
import "../../env-overrides-B5iBORFt.js";
import "../../skills-DAP6Yelo.js";
import "../../skills-remote-BWQnXFJW.js";
import "../../workspace-dirs-DhbPcAGw.js";
import "../../pairing-token-CQLjZNmB.js";
import "../../skill-commands-I_17sL0j.js";
import "../../level-overrides-Ds8ppNbt.js";
import "../../config-BnJ6TZJ3.js";
import "../../routes-sPyjSxSY.js";
import "../../ssh-tunnel-BSFzY8ID.js";
import "../../server-middleware-Nz-5LEFG.js";
import "../../logging-DFpZCDzJ.js";
import "../../config-runtime-gAoemo4L.js";
import "../../exec-approvals-CFIa5yg0.js";
import "../../webhook-ingress-BBdXea0x.js";
import "../../system-events-C2ngfmKm.js";
import "../../ssrf-policy-BlgZt63j.js";
import { n as promptSecretRefForSetup } from "../../provider-auth-ref-Dd8Ctvr9.js";
import { t as applyAuthProfileConfig } from "../../provider-auth-helpers-CK3uH19P.js";
import { t as createProviderApiKeyAuthMethod } from "../../provider-api-key-auth-CyO5NBeZ.js";
import "../../runtime-env-CAlGDNNR.js";
import "../../pairing-labels-BAkUxk5K.js";
import "../../directory-runtime-BgPR38m_.js";
import "../../read-only-account-inspect-B8C32NVU.js";
import "../../src-Bc4Fm2Ge.js";
import "../../web-media-Bo_Nxwjv.js";
import "../../temp-path-Dgv6Kihk.js";
import { a as describeImagesWithModel, i as describeImageWithModel } from "../../media-understanding-Ch_w554d.js";
import "../../web-media-BvqWsq_l.js";
import "../../state-paths-C-WJBTWG.js";
import "../../llm-task-Djjqflw_.js";
import "../../pi-model-discovery-BNyYuKdC.js";
import "../../exec-inline-eval-Di27nCtK.js";
import "../../target-registry-CofRQwTE.js";
import "../../external-content-Bq0zNXEh.js";
import "../../brave-Hgb0hhUH.js";
import "../../duckduckgo-B0xgAU7B.js";
import "../../exa-Ce69v825.js";
import "../../security-runtime-CoZ_sfgg.js";
import { a as fetchClaudeUsage } from "../../provider-usage-qQL8PJ0p.js";
import "../../provider-onboard-CqAihpGi.js";
import "../../perplexity-BNVyHCAP.js";
import "../../stagger-DZXzssdr.js";
import "../../command-secret-targets-wRPHRRib.js";
import "../../delivery-queue-1A7pbB9e.js";
import "../../channel-summary-DHY50TCU.js";
import "../../follow-redirects-0oww1mdf.js";
import "../../tool-policy-match-DnEs8-IP.js";
import "../../runtime-CUG7pnyw.js";
import "../../channel-status-C_NTcHvs.js";
import "../../discord-core-DpMUVJg9.js";
import "../../cli-runtime-CYhFvgcO.js";
import "../../config-presence-Cqse_N2e.js";
import "../../query-expansion-CZbIeYUY.js";
import "../../search-manager-DsKEM_vh.js";
import "../../acp-runtime-BS3aFfwU.js";
import "../../telegram-core-BFUHv9d-.js";
import "../../audit--q0z8yGd.js";
import "../../gateway-runtime-Czf7LoCY.js";
import "../../connection-auth-ZVYO9ARZ.js";
import "../../mcp-config-CpCZwCwq.js";
//#region extensions/anthropic/media-understanding-provider.ts
const anthropicMediaUnderstandingProvider = {
	id: "anthropic",
	capabilities: ["image"],
	describeImage: describeImageWithModel,
	describeImages: describeImagesWithModel
};
//#endregion
//#region extensions/anthropic/index.ts
const PROVIDER_ID = "anthropic";
const DEFAULT_ANTHROPIC_MODEL = "anthropic/claude-sonnet-4-6";
const ANTHROPIC_OPUS_46_MODEL_ID = "claude-opus-4-6";
const ANTHROPIC_OPUS_46_DOT_MODEL_ID = "claude-opus-4.6";
const ANTHROPIC_OPUS_TEMPLATE_MODEL_IDS = ["claude-opus-4-5", "claude-opus-4.5"];
const ANTHROPIC_SONNET_46_MODEL_ID = "claude-sonnet-4-6";
const ANTHROPIC_SONNET_46_DOT_MODEL_ID = "claude-sonnet-4.6";
const ANTHROPIC_SONNET_TEMPLATE_MODEL_IDS = ["claude-sonnet-4-5", "claude-sonnet-4.5"];
const ANTHROPIC_MODERN_MODEL_PREFIXES = [
	"claude-opus-4-6",
	"claude-sonnet-4-6",
	"claude-opus-4-5",
	"claude-sonnet-4-5",
	"claude-haiku-4-5"
];
const ANTHROPIC_OAUTH_ALLOWLIST = [
	"anthropic/claude-sonnet-4-6",
	"anthropic/claude-opus-4-6",
	"anthropic/claude-opus-4-5",
	"anthropic/claude-sonnet-4-5",
	"anthropic/claude-haiku-4-5"
];
function cloneFirstTemplateModel(params) {
	const trimmedModelId = params.modelId.trim();
	for (const templateId of [...new Set(params.templateIds)].filter(Boolean)) {
		const template = params.ctx.modelRegistry.find(PROVIDER_ID, templateId);
		if (!template) continue;
		return normalizeModelCompat({
			...template,
			id: trimmedModelId,
			name: trimmedModelId
		});
	}
}
function resolveAnthropic46ForwardCompatModel(params) {
	const trimmedModelId = params.ctx.modelId.trim();
	const lower = trimmedModelId.toLowerCase();
	if (!(lower === params.dashModelId || lower === params.dotModelId || lower.startsWith(`${params.dashModelId}-`) || lower.startsWith(`${params.dotModelId}-`))) return;
	const templateIds = [];
	if (lower.startsWith(params.dashModelId)) templateIds.push(lower.replace(params.dashModelId, params.dashTemplateId));
	if (lower.startsWith(params.dotModelId)) templateIds.push(lower.replace(params.dotModelId, params.dotTemplateId));
	templateIds.push(...params.fallbackTemplateIds);
	return cloneFirstTemplateModel({
		modelId: trimmedModelId,
		templateIds,
		ctx: params.ctx
	});
}
function resolveAnthropicForwardCompatModel(ctx) {
	return resolveAnthropic46ForwardCompatModel({
		ctx,
		dashModelId: ANTHROPIC_OPUS_46_MODEL_ID,
		dotModelId: ANTHROPIC_OPUS_46_DOT_MODEL_ID,
		dashTemplateId: "claude-opus-4-5",
		dotTemplateId: "claude-opus-4.5",
		fallbackTemplateIds: ANTHROPIC_OPUS_TEMPLATE_MODEL_IDS
	}) ?? resolveAnthropic46ForwardCompatModel({
		ctx,
		dashModelId: ANTHROPIC_SONNET_46_MODEL_ID,
		dotModelId: ANTHROPIC_SONNET_46_DOT_MODEL_ID,
		dashTemplateId: "claude-sonnet-4-5",
		dotTemplateId: "claude-sonnet-4.5",
		fallbackTemplateIds: ANTHROPIC_SONNET_TEMPLATE_MODEL_IDS
	});
}
function matchesAnthropicModernModel(modelId) {
	const lower = modelId.trim().toLowerCase();
	return ANTHROPIC_MODERN_MODEL_PREFIXES.some((prefix) => lower.startsWith(prefix));
}
function buildAnthropicAuthDoctorHint(params) {
	const legacyProfileId = params.profileId ?? "anthropic:default";
	const suggested = suggestOAuthProfileIdForLegacyDefault({
		cfg: params.config,
		store: params.store,
		provider: PROVIDER_ID,
		legacyProfileId
	});
	if (!suggested || suggested === legacyProfileId) return "";
	const storeOauthProfiles = listProfilesForProvider(params.store, PROVIDER_ID).filter((id) => params.store.profiles[id]?.type === "oauth").join(", ");
	const cfgMode = params.config?.auth?.profiles?.[legacyProfileId]?.mode;
	const cfgProvider = params.config?.auth?.profiles?.[legacyProfileId]?.provider;
	return [
		"Doctor hint (for GitHub issue):",
		`- provider: ${PROVIDER_ID}`,
		`- config: ${legacyProfileId}${cfgProvider || cfgMode ? ` (provider=${cfgProvider ?? "?"}, mode=${cfgMode ?? "?"})` : ""}`,
		`- auth store oauth profiles: ${storeOauthProfiles || "(none)"}`,
		`- suggested profile: ${suggested}`,
		`Fix: run "${formatCliCommand("openclaw doctor --yes")}"`
	].join("\n");
}
async function runAnthropicSetupToken(ctx) {
	await ctx.prompter.note(["Run `claude setup-token` in your terminal.", "Then paste the generated token below."].join("\n"), "Anthropic setup-token");
	const requestedSecretInputMode = normalizeSecretInputModeInput(ctx.secretInputMode);
	const selectedMode = ctx.allowSecretRefPrompt ? await resolveSecretInputModeForEnvSelection({
		prompter: ctx.prompter,
		explicitMode: requestedSecretInputMode,
		copy: {
			modeMessage: "How do you want to provide this setup token?",
			plaintextLabel: "Paste setup token now",
			plaintextHint: "Stores the token directly in the auth profile"
		}
	}) : "plaintext";
	let token = "";
	let tokenRef;
	if (selectedMode === "ref") {
		const resolved = await promptSecretRefForSetup({
			provider: "anthropic-setup-token",
			config: ctx.config,
			prompter: ctx.prompter,
			preferredEnvVar: "ANTHROPIC_SETUP_TOKEN",
			copy: {
				sourceMessage: "Where is this Anthropic setup token stored?",
				envVarPlaceholder: "ANTHROPIC_SETUP_TOKEN"
			}
		});
		token = resolved.resolvedValue.trim();
		tokenRef = resolved.ref;
	} else {
		const tokenRaw = await ctx.prompter.text({
			message: "Paste Anthropic setup-token",
			validate: (value) => validateAnthropicSetupToken(String(value ?? ""))
		});
		token = String(tokenRaw ?? "").trim();
	}
	const tokenError = validateAnthropicSetupToken(token);
	if (tokenError) throw new Error(tokenError);
	const profileNameRaw = await ctx.prompter.text({
		message: "Token name (blank = default)",
		placeholder: "default"
	});
	return { profiles: [{
		profileId: buildTokenProfileId({
			provider: PROVIDER_ID,
			name: String(profileNameRaw ?? "")
		}),
		credential: {
			type: "token",
			provider: PROVIDER_ID,
			token,
			...tokenRef ? { tokenRef } : {}
		}
	}] };
}
async function runAnthropicSetupTokenNonInteractive(ctx) {
	const provider = ctx.opts.tokenProvider?.trim().toLowerCase();
	if (!provider) {
		ctx.runtime.error("Missing --token-provider for --auth-choice token.");
		ctx.runtime.exit(1);
		return null;
	}
	if (provider !== PROVIDER_ID) {
		ctx.runtime.error("Only --token-provider anthropic is supported for --auth-choice token.");
		ctx.runtime.exit(1);
		return null;
	}
	const token = normalizeSecretInput(ctx.opts.token);
	if (!token) {
		ctx.runtime.error("Missing --token for --auth-choice token.");
		ctx.runtime.exit(1);
		return null;
	}
	const tokenError = validateAnthropicSetupToken(token);
	if (tokenError) {
		ctx.runtime.error(tokenError);
		ctx.runtime.exit(1);
		return null;
	}
	let expires;
	const expiresInRaw = ctx.opts.tokenExpiresIn?.trim();
	if (expiresInRaw) try {
		expires = Date.now() + parseDurationMs(expiresInRaw, { defaultUnit: "d" });
	} catch (err) {
		ctx.runtime.error(`Invalid --token-expires-in: ${String(err)}`);
		ctx.runtime.exit(1);
		return null;
	}
	const profileId = ctx.opts.tokenProfileId?.trim() || buildTokenProfileId({
		provider: PROVIDER_ID,
		name: ""
	});
	upsertAuthProfile({
		profileId,
		agentDir: ctx.agentDir,
		credential: {
			type: "token",
			provider: PROVIDER_ID,
			token,
			...expires ? { expires } : {}
		}
	});
	return applyAuthProfileConfig(ctx.config, {
		profileId,
		provider: PROVIDER_ID,
		mode: "token"
	});
}
var anthropic_default = definePluginEntry({
	id: PROVIDER_ID,
	name: "Anthropic Provider",
	description: "Bundled Anthropic provider plugin",
	register(api) {
		api.registerProvider({
			id: PROVIDER_ID,
			label: "Anthropic",
			docsPath: "/providers/models",
			envVars: ["ANTHROPIC_OAUTH_TOKEN", "ANTHROPIC_API_KEY"],
			deprecatedProfileIds: [CLAUDE_CLI_PROFILE_ID],
			auth: [{
				id: "setup-token",
				label: "setup-token (claude)",
				hint: "Paste a setup-token from `claude setup-token`",
				kind: "token",
				wizard: {
					choiceId: "token",
					choiceLabel: "Anthropic token (paste setup-token)",
					choiceHint: "Run `claude setup-token` elsewhere, then paste the token here",
					groupId: "anthropic",
					groupLabel: "Anthropic",
					groupHint: "setup-token + API key",
					modelAllowlist: {
						allowedKeys: [...ANTHROPIC_OAUTH_ALLOWLIST],
						initialSelections: ["anthropic/claude-sonnet-4-6"],
						message: "Anthropic OAuth models"
					}
				},
				run: async (ctx) => await runAnthropicSetupToken(ctx),
				runNonInteractive: async (ctx) => await runAnthropicSetupTokenNonInteractive({
					config: ctx.config,
					opts: ctx.opts,
					runtime: ctx.runtime,
					agentDir: ctx.agentDir
				})
			}, createProviderApiKeyAuthMethod({
				providerId: PROVIDER_ID,
				methodId: "api-key",
				label: "Anthropic API key",
				hint: "Direct Anthropic API key",
				optionKey: "anthropicApiKey",
				flagName: "--anthropic-api-key",
				envVar: "ANTHROPIC_API_KEY",
				promptMessage: "Enter Anthropic API key",
				defaultModel: DEFAULT_ANTHROPIC_MODEL,
				expectedProviders: ["anthropic"],
				wizard: {
					choiceId: "apiKey",
					choiceLabel: "Anthropic API key",
					groupId: "anthropic",
					groupLabel: "Anthropic",
					groupHint: "setup-token + API key"
				}
			})],
			resolveDynamicModel: (ctx) => resolveAnthropicForwardCompatModel(ctx),
			capabilities: {
				providerFamily: "anthropic",
				dropThinkingBlockModelHints: ["claude"]
			},
			isModernModelRef: ({ modelId }) => matchesAnthropicModernModel(modelId),
			resolveDefaultThinkingLevel: ({ modelId }) => matchesAnthropicModernModel(modelId) && (modelId.toLowerCase().startsWith(ANTHROPIC_OPUS_46_MODEL_ID) || modelId.toLowerCase().startsWith(ANTHROPIC_OPUS_46_DOT_MODEL_ID) || modelId.toLowerCase().startsWith(ANTHROPIC_SONNET_46_MODEL_ID) || modelId.toLowerCase().startsWith(ANTHROPIC_SONNET_46_DOT_MODEL_ID)) ? "adaptive" : void 0,
			resolveUsageAuth: async (ctx) => await ctx.resolveOAuthToken(),
			fetchUsageSnapshot: async (ctx) => await fetchClaudeUsage(ctx.token, ctx.timeoutMs, ctx.fetchFn),
			isCacheTtlEligible: () => true,
			buildAuthDoctorHint: (ctx) => buildAnthropicAuthDoctorHint({
				config: ctx.config,
				store: ctx.store,
				profileId: ctx.profileId
			})
		});
		api.registerMediaUnderstandingProvider(anthropicMediaUnderstandingProvider);
	}
});
//#endregion
export { anthropic_default as default };
